// @strapi/content-manager/dist/admin/pages/EditView/components/FormInputs/DynamicZone/ComponentPicker.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var strapiAdmin = require('@strapi/admin/strapi-admin');
var designSystem = require('@strapi/design-system');
var reactIntl = require('react-intl');
var translations = require('../../../../../utils/translations.js');
var ComponentCategory = require('./ComponentCategory.js');

const ComponentPicker = ({ dynamicComponentsByCategory = {}, isOpen, onClickAddComponent })=>{
    const { formatMessage } = reactIntl.useIntl();
    const isMobile = strapiAdmin.useIsMobile();
    const handleAddComponentToDz = (componentUid)=>()=>{
            onClickAddComponent(componentUid);
        };
    if (!isOpen) {
        return null;
    }
    return /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
        paddingTop: {
            initial: 0,
            medium: 6
        },
        paddingBottom: {
            initial: 0,
            medium: 6
        },
        paddingLeft: {
            initial: 0,
            medium: 5
        },
        paddingRight: {
            initial: 0,
            medium: 5
        },
        background: isMobile ? 'transparent' : 'neutral0',
        shadow: isMobile ? 'none' : 'tableShadow',
        borderColor: isMobile ? 'transparent' : 'neutral150',
        hasRadius: !isMobile,
        gap: {
            initial: 4,
            medium: 2
        },
        direction: "column",
        alignItems: "stretch",
        children: [
            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
                justifyContent: "center",
                children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                    fontWeight: "bold",
                    textColor: "neutral600",
                    fontSize: 1,
                    children: formatMessage({
                        id: translations.getTranslation('components.DynamicZone.ComponentPicker-label'),
                        defaultMessage: 'Pick one component'
                    })
                })
            }),
            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Box, {
                children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Accordion.Root, {
                    defaultValue: Object.keys(dynamicComponentsByCategory)[0],
                    children: Object.entries(dynamicComponentsByCategory).map(([category, components], index)=>/*#__PURE__*/ jsxRuntime.jsx(ComponentCategory.ComponentCategory, {
                            category: category,
                            components: components,
                            onAddComponent: handleAddComponentToDz,
                            variant: index % 2 === 1 ? 'primary' : 'secondary'
                        }, category))
                })
            })
        ]
    });
};

exports.ComponentPicker = ComponentPicker;
