// @strapi/content-manager/dist/admin/components/ActionsDrawer.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var strapiAdmin = require('@strapi/admin/strapi-admin');
var designSystem = require('@strapi/design-system');
var Icons = require('@strapi/icons');
var reactIntl = require('react-intl');
var styledComponents = require('styled-components');

function _interopNamespaceDefault(e) {
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespaceDefault(React);

const [ActionsDrawerProvider, useActionsDrawer] = strapiAdmin.createContext('ActionsDrawer', null);
/* -------------------------------------------------------------------------------------------------
 * Styled Components
 * -----------------------------------------------------------------------------------------------*/ const DrawerContainer = styledComponents.styled(designSystem.Portal)`
  position: fixed;
  left: 0;
  right: 0;
  bottom: 0;
  z-index: ${({ $isOpen })=>$isOpen ? 4 : 2};
  display: flex;
  flex-direction: column;
  align-items: stretch;
  justify-content: flex-end;

  ${({ theme })=>theme.breakpoints.medium} {
    z-index: 2;
    left: ${({ $hasSideNav })=>$hasSideNav ? strapiAdmin.WIDTH_SIDE_NAVIGATION : 0};
  }
`;
const DrawerOverlay = styledComponents.styled(designSystem.Box)`
  position: fixed;
  left: 0;
  top: 0;
  right: 0;
  bottom: 0;
  background-color: rgba(0, 0, 0, 0.5);
  transition: opacity ${(props)=>props.theme.motion.timings['200']}
    ${(props)=>props.theme.motion.easings.easeInOutQuad};
  opacity: ${({ $isOpen })=>$isOpen ? 1 : 0};
  pointer-events: ${({ $isOpen })=>$isOpen ? 'auto' : 'none'};

  ${({ theme })=>theme.breakpoints.medium} {
    left: ${({ $hasSideNav })=>$hasSideNav ? strapiAdmin.WIDTH_SIDE_NAVIGATION : 0};
  }
`;
const ToggleButton = styledComponents.styled(designSystem.IconButton)`
  padding: 0;
  border: none;
  background: ${({ theme })=>theme.colors.neutral200};
  width: 3.2rem;
  height: 3.2rem;
  border-radius: 1.6rem;
  display: flex;
  align-items: center;
  justify-content: center;
`;
const DrawerContent = styledComponents.styled(designSystem.Flex)`
  flex-direction: column;
  align-items: stretch;
  position: relative;
  z-index: 1;
  pointer-events: auto;
`;
const DrawerContentInner = styledComponents.styled(designSystem.ScrollArea)`
  max-height: ${({ $isOpen })=>$isOpen ? 'calc(100vh - 25rem)' // 25rem is arbitrary, to be able to see a bit of the content behind (navigation and header)
     : '0'};
  overflow: hidden;
  transition: max-height ${(props)=>props.theme.motion.timings['200']}
    ${(props)=>props.theme.motion.easings.easeInOutQuad};
  z-index: 1;
  pointer-events: auto;
`;
const Root = ({ children, hasContent = false, hasSideNav = false })=>{
    const [isOpen, setIsOpen] = React__namespace.useState(false);
    // Close drawer when there's no content
    React__namespace.useEffect(()=>{
        if (!hasContent) {
            setIsOpen(false);
        }
    }, [
        hasContent
    ]);
    return /*#__PURE__*/ jsxRuntime.jsx(ActionsDrawerProvider, {
        isOpen: isOpen,
        setIsOpen: setIsOpen,
        hasContent: hasContent,
        hasSideNav: hasSideNav,
        children: /*#__PURE__*/ jsxRuntime.jsx(DrawerContainer, {
            $hasSideNav: hasSideNav,
            $isOpen: isOpen,
            children: children
        })
    });
};
/* -------------------------------------------------------------------------------------------------
 * Overlay
 * -----------------------------------------------------------------------------------------------*/ const Overlay = ()=>{
    const ctx = useActionsDrawer('ActionsDrawer.Overlay', (s)=>s, false);
    const isOpen = ctx?.isOpen ?? false;
    const hasContent = ctx?.hasContent ?? false;
    const hasSideNav = ctx?.hasSideNav ?? false;
    const setIsOpen = ctx?.setIsOpen;
    if (!hasContent) {
        return null;
    }
    return /*#__PURE__*/ jsxRuntime.jsx(DrawerOverlay, {
        $isOpen: isOpen,
        $hasSideNav: hasSideNav,
        onClick: ()=>setIsOpen?.(false),
        "data-testid": "actions-drawer-overlay"
    });
};
const Header = ({ children })=>{
    const ctx = useActionsDrawer('ActionsDrawer.Header', (s)=>s, false);
    const isOpen = ctx?.isOpen ?? false;
    const hasContent = ctx?.hasContent ?? false;
    const setIsOpen = ctx?.setIsOpen;
    const { formatMessage } = reactIntl.useIntl();
    const toggleOpen = ()=>{
        setIsOpen?.((prev)=>!prev);
    };
    return /*#__PURE__*/ jsxRuntime.jsx(DrawerContent, {
        background: "neutral0",
        children: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
            paddingTop: 3,
            paddingBottom: 3,
            paddingLeft: 4,
            paddingRight: 4,
            gap: 3,
            borderStyle: "solid",
            borderWidth: isOpen ? '1px 0' : '1px 0 0 0',
            borderColor: "neutral150",
            children: [
                /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
                    flex: 1,
                    gap: 2,
                    alignItems: "center",
                    overflow: "hidden",
                    children: children
                }),
                hasContent && /*#__PURE__*/ jsxRuntime.jsx(ToggleButton, {
                    onClick: toggleOpen,
                    label: isOpen ? formatMessage({
                        id: 'content-manager.actions-drawer.close',
                        defaultMessage: 'Close more actions'
                    }) : formatMessage({
                        id: 'content-manager.actions-drawer.open',
                        defaultMessage: 'Open more actions'
                    }),
                    children: isOpen ? /*#__PURE__*/ jsxRuntime.jsx(Icons.CaretUp, {
                        fill: "neutral600"
                    }) : /*#__PURE__*/ jsxRuntime.jsx(Icons.CaretDown, {
                        fill: "neutral600"
                    })
                })
            ]
        })
    });
};
const Content = ({ children })=>{
    const ctx = useActionsDrawer('ActionsDrawer.Content', (s)=>s, false);
    const isOpen = ctx?.isOpen ?? false;
    return /*#__PURE__*/ jsxRuntime.jsx(DrawerContentInner, {
        $isOpen: isOpen,
        children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
            direction: "column",
            alignItems: "stretch",
            justifyContent: "flex-start",
            padding: {
                initial: 4,
                large: 0
            },
            background: "neutral0",
            children: children
        })
    });
};
/* -------------------------------------------------------------------------------------------------
 * ActionsDrawer
 * -----------------------------------------------------------------------------------------------*/ const ActionsDrawer = {
    Root,
    Overlay,
    Header,
    Content
};

exports.ActionsDrawer = ActionsDrawer;
exports.useActionsDrawer = useActionsDrawer;
