// @strapi/admin/dist/server/server/src/services/permission/permissions-manager/sanitize.js
'use strict';

var ability = require('@casl/ability');
var fp = require('lodash/fp');
var utils = require('@strapi/utils');
var permissionFields = require('./permission-fields.js');
var user = require('../../../domain/user.js');

const { visitors: { removePassword, expandWildcardPopulate } } = utils.sanitize;
const { constants, isScalarAttribute, getNonVisibleAttributes, getNonWritableAttributes, getWritableAttributes } = utils.contentTypes;
const { ID_ATTRIBUTE, DOC_ID_ATTRIBUTE, CREATED_AT_ATTRIBUTE, UPDATED_AT_ATTRIBUTE, PUBLISHED_AT_ATTRIBUTE, CREATED_BY_ATTRIBUTE, UPDATED_BY_ATTRIBUTE } = constants;
const COMPONENT_FIELDS = [
    '__component'
];
const STATIC_FIELDS = [
    ID_ATTRIBUTE,
    DOC_ID_ATTRIBUTE
];
var createSanitizeHelpers = (({ action, ability: ability$1, model })=>{
    const schema = strapi.getModel(model);
    const { removeDisallowedFields } = utils.sanitize.visitors;
    // Create request-scoped model cache to avoid redundant getModel() calls
    const modelCache = utils.createModelCache(strapi.getModel.bind(strapi));
    const ctx = {
        schema,
        getModel: modelCache.getModel
    };
    const createSanitizeQuery = (options = {})=>{
        const { fields } = options;
        // TODO: sanitize relations to admin users in all sanitizers
        const permittedFields = fields.shouldIncludeAll ? null : getQueryFields(fields.permitted);
        const createSanitizeFilters = (ctx)=>utils.async.pipe(utils.traverse.traverseQueryFilters(removeDisallowedFields(permittedFields), ctx), utils.traverse.traverseQueryFilters(omitDisallowedAdminUserFields, ctx), utils.traverse.traverseQueryFilters(omitHiddenFields, ctx), utils.traverse.traverseQueryFilters(removePassword, ctx), utils.traverse.traverseQueryFilters(({ key, value }, { remove })=>{
                if (fp.isObject(value) && fp.isEmpty(value)) {
                    remove(key);
                }
            }, ctx));
        const createSanitizeSort = (ctx)=>utils.async.pipe(utils.traverse.traverseQuerySort(removeDisallowedFields(permittedFields), ctx), utils.traverse.traverseQuerySort(omitDisallowedAdminUserFields, ctx), utils.traverse.traverseQuerySort(omitHiddenFields, ctx), utils.traverse.traverseQuerySort(removePassword, ctx), utils.traverse.traverseQuerySort(({ key, attribute, value }, { remove })=>{
                if (!isScalarAttribute(attribute) && fp.isEmpty(value)) {
                    remove(key);
                }
            }, ctx));
        const createSanitizeFields = (ctx)=>utils.async.pipe(utils.traverse.traverseQueryFields(removeDisallowedFields(permittedFields), ctx), utils.traverse.traverseQueryFields(omitHiddenFields, ctx), utils.traverse.traverseQueryFields(removePassword, ctx));
        const sanitizeFilters = createSanitizeFilters(ctx);
        const sanitizeSort = createSanitizeSort(ctx);
        const sanitizeFields = createSanitizeFields(ctx);
        /** Sanitize nested filters, sort, and fields inside populate. */ const sanitizeNestedPopulate = async ({ key, value, schema, attribute, getModel }, { set })=>{
            if (attribute) {
                return;
            }
            const nestedCtx = {
                schema,
                getModel
            };
            if (key === 'sort') {
                const sanitizedSort = await createSanitizeSort(nestedCtx)(value);
                set(key, sanitizedSort);
            }
            if (key === 'filters') {
                const sanitizedFilters = await createSanitizeFilters(nestedCtx)(value);
                set(key, sanitizedFilters);
            }
            if (key === 'fields') {
                const sanitizedFields = await createSanitizeFields(nestedCtx)(value);
                set(key, sanitizedFields);
            }
        };
        const sanitizePopulate = utils.async.pipe(utils.traverse.traverseQueryPopulate(expandWildcardPopulate, ctx), utils.traverse.traverseQueryPopulate(removeDisallowedFields(permittedFields), ctx), utils.traverse.traverseQueryPopulate(omitDisallowedAdminUserFields, ctx), utils.traverse.traverseQueryPopulate(omitHiddenFields, ctx), utils.traverse.traverseQueryPopulate(removePassword, ctx), utils.traverse.traverseQueryPopulate(sanitizeNestedPopulate, ctx));
        return async (query)=>{
            const sanitizedQuery = fp.cloneDeep(query);
            const [sanitizedFilters, sanitizedSort, sanitizedPopulate, sanitizedFields] = await Promise.all([
                query.filters ? sanitizeFilters(query.filters) : undefined,
                query.sort ? sanitizeSort(query.sort) : undefined,
                query.populate ? sanitizePopulate(query.populate) : undefined,
                query.fields ? sanitizeFields(query.fields) : undefined
            ]);
            if (sanitizedFilters !== undefined) sanitizedQuery.filters = sanitizedFilters;
            if (sanitizedSort !== undefined) sanitizedQuery.sort = sanitizedSort;
            if (sanitizedPopulate !== undefined) sanitizedQuery.populate = sanitizedPopulate;
            if (sanitizedFields !== undefined) sanitizedQuery.fields = sanitizedFields;
            return sanitizedQuery;
        };
    };
    const createSanitizeOutput = (options = {})=>{
        const { fields } = options;
        const permittedFields = fields.shouldIncludeAll ? null : getOutputFields(fields.permitted);
        return utils.async.pipe(// Remove fields hidden from the admin
        utils.traverseEntity(omitHiddenFields, ctx), // Remove unallowed fields from admin::user relations
        utils.traverseEntity(pickAllowedAdminUserFields, ctx), // Remove not allowed fields (RBAC)
        utils.traverseEntity(removeDisallowedFields(permittedFields), ctx), // Remove all fields of type 'password'
        utils.sanitize.sanitizers.sanitizePasswords({
            schema,
            getModel (uid) {
                return strapi.getModel(uid);
            }
        }));
    };
    const createSanitizeInput = (options = {})=>{
        const { fields } = options;
        const permittedFields = fields.shouldIncludeAll ? null : getInputFields(fields.permitted);
        return utils.async.pipe(// Remove fields hidden from the admin
        utils.traverseEntity(omitHiddenFields, ctx), // Remove not allowed fields (RBAC)
        utils.traverseEntity(removeDisallowedFields(permittedFields), ctx), // Remove roles from createdBy & updatedBy fields
        omitCreatorRoles);
    };
    const wrapSanitize = (createSanitizeFunction)=>{
        const { getPermissionFields } = permissionFields.createPermissionFieldsCache(ability$1);
        // TODO
        // @ts-expect-error define the correct return type
        const wrappedSanitize = async (data, options = {})=>{
            if (fp.isArray(data)) {
                return Promise.all(data.map((entity)=>wrappedSanitize(entity, options)));
            }
            const { subject, action: actionOverride } = getDefaultOptions(data, options);
            const { permittedFields, hasAtLeastOneRegistered, shouldIncludeAll } = getPermissionFields(actionOverride, subject);
            const sanitizeOptions = {
                ...options,
                fields: {
                    shouldIncludeAll,
                    permitted: permittedFields,
                    hasAtLeastOneRegistered
                }
            };
            const sanitizeFunction = createSanitizeFunction(sanitizeOptions);
            return sanitizeFunction(data);
        };
        return wrappedSanitize;
    };
    const getDefaultOptions = (data, options)=>{
        return fp.defaults({
            subject: ability.subject(model, data),
            action
        }, options);
    };
    /**
   * Omit creator fields' (createdBy & updatedBy) roles from the admin API responses
   */ const omitCreatorRoles = fp.omit([
        `${CREATED_BY_ATTRIBUTE}.roles`,
        `${UPDATED_BY_ATTRIBUTE}.roles`
    ]);
    /**
   * Visitor used to remove hidden fields from the admin API responses
   */ const omitHiddenFields = ({ key, schema }, { remove })=>{
        const isHidden = fp.getOr(false, [
            'config',
            'attributes',
            key,
            'hidden'
        ], schema);
        if (isHidden) {
            remove(key);
        }
    };
    /**
   * Visitor used to only select needed fields from the admin users entities & avoid leaking sensitive information
   */ const pickAllowedAdminUserFields = ({ attribute, key, value }, { set })=>{
        const pickAllowedFields = fp.pick(user.ADMIN_USER_ALLOWED_FIELDS);
        if (!attribute) {
            return;
        }
        if (attribute.type === 'relation' && attribute.target === 'admin::user' && value) {
            if (Array.isArray(value)) {
                set(key, value.map(pickAllowedFields));
            } else {
                set(key, pickAllowedFields(value));
            }
        }
    };
    /**
   * Visitor used to omit disallowed fields from the admin users entities & avoid leaking sensitive information
   */ const omitDisallowedAdminUserFields = ({ key, attribute, schema }, { remove })=>{
        if (schema.uid === 'admin::user' && attribute && !user.ADMIN_USER_ALLOWED_FIELDS.includes(key)) {
            remove(key);
        }
    };
    const getInputFields = (fields = [])=>{
        const nonVisibleAttributes = getNonVisibleAttributes(schema);
        const writableAttributes = getWritableAttributes(schema);
        const nonVisibleWritableAttributes = fp.intersection(nonVisibleAttributes, writableAttributes);
        return fp.uniq([
            ...fields,
            ...COMPONENT_FIELDS,
            ...nonVisibleWritableAttributes
        ]);
    };
    const getOutputFields = (fields = [])=>{
        const nonWritableAttributes = getNonWritableAttributes(schema);
        const nonVisibleAttributes = getNonVisibleAttributes(schema);
        return fp.uniq([
            ...fields,
            ...STATIC_FIELDS,
            ...COMPONENT_FIELDS,
            ...nonWritableAttributes,
            ...nonVisibleAttributes,
            CREATED_AT_ATTRIBUTE,
            UPDATED_AT_ATTRIBUTE
        ]);
    };
    const getQueryFields = (fields = [])=>{
        const nonVisibleAttributes = getNonVisibleAttributes(schema);
        const writableAttributes = getWritableAttributes(schema);
        const nonVisibleWritableAttributes = fp.intersection(nonVisibleAttributes, writableAttributes);
        return fp.uniq([
            ...fields,
            ...STATIC_FIELDS,
            ...COMPONENT_FIELDS,
            ...nonVisibleWritableAttributes,
            CREATED_AT_ATTRIBUTE,
            UPDATED_AT_ATTRIBUTE,
            PUBLISHED_AT_ATTRIBUTE,
            CREATED_BY_ATTRIBUTE,
            UPDATED_BY_ATTRIBUTE
        ]);
    };
    return {
        sanitizeOutput: wrapSanitize(createSanitizeOutput),
        sanitizeInput: wrapSanitize(createSanitizeInput),
        sanitizeQuery: wrapSanitize(createSanitizeQuery)
    };
});

module.exports = createSanitizeHelpers;
