// @strapi/content-releases/dist/server/index.js
'use strict';

var register = require('./register.js');
var bootstrap = require('./bootstrap.js');
var destroy = require('./destroy.js');
var index$4 = require('./content-types/index.js');
var index$3 = require('./services/index.js');
var index$2 = require('./controllers/index.js');
var index$1 = require('./routes/index.js');

const getPlugin = ()=>{
    if (strapi.ee.features.isEnabled('cms-content-releases')) {
        return {
            register: register.register,
            bootstrap: bootstrap.bootstrap,
            destroy: destroy.destroy,
            contentTypes: index$4.contentTypes,
            services: index$3.services,
            controllers: index$2.controllers,
            routes: index$1.routes
        };
    }
    return {
        // Always return register, it handles its own feature check
        register: register.register,
        // Always return contentTypes to avoid losing data when the feature is disabled
        contentTypes: index$4.contentTypes
    };
};
var index = getPlugin();

module.exports = index;
