// @strapi/admin@5.42.1/dist/admin/admin/src/components/Layouts/ActionLayout.mjs
import { jsxs, jsx } from 'react/jsx-runtime';
import 'react';
import { Flex } from '@strapi/design-system';
import { styled } from 'styled-components';
import { RESPONSIVE_DEFAULT_SPACING } from '../../constants/theme.mjs';

/** On mobile: first item stays left, rest stick to the right (margin-right: auto on first child). */ const EndActionsFlex = styled(Flex)`
  width: 100%;
  & > *:first-child {
    margin-right: auto;
  }

  ${({ theme })=>theme.breakpoints.medium} {
    width: auto;
    & > *:first-child {
      margin-right: 0;
    }
  }
`;
const ActionLayout = ({ startActions, endActions, bottomActions })=>{
    if (!startActions && !endActions) {
        return null;
    }
    return /*#__PURE__*/ jsxs(Flex, {
        direction: {
            initial: 'column',
            medium: 'row'
        },
        gap: {
            initial: 4,
            medium: 3
        },
        alignItems: "flex-start",
        paddingBottom: {
            initial: 2,
            medium: 4
        },
        paddingTop: 0,
        paddingLeft: RESPONSIVE_DEFAULT_SPACING,
        paddingRight: RESPONSIVE_DEFAULT_SPACING,
        wrap: {
            initial: 'nowrap',
            medium: 'wrap'
        },
        children: [
            /*#__PURE__*/ jsx(Flex, {
                gap: 2,
                wrap: "wrap",
                flex: 1,
                width: "100%",
                children: startActions
            }),
            /*#__PURE__*/ jsx(EndActionsFlex, {
                gap: {
                    initial: 3,
                    medium: 2
                },
                shrink: 0,
                wrap: "wrap",
                children: endActions
            }),
            bottomActions && /*#__PURE__*/ jsx(Flex, {
                gap: 2,
                wrap: "wrap",
                width: "100%",
                children: bottomActions
            })
        ]
    });
};

export { ActionLayout };
