// @strapi/upload@5.42.1/dist/admin/future/pages/Assets/components/AssetsGrid.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var designSystem = require('@strapi/design-system');
var icons = require('@strapi/icons');
var reactIntl = require('react-intl');
var styledComponents = require('styled-components');
var enums = require('../../../enums.js');
var files = require('../../../utils/files.js');
var getAssetIcon = require('../../../utils/getAssetIcon.js');
var translations = require('../../../utils/translations.js');
var useFolderNavigation = require('../hooks/useFolderNavigation.js');

/* -------------------------------------------------------------------------------------------------
 * AssetsGrid
 * -----------------------------------------------------------------------------------------------*/ const StyledCard = styledComponents.styled(designSystem.Card)`
  border: 1px solid ${({ theme })=>theme.colors.neutral200};
  border-radius: 8px;
  overflow: hidden;
  cursor: pointer;

  &:hover {
    background: ${({ theme })=>theme.colors.primary100};
  }

  &:focus-visible {
    outline: 2px solid ${({ theme })=>theme.colors.primary600};
    outline-offset: 2px;
  }
`;
/* -------------------------------------------------------------------------------------------------
 * FolderCard
 * -----------------------------------------------------------------------------------------------*/ const FoldersRow = styledComponents.styled(designSystem.Box)`
  grid-column: 1 / -1;
`;
const StyledFolderCard = styledComponents.styled(designSystem.Flex)`
  width: 100%;
  padding: ${({ theme })=>`${theme.spaces[2]} ${theme.spaces[3]}`}; // 8px 12px
  align-items: center;
  gap: ${({ theme })=>theme.spaces[2]}; // 8px
  border: 1px solid ${({ theme })=>theme.colors.neutral200};
  border-radius: ${({ theme })=>theme.borderRadius};
  background: ${({ theme })=>theme.colors.neutral0};
  cursor: pointer;
  transition: background 0.2s;

  &:hover {
    background: ${({ theme })=>theme.colors.primary100};
  }

  &:focus-visible {
    outline: 2px solid ${({ theme })=>theme.colors.primary600};
    outline-offset: 2px;
  }
`;
const FolderIconContainer = styledComponents.styled(designSystem.Flex)`
  flex-shrink: 0;
  color: ${({ theme })=>theme.colors.neutral600};
`;
const FolderName = styledComponents.styled(designSystem.Typography)`
  flex: 1;
  min-width: 0;
`;
const FolderCard = ({ folder })=>{
    const { formatMessage } = reactIntl.useIntl();
    const { navigateToFolder } = useFolderNavigation.useFolderNavigation();
    const handleKeyDown = (e)=>{
        if (e.key === 'Enter' || e.key === ' ') {
            e.preventDefault();
            navigateToFolder(folder);
        }
    };
    return /*#__PURE__*/ jsxRuntime.jsxs(StyledFolderCard, {
        onClick: ()=>navigateToFolder(folder),
        onKeyDown: handleKeyDown,
        role: "listitem",
        tabIndex: 0,
        children: [
            /*#__PURE__*/ jsxRuntime.jsx(FolderIconContainer, {
                children: /*#__PURE__*/ jsxRuntime.jsx(icons.Folder, {
                    width: 20,
                    height: 20
                })
            }),
            /*#__PURE__*/ jsxRuntime.jsx(FolderName, {
                textColor: "neutral800",
                ellipsis: true,
                children: folder.name
            }),
            /*#__PURE__*/ jsxRuntime.jsx(designSystem.IconButton, {
                label: formatMessage({
                    id: translations.getTranslationKey('control-card.more-actions'),
                    defaultMessage: 'More actions'
                }),
                variant: "ghost",
                onClick: (e)=>e.stopPropagation(),
                children: /*#__PURE__*/ jsxRuntime.jsx(icons.More, {})
            })
        ]
    });
};
/* -------------------------------------------------------------------------------------------------
 * AssetPreview
 * -----------------------------------------------------------------------------------------------*/ const PreviewContainer = styledComponents.styled(designSystem.Box)`
  position: relative;
  width: 100%;
  padding-bottom: 62.5%;
  height: 0;
  overflow: hidden;
  background: repeating-conic-gradient(
      ${({ theme })=>theme.colors.neutral100} 0% 25%,
      transparent 0% 50%
    )
    50% / 20px 20px;
`;
const StyledImage = styledComponents.styled.img`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  object-fit: cover;
`;
const IconPreview = styledComponents.styled(designSystem.Flex)`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  color: ${({ theme })=>theme.colors.neutral500};
  background: ${({ theme })=>theme.colors.neutral100};
`;
const AssetPreview = ({ asset })=>{
    const { alternativeText, ext, formats, mime, url } = asset;
    if (mime?.includes(enums.AssetType.Image)) {
        const mediaURL = files.prefixFileUrlWithBackendUrl(formats?.thumbnail?.url) ?? files.prefixFileUrlWithBackendUrl(url);
        if (mediaURL) {
            return /*#__PURE__*/ jsxRuntime.jsx(PreviewContainer, {
                children: /*#__PURE__*/ jsxRuntime.jsx(StyledImage, {
                    src: mediaURL,
                    alt: alternativeText || ''
                })
            });
        }
    }
    const DocIcon = getAssetIcon.getAssetIcon(mime, ext);
    return /*#__PURE__*/ jsxRuntime.jsx(PreviewContainer, {
        children: /*#__PURE__*/ jsxRuntime.jsx(IconPreview, {
            justifyContent: "center",
            alignItems: "center",
            children: /*#__PURE__*/ jsxRuntime.jsx(DocIcon, {
                width: 48,
                height: 48
            })
        })
    });
};
/* -------------------------------------------------------------------------------------------------
 * AssetCard
 * -----------------------------------------------------------------------------------------------*/ const StyledCardHeader = styledComponents.styled(designSystem.CardHeader)`
  border-bottom: 1px solid ${({ theme })=>theme.colors.neutral200};
`;
const CardFooter = styledComponents.styled(designSystem.Flex)`
  min-width: 0;
  width: 100%;
`;
const FileTypeIcon = styledComponents.styled(designSystem.Flex)`
  color: ${({ theme })=>theme.colors.neutral600};
  flex-shrink: 0;
`;
const FileName = styledComponents.styled(designSystem.Typography)`
  flex: 1;
  min-width: 0;
`;
const AssetCard = ({ asset, onAssetItemClick })=>{
    const { formatMessage } = reactIntl.useIntl();
    const TypeIcon = getAssetIcon.getAssetIcon(asset.mime, asset.ext);
    const handleKeyDown = (e)=>{
        if (e.key === 'Enter' || e.key === ' ') {
            e.preventDefault();
            onAssetItemClick(asset.id);
        }
    };
    return /*#__PURE__*/ jsxRuntime.jsxs(StyledCard, {
        tabIndex: 0,
        role: "listitem",
        onClick: ()=>onAssetItemClick(asset.id),
        onKeyDown: handleKeyDown,
        children: [
            /*#__PURE__*/ jsxRuntime.jsx(StyledCardHeader, {
                children: /*#__PURE__*/ jsxRuntime.jsx(AssetPreview, {
                    asset: asset
                })
            }),
            /*#__PURE__*/ jsxRuntime.jsx(designSystem.CardBody, {
                children: /*#__PURE__*/ jsxRuntime.jsxs(CardFooter, {
                    alignItems: "center",
                    gap: 2,
                    children: [
                        /*#__PURE__*/ jsxRuntime.jsx(FileTypeIcon, {
                            children: /*#__PURE__*/ jsxRuntime.jsx(TypeIcon, {
                                width: 20,
                                height: 20
                            })
                        }),
                        /*#__PURE__*/ jsxRuntime.jsx(FileName, {
                            textColor: "primary800",
                            ellipsis: true,
                            children: asset.name
                        }),
                        /*#__PURE__*/ jsxRuntime.jsx(designSystem.IconButton, {
                            label: formatMessage({
                                id: translations.getTranslationKey('control-card.more-actions'),
                                defaultMessage: 'More actions'
                            }),
                            variant: "ghost",
                            children: /*#__PURE__*/ jsxRuntime.jsx(icons.More, {})
                        })
                    ]
                })
            })
        ]
    });
};
const AssetsGrid = ({ assets, folders = [], onAssetItemClick })=>{
    const { formatMessage } = reactIntl.useIntl();
    const totalItems = folders.length + assets.length;
    if (totalItems === 0) {
        return /*#__PURE__*/ jsxRuntime.jsx(designSystem.Box, {
            padding: 8,
            children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                textColor: "neutral600",
                children: formatMessage({
                    id: 'app.components.EmptyStateLayout.content-document',
                    defaultMessage: 'No content found'
                })
            })
        });
    }
    return /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Grid.Root, {
        gap: 4,
        role: "list",
        children: [
            folders.length > 0 && /*#__PURE__*/ jsxRuntime.jsx(FoldersRow, {
                children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Grid.Root, {
                    gap: 4,
                    children: folders.map((folder)=>/*#__PURE__*/ jsxRuntime.jsx(designSystem.Grid.Item, {
                            col: 3,
                            m: 4,
                            s: 6,
                            xs: 12,
                            children: /*#__PURE__*/ jsxRuntime.jsx(FolderCard, {
                                folder: folder
                            })
                        }, `folder-${folder.id}`))
                })
            }),
            assets.map((asset)=>/*#__PURE__*/ jsxRuntime.jsx(designSystem.Grid.Item, {
                    col: 3,
                    m: 4,
                    s: 6,
                    xs: 12,
                    direction: "column",
                    alignItems: "stretch",
                    children: /*#__PURE__*/ jsxRuntime.jsx(AssetCard, {
                        asset: asset,
                        onAssetItemClick: onAssetItemClick
                    })
                }, asset.id))
        ]
    });
};

exports.AssetsGrid = AssetsGrid;
