// @strapi/upload@5.42.1/dist/admin/future/pages/Assets/components/AssetsTable.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var strapiAdmin = require('@strapi/admin/strapi-admin');
var designSystem = require('@strapi/design-system');
var icons = require('@strapi/icons');
var reactIntl = require('react-intl');
var styledComponents = require('styled-components');
var files = require('../../../utils/files.js');
var getAssetIcon = require('../../../utils/getAssetIcon.js');
var translations = require('../../../utils/translations.js');
var constants = require('../constants.js');
var useFolderNavigation = require('../hooks/useFolderNavigation.js');

const StyledTable = styledComponents.styled(designSystem.RawTable)`
  width: 100%;
  border-collapse: separate;
  border-spacing: 0;
  border: 1px solid ${({ theme })=>theme.colors.neutral150};
  border-radius: 4px;
  overflow: hidden;
`;
const StyledThead = styledComponents.styled(designSystem.RawThead)`
  background: ${({ theme })=>theme.colors.neutral100};

  tr {
    border-bottom: 1px solid ${({ theme })=>theme.colors.neutral150};
  }
`;
const StyledTh = styledComponents.styled(designSystem.RawTh)`
  height: 40px;
  padding: 0 ${({ theme })=>theme.spaces[4]};
  text-align: left;
`;
const StyledTd = styledComponents.styled(designSystem.RawTd)`
  padding: 0 ${({ theme })=>theme.spaces[4]};
  border-bottom: 1px solid ${({ theme })=>theme.colors.neutral150};
`;
const StyledTr = styledComponents.styled(designSystem.RawTr)`
  height: 48px;
  background: ${({ theme })=>theme.colors.neutral0};
  cursor: pointer;

  &:hover {
    background: ${({ theme })=>theme.colors.primary100};
  }

  &:focus-visible {
    outline: 2px solid ${({ theme })=>theme.colors.primary600};
    outline-offset: -2px;
  }

  &:last-child {
    ${StyledTd} {
      border-bottom: 0;
    }
  }
`;
const StyledBodyTd = styledComponents.styled(designSystem.RawTd)`
  padding: ${({ theme })=>theme.spaces[4]};
  border-bottom: 1px solid ${({ theme })=>theme.colors.neutral150};
`;
const AssetPreviewCell = ({ asset })=>{
    const { ext, mime } = asset;
    const DocIcon = getAssetIcon.getAssetIcon(mime, ext);
    return /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
        justifyContent: "center",
        alignItems: "center",
        borderRadius: "4px",
        color: "neutral500",
        width: "3.2rem",
        height: "3.2rem",
        shrink: 0,
        children: /*#__PURE__*/ jsxRuntime.jsx(DocIcon, {
            width: 20,
            height: 20
        })
    });
};
const AssetRow = ({ asset, onAssetItemClick })=>{
    const isMobile = strapiAdmin.useIsMobile();
    const { formatDate, formatMessage } = reactIntl.useIntl();
    const handleKeyDown = (e)=>{
        if (e.key === 'Enter' || e.key === ' ') {
            e.preventDefault();
            onAssetItemClick(asset.id);
        }
    };
    return /*#__PURE__*/ jsxRuntime.jsxs(StyledTr, {
        tabIndex: 0,
        role: "row",
        onClick: ()=>onAssetItemClick(asset.id),
        onKeyDown: handleKeyDown,
        children: [
            /*#__PURE__*/ jsxRuntime.jsx(StyledTd, {
                children: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                    gap: 3,
                    alignItems: "center",
                    children: [
                        /*#__PURE__*/ jsxRuntime.jsx(AssetPreviewCell, {
                            asset: asset
                        }),
                        /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                            direction: "column",
                            alignItems: "flex-start",
                            children: [
                                /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                                    textColor: "neutral800",
                                    fontWeight: "semiBold",
                                    ellipsis: true,
                                    children: asset.name
                                }),
                                isMobile && /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                                    textColor: "neutral600",
                                    variant: "pi",
                                    children: asset.size ? files.formatBytes(asset.size, 1) : '-'
                                })
                            ]
                        })
                    ]
                })
            }),
            !isMobile && /*#__PURE__*/ jsxRuntime.jsxs(jsxRuntime.Fragment, {
                children: [
                    /*#__PURE__*/ jsxRuntime.jsx(StyledTd, {
                        children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                            textColor: "neutral600",
                            children: asset.createdAt ? formatDate(new Date(asset.createdAt), {
                                dateStyle: 'long'
                            }) : '-'
                        })
                    }),
                    /*#__PURE__*/ jsxRuntime.jsx(StyledTd, {
                        children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                            textColor: "neutral600",
                            children: asset.updatedAt ? formatDate(new Date(asset.updatedAt), {
                                dateStyle: 'long'
                            }) : '-'
                        })
                    }),
                    /*#__PURE__*/ jsxRuntime.jsx(StyledTd, {
                        children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                            textColor: "neutral600",
                            children: asset.size ? files.formatBytes(asset.size, 1) : '-'
                        })
                    })
                ]
            }),
            /*#__PURE__*/ jsxRuntime.jsx(StyledTd, {
                children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
                    justifyContent: "flex-end",
                    children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.IconButton, {
                        label: formatMessage({
                            id: translations.getTranslationKey('control-card.more-actions'),
                            defaultMessage: 'More actions'
                        }),
                        variant: "ghost",
                        children: /*#__PURE__*/ jsxRuntime.jsx(icons.More, {})
                    })
                })
            })
        ]
    });
};
const FolderTr = styledComponents.styled(StyledTr)`
  cursor: pointer;

  &:hover {
    background: ${({ theme })=>theme.colors.primary100};
  }
`;
const FolderRow = ({ folder })=>{
    const isMobile = strapiAdmin.useIsMobile();
    const { formatDate, formatMessage } = reactIntl.useIntl();
    const { navigateToFolder } = useFolderNavigation.useFolderNavigation();
    const handleKeyDown = (e)=>{
        if (e.key === 'Enter' || e.key === ' ') {
            e.preventDefault();
            navigateToFolder(folder);
        }
    };
    return /*#__PURE__*/ jsxRuntime.jsxs(FolderTr, {
        tabIndex: 0,
        role: "row",
        onClick: ()=>navigateToFolder(folder),
        onKeyDown: handleKeyDown,
        children: [
            /*#__PURE__*/ jsxRuntime.jsx(StyledTd, {
                children: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                    gap: 3,
                    alignItems: "center",
                    children: [
                        /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
                            justifyContent: "center",
                            alignItems: "center",
                            borderRadius: "4px",
                            color: "neutral600",
                            width: "3.2rem",
                            height: "3.2rem",
                            shrink: 0,
                            children: /*#__PURE__*/ jsxRuntime.jsx(icons.Folder, {
                                width: 20,
                                height: 20
                            })
                        }),
                        /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                            textColor: "neutral800",
                            fontWeight: "semiBold",
                            ellipsis: true,
                            children: folder.name
                        })
                    ]
                })
            }),
            !isMobile && /*#__PURE__*/ jsxRuntime.jsxs(jsxRuntime.Fragment, {
                children: [
                    /*#__PURE__*/ jsxRuntime.jsx(StyledTd, {
                        children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                            textColor: "neutral600",
                            children: folder.createdAt ? formatDate(new Date(folder.createdAt), {
                                dateStyle: 'long'
                            }) : '-'
                        })
                    }),
                    /*#__PURE__*/ jsxRuntime.jsx(StyledTd, {
                        children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                            textColor: "neutral600",
                            children: folder.updatedAt ? formatDate(new Date(folder.updatedAt), {
                                dateStyle: 'long'
                            }) : '-'
                        })
                    }),
                    /*#__PURE__*/ jsxRuntime.jsx(StyledTd, {
                        children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                            textColor: "neutral600",
                            children: "-"
                        })
                    })
                ]
            }),
            /*#__PURE__*/ jsxRuntime.jsx(StyledTd, {
                children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
                    justifyContent: "flex-end",
                    children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.IconButton, {
                        label: formatMessage({
                            id: translations.getTranslationKey('control-card.more-actions'),
                            defaultMessage: 'More actions'
                        }),
                        variant: "ghost",
                        onClick: (e)=>e.stopPropagation(),
                        children: /*#__PURE__*/ jsxRuntime.jsx(icons.More, {})
                    })
                })
            })
        ]
    });
};
const AssetsTable = ({ assets, folders = [], onAssetItemClick })=>{
    const isMobile = strapiAdmin.useIsMobile();
    const { formatMessage } = reactIntl.useIntl();
    const visibleHeaders = isMobile ? constants.TABLE_HEADERS.filter((h)=>h.name === 'name' || h.name === 'actions') : constants.TABLE_HEADERS;
    const totalRows = folders.length + assets.length;
    return /*#__PURE__*/ jsxRuntime.jsxs(StyledTable, {
        colCount: visibleHeaders.length,
        rowCount: totalRows + 1,
        children: [
            /*#__PURE__*/ jsxRuntime.jsx(StyledThead, {
                children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.RawTr, {
                    children: visibleHeaders.map((header)=>{
                        const tableHeaderLabel = formatMessage(header.label);
                        const isVisuallyHidden = 'isVisuallyHidden' in header && header.isVisuallyHidden;
                        if (isVisuallyHidden) {
                            return /*#__PURE__*/ jsxRuntime.jsx(StyledTh, {
                                children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.VisuallyHidden, {
                                    children: formatMessage({
                                        id: translations.getTranslationKey('table.header.actions'),
                                        defaultMessage: 'actions'
                                    })
                                })
                            }, header.name);
                        }
                        return /*#__PURE__*/ jsxRuntime.jsx(StyledTh, {
                            children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                                textColor: "neutral600",
                                variant: "sigma",
                                children: tableHeaderLabel
                            })
                        }, header.name);
                    })
                })
            }),
            /*#__PURE__*/ jsxRuntime.jsx(designSystem.RawTbody, {
                children: totalRows === 0 ? /*#__PURE__*/ jsxRuntime.jsx(designSystem.RawTr, {
                    children: /*#__PURE__*/ jsxRuntime.jsx(StyledBodyTd, {
                        colSpan: visibleHeaders.length,
                        children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                            textColor: "neutral600",
                            children: formatMessage({
                                id: 'app.components.EmptyStateLayout.content-document',
                                defaultMessage: 'No content found'
                            })
                        })
                    })
                }) : /*#__PURE__*/ jsxRuntime.jsxs(jsxRuntime.Fragment, {
                    children: [
                        folders.map((folder)=>/*#__PURE__*/ jsxRuntime.jsx(FolderRow, {
                                folder: folder
                            }, `folder-${folder.id}`)),
                        assets.map((asset)=>/*#__PURE__*/ jsxRuntime.jsx(AssetRow, {
                                asset: asset,
                                onAssetItemClick: onAssetItemClick
                            }, asset.id))
                    ]
                })
            })
        ]
    });
};

exports.AssetsTable = AssetsTable;
