// @strapi/upload@5.42.1/dist/admin/future/pages/Assets/components/AssetsTable.mjs
import { jsxs, jsx, Fragment } from 'react/jsx-runtime';
import { useIsMobile } from '@strapi/admin/strapi-admin';
import { RawTr, VisuallyHidden, Typography, RawTbody, RawTable, RawThead, RawTh, RawTd, Flex, IconButton } from '@strapi/design-system';
import { Folder, More } from '@strapi/icons';
import { useIntl } from 'react-intl';
import { styled } from 'styled-components';
import { formatBytes } from '../../../utils/files.mjs';
import { getAssetIcon } from '../../../utils/getAssetIcon.mjs';
import { getTranslationKey } from '../../../utils/translations.mjs';
import { TABLE_HEADERS } from '../constants.mjs';
import { useFolderNavigation } from '../hooks/useFolderNavigation.mjs';

const StyledTable = styled(RawTable)`
  width: 100%;
  border-collapse: separate;
  border-spacing: 0;
  border: 1px solid ${({ theme })=>theme.colors.neutral150};
  border-radius: 4px;
  overflow: hidden;
`;
const StyledThead = styled(RawThead)`
  background: ${({ theme })=>theme.colors.neutral100};

  tr {
    border-bottom: 1px solid ${({ theme })=>theme.colors.neutral150};
  }
`;
const StyledTh = styled(RawTh)`
  height: 40px;
  padding: 0 ${({ theme })=>theme.spaces[4]};
  text-align: left;
`;
const StyledTd = styled(RawTd)`
  padding: 0 ${({ theme })=>theme.spaces[4]};
  border-bottom: 1px solid ${({ theme })=>theme.colors.neutral150};
`;
const StyledTr = styled(RawTr)`
  height: 48px;
  background: ${({ theme })=>theme.colors.neutral0};
  cursor: pointer;

  &:hover {
    background: ${({ theme })=>theme.colors.primary100};
  }

  &:focus-visible {
    outline: 2px solid ${({ theme })=>theme.colors.primary600};
    outline-offset: -2px;
  }

  &:last-child {
    ${StyledTd} {
      border-bottom: 0;
    }
  }
`;
const StyledBodyTd = styled(RawTd)`
  padding: ${({ theme })=>theme.spaces[4]};
  border-bottom: 1px solid ${({ theme })=>theme.colors.neutral150};
`;
const AssetPreviewCell = ({ asset })=>{
    const { ext, mime } = asset;
    const DocIcon = getAssetIcon(mime, ext);
    return /*#__PURE__*/ jsx(Flex, {
        justifyContent: "center",
        alignItems: "center",
        borderRadius: "4px",
        color: "neutral500",
        width: "3.2rem",
        height: "3.2rem",
        shrink: 0,
        children: /*#__PURE__*/ jsx(DocIcon, {
            width: 20,
            height: 20
        })
    });
};
const AssetRow = ({ asset, onAssetItemClick })=>{
    const isMobile = useIsMobile();
    const { formatDate, formatMessage } = useIntl();
    const handleKeyDown = (e)=>{
        if (e.key === 'Enter' || e.key === ' ') {
            e.preventDefault();
            onAssetItemClick(asset.id);
        }
    };
    return /*#__PURE__*/ jsxs(StyledTr, {
        tabIndex: 0,
        role: "row",
        onClick: ()=>onAssetItemClick(asset.id),
        onKeyDown: handleKeyDown,
        children: [
            /*#__PURE__*/ jsx(StyledTd, {
                children: /*#__PURE__*/ jsxs(Flex, {
                    gap: 3,
                    alignItems: "center",
                    children: [
                        /*#__PURE__*/ jsx(AssetPreviewCell, {
                            asset: asset
                        }),
                        /*#__PURE__*/ jsxs(Flex, {
                            direction: "column",
                            alignItems: "flex-start",
                            children: [
                                /*#__PURE__*/ jsx(Typography, {
                                    textColor: "neutral800",
                                    fontWeight: "semiBold",
                                    ellipsis: true,
                                    children: asset.name
                                }),
                                isMobile && /*#__PURE__*/ jsx(Typography, {
                                    textColor: "neutral600",
                                    variant: "pi",
                                    children: asset.size ? formatBytes(asset.size, 1) : '-'
                                })
                            ]
                        })
                    ]
                })
            }),
            !isMobile && /*#__PURE__*/ jsxs(Fragment, {
                children: [
                    /*#__PURE__*/ jsx(StyledTd, {
                        children: /*#__PURE__*/ jsx(Typography, {
                            textColor: "neutral600",
                            children: asset.createdAt ? formatDate(new Date(asset.createdAt), {
                                dateStyle: 'long'
                            }) : '-'
                        })
                    }),
                    /*#__PURE__*/ jsx(StyledTd, {
                        children: /*#__PURE__*/ jsx(Typography, {
                            textColor: "neutral600",
                            children: asset.updatedAt ? formatDate(new Date(asset.updatedAt), {
                                dateStyle: 'long'
                            }) : '-'
                        })
                    }),
                    /*#__PURE__*/ jsx(StyledTd, {
                        children: /*#__PURE__*/ jsx(Typography, {
                            textColor: "neutral600",
                            children: asset.size ? formatBytes(asset.size, 1) : '-'
                        })
                    })
                ]
            }),
            /*#__PURE__*/ jsx(StyledTd, {
                children: /*#__PURE__*/ jsx(Flex, {
                    justifyContent: "flex-end",
                    children: /*#__PURE__*/ jsx(IconButton, {
                        label: formatMessage({
                            id: getTranslationKey('control-card.more-actions'),
                            defaultMessage: 'More actions'
                        }),
                        variant: "ghost",
                        children: /*#__PURE__*/ jsx(More, {})
                    })
                })
            })
        ]
    });
};
const FolderTr = styled(StyledTr)`
  cursor: pointer;

  &:hover {
    background: ${({ theme })=>theme.colors.primary100};
  }
`;
const FolderRow = ({ folder })=>{
    const isMobile = useIsMobile();
    const { formatDate, formatMessage } = useIntl();
    const { navigateToFolder } = useFolderNavigation();
    const handleKeyDown = (e)=>{
        if (e.key === 'Enter' || e.key === ' ') {
            e.preventDefault();
            navigateToFolder(folder);
        }
    };
    return /*#__PURE__*/ jsxs(FolderTr, {
        tabIndex: 0,
        role: "row",
        onClick: ()=>navigateToFolder(folder),
        onKeyDown: handleKeyDown,
        children: [
            /*#__PURE__*/ jsx(StyledTd, {
                children: /*#__PURE__*/ jsxs(Flex, {
                    gap: 3,
                    alignItems: "center",
                    children: [
                        /*#__PURE__*/ jsx(Flex, {
                            justifyContent: "center",
                            alignItems: "center",
                            borderRadius: "4px",
                            color: "neutral600",
                            width: "3.2rem",
                            height: "3.2rem",
                            shrink: 0,
                            children: /*#__PURE__*/ jsx(Folder, {
                                width: 20,
                                height: 20
                            })
                        }),
                        /*#__PURE__*/ jsx(Typography, {
                            textColor: "neutral800",
                            fontWeight: "semiBold",
                            ellipsis: true,
                            children: folder.name
                        })
                    ]
                })
            }),
            !isMobile && /*#__PURE__*/ jsxs(Fragment, {
                children: [
                    /*#__PURE__*/ jsx(StyledTd, {
                        children: /*#__PURE__*/ jsx(Typography, {
                            textColor: "neutral600",
                            children: folder.createdAt ? formatDate(new Date(folder.createdAt), {
                                dateStyle: 'long'
                            }) : '-'
                        })
                    }),
                    /*#__PURE__*/ jsx(StyledTd, {
                        children: /*#__PURE__*/ jsx(Typography, {
                            textColor: "neutral600",
                            children: folder.updatedAt ? formatDate(new Date(folder.updatedAt), {
                                dateStyle: 'long'
                            }) : '-'
                        })
                    }),
                    /*#__PURE__*/ jsx(StyledTd, {
                        children: /*#__PURE__*/ jsx(Typography, {
                            textColor: "neutral600",
                            children: "-"
                        })
                    })
                ]
            }),
            /*#__PURE__*/ jsx(StyledTd, {
                children: /*#__PURE__*/ jsx(Flex, {
                    justifyContent: "flex-end",
                    children: /*#__PURE__*/ jsx(IconButton, {
                        label: formatMessage({
                            id: getTranslationKey('control-card.more-actions'),
                            defaultMessage: 'More actions'
                        }),
                        variant: "ghost",
                        onClick: (e)=>e.stopPropagation(),
                        children: /*#__PURE__*/ jsx(More, {})
                    })
                })
            })
        ]
    });
};
const AssetsTable = ({ assets, folders = [], onAssetItemClick })=>{
    const isMobile = useIsMobile();
    const { formatMessage } = useIntl();
    const visibleHeaders = isMobile ? TABLE_HEADERS.filter((h)=>h.name === 'name' || h.name === 'actions') : TABLE_HEADERS;
    const totalRows = folders.length + assets.length;
    return /*#__PURE__*/ jsxs(StyledTable, {
        colCount: visibleHeaders.length,
        rowCount: totalRows + 1,
        children: [
            /*#__PURE__*/ jsx(StyledThead, {
                children: /*#__PURE__*/ jsx(RawTr, {
                    children: visibleHeaders.map((header)=>{
                        const tableHeaderLabel = formatMessage(header.label);
                        const isVisuallyHidden = 'isVisuallyHidden' in header && header.isVisuallyHidden;
                        if (isVisuallyHidden) {
                            return /*#__PURE__*/ jsx(StyledTh, {
                                children: /*#__PURE__*/ jsx(VisuallyHidden, {
                                    children: formatMessage({
                                        id: getTranslationKey('table.header.actions'),
                                        defaultMessage: 'actions'
                                    })
                                })
                            }, header.name);
                        }
                        return /*#__PURE__*/ jsx(StyledTh, {
                            children: /*#__PURE__*/ jsx(Typography, {
                                textColor: "neutral600",
                                variant: "sigma",
                                children: tableHeaderLabel
                            })
                        }, header.name);
                    })
                })
            }),
            /*#__PURE__*/ jsx(RawTbody, {
                children: totalRows === 0 ? /*#__PURE__*/ jsx(RawTr, {
                    children: /*#__PURE__*/ jsx(StyledBodyTd, {
                        colSpan: visibleHeaders.length,
                        children: /*#__PURE__*/ jsx(Typography, {
                            textColor: "neutral600",
                            children: formatMessage({
                                id: 'app.components.EmptyStateLayout.content-document',
                                defaultMessage: 'No content found'
                            })
                        })
                    })
                }) : /*#__PURE__*/ jsxs(Fragment, {
                    children: [
                        folders.map((folder)=>/*#__PURE__*/ jsx(FolderRow, {
                                folder: folder
                            }, `folder-${folder.id}`)),
                        assets.map((asset)=>/*#__PURE__*/ jsx(AssetRow, {
                                asset: asset,
                                onAssetItemClick: onAssetItemClick
                            }, asset.id))
                    ]
                })
            })
        ]
    });
};

export { AssetsTable };
