// @strapi/upload@5.42.1/dist/admin/future/pages/Assets/components/CreateFolderDialog.mjs
import { jsx, jsxs } from 'react/jsx-runtime';
import { useState, useEffect } from 'react';
import { useNotification } from '@strapi/admin/strapi-admin';
import { Modal, Field, TextInput, Flex, Button } from '@strapi/design-system';
import { useIntl } from 'react-intl';
import { styled } from 'styled-components';
import { useCreateFolderMutation } from '../../../services/folders.mjs';
import { getTranslationKey } from '../../../utils/translations.mjs';

const StyledModalContent = styled(Modal.Content)`
  max-width: 51.6rem;
`;
const CreateFolderDialog = ({ open, folderName, parentFolderId, onClose })=>{
    const { formatMessage } = useIntl();
    const { toggleNotification } = useNotification();
    const [name, setName] = useState('');
    const [fieldError, setFieldError] = useState();
    const [createFolder, { isLoading }] = useCreateFolderMutation();
    useEffect(()=>{
        if (open) {
            setName('');
            setFieldError(undefined);
        }
    }, [
        open
    ]);
    const handleSubmit = async (e)=>{
        e.preventDefault();
        const trimmedName = name.trim();
        if (!trimmedName) {
            setFieldError(formatMessage({
                id: getTranslationKey('folder.create.form.error.name-required'),
                defaultMessage: 'Name is required'
            }));
            return;
        }
        try {
            await createFolder({
                name: trimmedName,
                parent: parentFolderId
            }).unwrap();
            toggleNotification({
                type: 'success',
                message: formatMessage({
                    id: getTranslationKey('folder.create.success'),
                    defaultMessage: 'Folder has been created'
                })
            });
            onClose();
        } catch (err) {
            const apiError = err;
            if (apiError?.message) {
                setFieldError(apiError.message);
            } else {
                toggleNotification({
                    type: 'danger',
                    message: formatMessage({
                        id: getTranslationKey('folder.create.form.error.unknown'),
                        defaultMessage: 'An error occurred while creating the folder'
                    })
                });
            }
        }
    };
    return /*#__PURE__*/ jsx(Modal.Root, {
        open: open,
        onOpenChange: onClose,
        children: /*#__PURE__*/ jsxs(StyledModalContent, {
            children: [
                /*#__PURE__*/ jsx(Modal.Header, {
                    children: /*#__PURE__*/ jsx(Modal.Title, {
                        children: formatMessage({
                            id: getTranslationKey('folder.create.title-in'),
                            defaultMessage: 'New folder in {folderName}'
                        }, {
                            folderName
                        })
                    })
                }),
                /*#__PURE__*/ jsxs("form", {
                    onSubmit: handleSubmit,
                    children: [
                        /*#__PURE__*/ jsx(Modal.Body, {
                            children: /*#__PURE__*/ jsxs(Field.Root, {
                                error: fieldError,
                                name: "name",
                                required: true,
                                children: [
                                    /*#__PURE__*/ jsx(Field.Label, {
                                        children: formatMessage({
                                            id: getTranslationKey('folder.form.name.label'),
                                            defaultMessage: 'Folder name'
                                        })
                                    }),
                                    /*#__PURE__*/ jsx(TextInput, {
                                        value: name,
                                        onChange: (e)=>{
                                            setName(e.target.value);
                                            setFieldError(undefined);
                                        },
                                        autoFocus: true
                                    }),
                                    /*#__PURE__*/ jsx(Field.Error, {})
                                ]
                            })
                        }),
                        /*#__PURE__*/ jsx(Modal.Footer, {
                            children: /*#__PURE__*/ jsxs(Flex, {
                                gap: 2,
                                justifyContent: "space-between",
                                width: "100%",
                                children: [
                                    /*#__PURE__*/ jsx(Button, {
                                        variant: "tertiary",
                                        onClick: onClose,
                                        type: "button",
                                        children: formatMessage({
                                            id: 'app.components.Button.cancel',
                                            defaultMessage: 'Cancel'
                                        })
                                    }),
                                    /*#__PURE__*/ jsx(Button, {
                                        type: "submit",
                                        loading: isLoading,
                                        children: formatMessage({
                                            id: getTranslationKey('folder.create.submit'),
                                            defaultMessage: 'Create folder'
                                        })
                                    })
                                ]
                            })
                        })
                    ]
                })
            ]
        })
    });
};

export { CreateFolderDialog };
