// @strapi/upload@5.42.1/dist/admin/future/pages/Assets/components/ImportFromUrlDialog.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var designSystem = require('@strapi/design-system');
var reactIntl = require('react-intl');
var files = require('../../../utils/files.js');
var translations = require('../../../utils/translations.js');

function _interopNamespaceDefault(e) {
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespaceDefault(React);

const ImportFromUrlDialog = ({ open, onClose, onUpload })=>{
    const { formatMessage } = reactIntl.useIntl();
    const [urls, setUrls] = React__namespace.useState('');
    const [error, setError] = React__namespace.useState(null);
    const handleClose = ()=>{
        setUrls('');
        setError(null);
        onClose();
    };
    const handleSubmit = async (e)=>{
        e.preventDefault();
        // Validate URLs
        const { urls: validUrls, error: validationError } = files.validateUrls(urls);
        if (validationError) {
            setError(validationError);
            return;
        }
        setError(null);
        // Close dialog and let the mutation handle fetching and uploading
        handleClose();
        // Pass URLs to mutation which handles fetching and upload progress
        await onUpload(validUrls);
    };
    return /*#__PURE__*/ jsxRuntime.jsx(designSystem.Modal.Root, {
        open: open,
        onOpenChange: (isOpen)=>!isOpen && handleClose(),
        children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Modal.Content, {
            children: /*#__PURE__*/ jsxRuntime.jsxs("form", {
                onSubmit: handleSubmit,
                children: [
                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Modal.Header, {
                        children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Modal.Title, {
                            children: formatMessage({
                                id: translations.getTranslationKey('modal.url.title'),
                                defaultMessage: 'Import from URL'
                            })
                        })
                    }),
                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Modal.Body, {
                        children: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Field.Root, {
                            error: error || undefined,
                            hint: formatMessage({
                                id: translations.getTranslationKey('input.url.description'),
                                defaultMessage: 'Separate your URL links by a carriage return.'
                            }),
                            children: [
                                /*#__PURE__*/ jsxRuntime.jsx(designSystem.Field.Label, {
                                    children: formatMessage({
                                        id: translations.getTranslationKey('input.url.label'),
                                        defaultMessage: 'URL(s)'
                                    })
                                }),
                                /*#__PURE__*/ jsxRuntime.jsx(designSystem.Textarea, {
                                    name: "urls",
                                    minHeight: "unset",
                                    rows: Math.min(urls.split('\n').length, 7),
                                    maxHeight: "10.5rem",
                                    placeholder: formatMessage({
                                        id: translations.getTranslationKey('input.url.placeholder'),
                                        defaultMessage: 'Empty'
                                    }),
                                    value: urls,
                                    onChange: (e)=>{
                                        setUrls(e.target.value);
                                        setError(null);
                                    }
                                }),
                                /*#__PURE__*/ jsxRuntime.jsx(designSystem.Field.Hint, {}),
                                /*#__PURE__*/ jsxRuntime.jsx(designSystem.Field.Error, {})
                            ]
                        })
                    }),
                    /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Modal.Footer, {
                        children: [
                            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Button, {
                                variant: "tertiary",
                                onClick: handleClose,
                                children: formatMessage({
                                    id: 'app.components.Button.cancel',
                                    defaultMessage: 'Cancel'
                                })
                            }),
                            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Button, {
                                type: "submit",
                                children: formatMessage({
                                    id: translations.getTranslationKey('modal.url.upload'),
                                    defaultMessage: 'Upload'
                                })
                            })
                        ]
                    })
                ]
            })
        })
    });
};

exports.ImportFromUrlDialog = ImportFromUrlDialog;
