// @strapi/admin@5.42.1/dist/admin/admin/src/components/SearchInput.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var designSystem = require('@strapi/design-system');
var icons = require('@strapi/icons');
var reactIntl = require('react-intl');
var Tracking = require('../features/Tracking.js');
var useMediaQuery = require('../hooks/useMediaQuery.js');
var useQueryParams = require('../hooks/useQueryParams.js');

function _interopNamespaceDefault(e) {
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespaceDefault(React);

const SearchInput = ({ disabled, label, placeholder, trackedEvent, trackedEventDetails })=>{
    const inputRef = React__namespace.useRef(null);
    const iconButtonRef = React__namespace.useRef(null);
    const [{ query }, setQuery] = useQueryParams.useQueryParams();
    const [value, setValue] = React__namespace.useState(query?._q || '');
    const [isOpen, setIsOpen] = React__namespace.useState(!!value);
    const { formatMessage } = reactIntl.useIntl();
    const { trackUsage } = Tracking.useTracking();
    const isMobile = useMediaQuery.useIsMobile();
    const handleToggle = ()=>setIsOpen((prev)=>!prev);
    React__namespace.useLayoutEffect(()=>{
        if (isOpen && inputRef.current) {
            inputRef.current.focus();
        }
    }, [
        isOpen
    ]);
    const handleClear = ()=>{
        setValue('');
        setQuery({
            _q: ''
        }, 'remove');
    };
    const handleSubmit = (e)=>{
        e.preventDefault();
        // Ensure value is a string
        if (value) {
            if (trackedEvent) {
                trackUsage(trackedEvent, trackedEventDetails);
            }
            setQuery({
                _q: encodeURIComponent(value),
                page: 1
            });
        } else {
            handleToggle();
            setQuery({
                _q: ''
            }, 'remove');
        }
    };
    const renderSearchForm = (opts)=>/*#__PURE__*/ jsxRuntime.jsx(designSystem.SearchForm, {
            onSubmit: handleSubmit,
            children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Searchbar, {
                ref: inputRef,
                name: "search",
                onChange: (e)=>setValue(e.target.value),
                value: value,
                clearLabel: formatMessage({
                    id: 'clearLabel',
                    defaultMessage: 'Clear'
                }),
                onClear: handleClear,
                placeholder: isMobile ? undefined : placeholder,
                disabled: disabled,
                onBlur: opts?.enableBlurClose ? (e)=>{
                    if (!e.currentTarget.contains(e.relatedTarget) && e.currentTarget.value === '') {
                        setIsOpen(false);
                    }
                } : undefined,
                children: label
            })
        });
    if (isMobile) {
        // Add wrapper so that the Searchbar takes up the rest of the available space.
        return /*#__PURE__*/ jsxRuntime.jsx(designSystem.Box, {
            width: "100%",
            children: renderSearchForm()
        });
    }
    if (isOpen) {
        return renderSearchForm({
            enableBlurClose: true
        });
    }
    return /*#__PURE__*/ jsxRuntime.jsx(designSystem.IconButton, {
        ref: iconButtonRef,
        disabled: disabled,
        label: formatMessage({
            id: 'global.search',
            defaultMessage: 'Search'
        }),
        onClick: handleToggle,
        children: /*#__PURE__*/ jsxRuntime.jsx(icons.Search, {})
    });
};

exports.SearchInput = SearchInput;
