// @strapi/upload@5.42.1/dist/admin/future/pages/Assets/hooks/useFolderInfo.js
'use strict';

var reactIntl = require('react-intl');
var assets = require('../../../services/assets.js');
var folders = require('../../../services/folders.js');
var translations = require('../../../utils/translations.js');

const useFolderInfo = (currentFolderId)=>{
    const { formatMessage } = reactIntl.useIntl();
    const { data: currentFolder, isLoading } = folders.useGetFolderQuery({
        id: currentFolderId
    }, {
        skip: currentFolderId === null
    });
    const { data: rootAssetsData, isLoading: isRootAssetsLoading } = assets.useGetAssetsQuery({
        folder: null,
        pageSize: 1
    }, {
        skip: currentFolderId !== null
    });
    const homeLabel = formatMessage({
        id: translations.getTranslationKey('plugin.home'),
        defaultMessage: 'Home'
    });
    if (currentFolderId === null) {
        if (isRootAssetsLoading) {
            return {
                title: homeLabel,
                itemCount: 0
            };
        }
        return {
            title: homeLabel,
            itemCount: rootAssetsData?.pagination?.total ?? 0
        };
    }
    if (isLoading) {
        return {
            title: homeLabel,
            itemCount: 0
        };
    }
    return {
        title: currentFolder?.name ?? homeLabel,
        itemCount: currentFolder?.files?.count ?? 0
    };
};

exports.useFolderInfo = useFolderInfo;
