// @strapi/content-releases/dist/server/utils/index.mjs
import { contentTypes } from '@strapi/utils';

const getService = (name, { strapi: strapi1 })=>{
    return strapi1.plugin('content-releases').service(name);
};
const getDraftEntryValidStatus = async ({ contentType, documentId, locale }, { strapi: strapi1 })=>{
    const populateBuilderService = strapi1.plugin('content-manager').service('populate-builder');
    // @ts-expect-error - populateBuilderService should be a function but is returning service
    const populate = await populateBuilderService(contentType).populateDeep(Infinity).build();
    const entry = await getEntry({
        contentType,
        documentId,
        locale,
        populate
    }, {
        strapi: strapi1
    });
    return isEntryValid(contentType, entry, {
        strapi: strapi1
    });
};
const isEntryValid = async (contentTypeUid, entry, { strapi: strapi1 })=>{
    try {
        // @TODO: When documents service has validateEntityCreation method, use it instead
        await strapi1.entityValidator.validateEntityCreation(strapi1.getModel(contentTypeUid), entry, undefined, // @ts-expect-error - FIXME: entity here is unnecessary
        entry);
        const workflowsService = strapi1.plugin('review-workflows').service('workflows');
        // Workflows service may not be available depending on the license
        const workflow = await workflowsService?.getAssignedWorkflow(contentTypeUid, {
            populate: 'stageRequiredToPublish'
        });
        if (workflow?.stageRequiredToPublish) {
            return entry.strapi_stage.id === workflow.stageRequiredToPublish.id;
        }
        return true;
    } catch  {
        return false;
    }
};
const getEntry = async ({ contentType, documentId, locale, populate, status = 'draft' }, { strapi: strapi1 })=>{
    if (documentId) {
        // Try to get an existing draft or published document
        const entry = await strapi1.documents(contentType).findOne({
            documentId,
            locale,
            populate,
            status
        });
        // The document isn't published yet, but the action is to publish it, fetch the draft
        if (status === 'published' && !entry) {
            return strapi1.documents(contentType).findOne({
                documentId,
                locale,
                populate,
                status: 'draft'
            });
        }
        return entry;
    }
    return strapi1.documents(contentType).findFirst({
        locale,
        populate,
        status
    });
};
const getEntryStatus = async (contentType, entry)=>{
    if (entry.publishedAt) {
        return 'published';
    }
    const publishedEntry = await strapi.documents(contentType).findOne({
        documentId: entry.documentId,
        locale: entry.locale,
        status: 'published',
        fields: [
            'updatedAt'
        ]
    });
    if (!publishedEntry) {
        return 'draft';
    }
    const entryUpdatedAt = new Date(entry.updatedAt).getTime();
    const publishedEntryUpdatedAt = new Date(publishedEntry.updatedAt).getTime();
    if (entryUpdatedAt > publishedEntryUpdatedAt) {
        return 'modified';
    }
    return 'published';
};
/**
 * Recursively collects content type UIDs that a model (content type or component) has relations to.
 * Go through component and dynamic zone attributes to find nested relations.
 */ const collectRelationTargets = (modelUid, strapi1, visited = new Set())=>{
    const targets = new Set();
    if (visited.has(modelUid)) {
        return targets;
    }
    visited.add(modelUid);
    const model = strapi1.getModel(modelUid);
    if (!model?.attributes) {
        return targets;
    }
    for (const attribute of Object.values(model.attributes)){
        if (attribute?.type === 'relation' && attribute.target) {
            targets.add(attribute.target);
        }
        if (attribute?.type === 'component' && attribute.component) {
            for (const t of collectRelationTargets(attribute.component, strapi1, visited)){
                targets.add(t);
            }
        }
        if (attribute?.type === 'dynamiczone' && attribute.components) {
            for (const compUid of attribute.components){
                for (const t of collectRelationTargets(compUid, strapi1, visited)){
                    targets.add(t);
                }
            }
        }
    }
    return targets;
};
/**
 * Returns content type UIDs sorted by relation dependency order for publishing.
 * When content type A has a relation to content type B (both with draft & publish),
 * B will appear before A in the result. This ensures that when publishing a release,
 * related entities are published first, so that relation IDs can be correctly
 * resolved (published target must exist when publishing source).
 *
 * Relations in components (nested or not) and dynamic zones are also considered.
 *
 * @param contentTypeUids - Content type UIDs that will be published in the release
 * @param strapi - Strapi instance
 * @returns Content type UIDs in publish order (dependencies first)
 */ const getPublishOrderForContentTypes = (contentTypeUids, { strapi: strapi1 })=>{
    const uidSet = new Set(contentTypeUids);
    // Build dependency graph: source depends on target (source must be published after target)
    const dependencies = new Map();
    for (const uid of contentTypeUids){
        const model = strapi1.getModel(uid);
        if (model && contentTypes.hasDraftAndPublish(model)) {
            const relationTargets = collectRelationTargets(uid, strapi1);
            for (const targetUid of relationTargets){
                const targetContentTypeUid = targetUid;
                const isTargetInRelease = uidSet.has(targetContentTypeUid) && targetContentTypeUid in strapi1.contentTypes;
                const targetModel = strapi1.getModel(targetContentTypeUid);
                const targetHasDraftAndPublish = targetModel && contentTypes.hasDraftAndPublish(targetModel);
                if (isTargetInRelease && targetHasDraftAndPublish) {
                    let dependencySet = dependencies.get(uid);
                    if (!dependencySet) {
                        dependencySet = new Set();
                        dependencies.set(uid, dependencySet);
                    }
                    dependencySet.add(targetContentTypeUid);
                }
            }
        }
    }
    // Topological sort: dependencies first
    const sorted = [];
    const visited = new Set();
    const visiting = new Set();
    const visit = (uid)=>{
        if (visited.has(uid)) return;
        if (visiting.has(uid)) return; // No cycle in valid schemas
        visiting.add(uid);
        for (const dep of dependencies.get(uid) ?? []){
            visit(dep);
        }
        visiting.delete(uid);
        visited.add(uid);
        sorted.push(uid);
    };
    for (const uid of contentTypeUids){
        visit(uid);
    }
    return sorted;
};

export { getDraftEntryValidStatus, getEntry, getEntryStatus, getPublishOrderForContentTypes, getService, isEntryValid };
