// @strapi/upload@5.42.1/dist/admin/future/pages/Assets/AssetsPage.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var ToggleGroup = require('@radix-ui/react-toggle-group');
var strapiAdmin = require('@strapi/admin/strapi-admin');
var designSystem = require('@strapi/design-system');
var icons = require('@strapi/icons');
var reactIntl = require('react-intl');
var styledComponents = require('styled-components');
var api = require('../../services/api.js');
var folders = require('../../services/folders.js');
var translations = require('../../utils/translations.js');
var AssetDetailsDrawer = require('./components/AssetDetails/AssetDetailsDrawer.js');
var AssetsGrid = require('./components/AssetsGrid.js');
var AssetsTable = require('./components/AssetsTable.js');
var CreateFolderDialog = require('./components/CreateFolderDialog.js');
var UploadDropZone = require('./components/DropZone/UploadDropZone.js');
var UploadDropZoneContext = require('./components/DropZone/UploadDropZoneContext.js');
var ImportFromUrlDialog = require('./components/ImportFromUrlDialog.js');
var constants = require('./constants.js');
var useFolderInfo = require('./hooks/useFolderInfo.js');
var useFolderNavigation = require('./hooks/useFolderNavigation.js');
var useInfiniteAssets = require('./hooks/useInfiniteAssets.js');

function _interopNamespaceDefault(e) {
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var ToggleGroup__namespace = /*#__PURE__*/_interopNamespaceDefault(ToggleGroup);

const INTERSECTION_OPTIONS = {
    threshold: 0.1
};
const AssetsView = ({ view, folderId, onAssetItemClick })=>{
    const { formatMessage } = reactIntl.useIntl();
    const { assets, isLoading: isLoadingAssets, isFetchingMore, hasNextPage, fetchNextPage, error } = useInfiniteAssets.useInfiniteAssets({
        folder: folderId
    });
    const { data: folders$1 = [], isLoading: isLoadingFolders } = folders.useGetFoldersQuery({
        parentId: folderId
    });
    const isGridView = view === constants.viewOptions.GRID;
    const isLoading = isLoadingAssets || isLoadingFolders;
    const loadMoreRef = strapiAdmin.useElementOnScreen(React.useCallback((isVisible)=>{
        if (isVisible && hasNextPage && !isFetchingMore) {
            fetchNextPage();
        }
    }, [
        hasNextPage,
        isFetchingMore,
        fetchNextPage
    ]), INTERSECTION_OPTIONS);
    if (isLoading) {
        return /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
            justifyContent: "center",
            padding: 8,
            children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Loader, {
                children: formatMessage({
                    id: 'app.loading',
                    defaultMessage: 'Loading...'
                })
            })
        });
    }
    if (error) {
        return /*#__PURE__*/ jsxRuntime.jsx(designSystem.Box, {
            padding: 8,
            children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                textColor: "danger600",
                children: formatMessage({
                    id: translations.getTranslationKey('list.assets.error'),
                    defaultMessage: 'An error occurred while fetching assets.'
                })
            })
        });
    }
    if (folders$1.length === 0 && assets.length === 0) {
        return /*#__PURE__*/ jsxRuntime.jsx(designSystem.Box, {
            padding: 8,
            children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                textColor: "neutral600",
                children: formatMessage({
                    id: 'app.components.EmptyStateLayout.content-document',
                    defaultMessage: 'No content found'
                })
            })
        });
    }
    return /*#__PURE__*/ jsxRuntime.jsxs(jsxRuntime.Fragment, {
        children: [
            isGridView ? /*#__PURE__*/ jsxRuntime.jsx(AssetsGrid.AssetsGrid, {
                folders: folders$1,
                assets: assets,
                onAssetItemClick: onAssetItemClick
            }) : /*#__PURE__*/ jsxRuntime.jsx(AssetsTable.AssetsTable, {
                assets: assets,
                folders: folders$1,
                onAssetItemClick: onAssetItemClick
            }),
            /*#__PURE__*/ jsxRuntime.jsx("div", {
                ref: loadMoreRef,
                style: {
                    height: 1
                }
            }),
            isFetchingMore && /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
                justifyContent: "center",
                padding: 4,
                children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Loader, {
                    children: formatMessage({
                        id: translations.getTranslationKey('list.assets.loading-more'),
                        defaultMessage: 'Loading more assets...'
                    })
                })
            })
        ]
    });
};
/* -------------------------------------------------------------------------------------------------
 * AssetsPage
 * -----------------------------------------------------------------------------------------------*/ const StyledToggleGroup = styledComponents.styled(ToggleGroup__namespace.Root)`
  display: flex;
  border: 1px solid ${({ theme })=>theme.colors.neutral200};
  border-radius: ${({ theme })=>theme.borderRadius};
  overflow: hidden;
`;
const StyledToggleItem = styledComponents.styled(ToggleGroup__namespace.Item)`
  display: flex;
  align-items: center;
  gap: ${({ theme })=>theme.spaces[2]};
  padding: ${({ theme })=>`${theme.spaces[2]} ${theme.spaces[4]}`};
  border: none;
  background: ${({ theme })=>theme.colors.neutral0};
  color: ${({ theme })=>theme.colors.neutral800};
  cursor: pointer;
  font-size: ${({ theme })=>theme.fontSizes[1]};
  font-weight: ${({ theme })=>theme.fontWeights.semiBold};

  &:hover {
    background: ${({ theme })=>theme.colors.primary100};
  }

  &[data-state='on'] {
    background: ${({ theme })=>theme.colors.neutral150};
  }

  svg {
    width: 1.6rem;
    height: 1.6rem;
  }
`;
const HeaderWrapper = styledComponents.styled(designSystem.Box)`
  [data-strapi-header] {
    background: ${({ theme })=>theme.colors.neutral0};

    h1 {
      font-size: 1.8rem;
    }
  }
`;
const AssetsPage = ()=>{
    const { formatMessage } = reactIntl.useIntl();
    const { openDetails } = AssetDetailsDrawer.useAssetDetailsParam();
    const { currentFolderId } = useFolderNavigation.useFolderNavigation();
    const { title, itemCount } = useFolderInfo.useFolderInfo(currentFolderId);
    const itemCountLabel = formatMessage({
        id: translations.getTranslationKey('header.content.item-count'),
        defaultMessage: '{count, plural, =1 {# item} other {# items}}'
    }, {
        count: itemCount
    });
    const [isCreateFolderDialogOpen, setIsCreateFolderDialogOpen] = React.useState(false);
    // View state
    const [view, setView] = strapiAdmin.usePersistentState(constants.localStorageKeys.view, constants.viewOptions.GRID);
    const isGridView = view === constants.viewOptions.GRID;
    // Dialog state
    const [isUrlDialogOpen, setIsUrlDialogOpen] = React.useState(false);
    // Refs
    const fileInputRef = React.useRef(null);
    const uploadDropZoneRef = React.useRef(null);
    // Upload handlers
    const [uploadFilesStream] = api.useUploadFilesStreamMutation();
    const [uploadFromUrls] = api.useUploadFromUrlsMutation();
    const uploadFilesToFolder = async (files, folderId)=>{
        if (files.length === 0) return;
        const formData = new FormData();
        const fileInfoArray = [];
        files.forEach((file)=>{
            formData.append('files', file);
            fileInfoArray.push({
                name: file.name,
                caption: null,
                alternativeText: null,
                folder: folderId
            });
        });
        formData.append('fileInfo', JSON.stringify(fileInfoArray));
        try {
            await uploadFilesStream({
                formData,
                totalFiles: files.length
            }).unwrap();
        } catch (error) {
        // Error is already dispatched to store from the API queryFn
        }
    };
    const handleFileSelect = ()=>{
        fileInputRef.current?.click();
    };
    const handleFileChange = async (e)=>{
        const files = e.target.files;
        if (files && files.length > 0) {
            await uploadFilesToFolder(Array.from(files), currentFolderId);
        }
        e.target.value = '';
    };
    const handleDrop = async (files)=>{
        await uploadFilesToFolder(files, currentFolderId);
    };
    const handleUrlUpload = async (urls)=>{
        try {
            await uploadFromUrls({
                urls,
                folderId: currentFolderId
            }).unwrap();
        } catch (error) {
        // Error is already dispatched to store from the API queryFn
        }
    };
    return /*#__PURE__*/ jsxRuntime.jsxs(jsxRuntime.Fragment, {
        children: [
            /*#__PURE__*/ jsxRuntime.jsx(UploadDropZoneContext.UploadDropZoneProvider, {
                onDrop: handleDrop,
                children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Box, {
                    ref: uploadDropZoneRef,
                    children: /*#__PURE__*/ jsxRuntime.jsxs(strapiAdmin.Layouts.Root, {
                        minHeight: "100vh",
                        background: "neutral0",
                        children: [
                            /*#__PURE__*/ jsxRuntime.jsx(designSystem.VisuallyHidden, {
                                children: /*#__PURE__*/ jsxRuntime.jsx("input", {
                                    type: "file",
                                    ref: fileInputRef,
                                    onChange: handleFileChange,
                                    multiple: true
                                })
                            }),
                            /*#__PURE__*/ jsxRuntime.jsx(HeaderWrapper, {
                                children: /*#__PURE__*/ jsxRuntime.jsx(strapiAdmin.Layouts.Header, {
                                    title: `${title} (${itemCountLabel})`,
                                    primaryAction: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.SimpleMenu, {
                                        popoverPlacement: "bottom-end",
                                        variant: "default",
                                        endIcon: /*#__PURE__*/ jsxRuntime.jsx(icons.ChevronDown, {}),
                                        label: formatMessage({
                                            id: translations.getTranslationKey('new'),
                                            defaultMessage: 'New'
                                        }),
                                        children: [
                                            /*#__PURE__*/ jsxRuntime.jsx(designSystem.MenuItem, {
                                                onSelect: ()=>setIsCreateFolderDialogOpen(true),
                                                startIcon: /*#__PURE__*/ jsxRuntime.jsx(icons.Folder, {}),
                                                children: formatMessage({
                                                    id: translations.getTranslationKey('folder.create.title'),
                                                    defaultMessage: 'New folder'
                                                })
                                            }),
                                            /*#__PURE__*/ jsxRuntime.jsx(designSystem.MenuItem, {
                                                onSelect: handleFileSelect,
                                                startIcon: /*#__PURE__*/ jsxRuntime.jsx(icons.Files, {}),
                                                children: formatMessage({
                                                    id: translations.getTranslationKey('import-files'),
                                                    defaultMessage: 'Import files'
                                                })
                                            }),
                                            /*#__PURE__*/ jsxRuntime.jsx(designSystem.MenuItem, {
                                                onSelect: ()=>setIsUrlDialogOpen(true),
                                                startIcon: /*#__PURE__*/ jsxRuntime.jsx(icons.Link, {}),
                                                children: formatMessage({
                                                    id: translations.getTranslationKey('import-from-url'),
                                                    defaultMessage: 'Import from URL'
                                                })
                                            })
                                        ]
                                    }),
                                    subtitle: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                                        justifyContent: "space-between",
                                        alignItems: "center",
                                        gap: 4,
                                        width: "100%",
                                        children: [
                                            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
                                                gap: 4,
                                                alignItems: "center",
                                                children: "TODO: Filters and search"
                                            }),
                                            /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                                                gap: 4,
                                                alignItems: "center",
                                                children: [
                                                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Box, {
                                                        children: "TODO: Sort"
                                                    }),
                                                    /*#__PURE__*/ jsxRuntime.jsxs(StyledToggleGroup, {
                                                        type: "single",
                                                        value: isGridView ? 'grid' : 'table',
                                                        onValueChange: (value)=>value && setView(value === 'grid' ? constants.viewOptions.GRID : constants.viewOptions.TABLE),
                                                        "aria-label": formatMessage({
                                                            id: translations.getTranslationKey('view.switch.label'),
                                                            defaultMessage: 'View options'
                                                        }),
                                                        children: [
                                                            /*#__PURE__*/ jsxRuntime.jsxs(StyledToggleItem, {
                                                                value: "table",
                                                                "aria-label": formatMessage({
                                                                    id: translations.getTranslationKey('view.table'),
                                                                    defaultMessage: 'Table view'
                                                                }),
                                                                children: [
                                                                    /*#__PURE__*/ jsxRuntime.jsx(icons.List, {}),
                                                                    formatMessage({
                                                                        id: translations.getTranslationKey('view.table'),
                                                                        defaultMessage: 'Table view'
                                                                    })
                                                                ]
                                                            }),
                                                            /*#__PURE__*/ jsxRuntime.jsxs(StyledToggleItem, {
                                                                value: "grid",
                                                                "aria-label": formatMessage({
                                                                    id: translations.getTranslationKey('view.grid'),
                                                                    defaultMessage: 'Grid view'
                                                                }),
                                                                children: [
                                                                    /*#__PURE__*/ jsxRuntime.jsx(icons.GridFour, {}),
                                                                    formatMessage({
                                                                        id: translations.getTranslationKey('view.grid'),
                                                                        defaultMessage: 'Grid view'
                                                                    })
                                                                ]
                                                            })
                                                        ]
                                                    })
                                                ]
                                            })
                                        ]
                                    })
                                })
                            }),
                            /*#__PURE__*/ jsxRuntime.jsx(strapiAdmin.Layouts.Content, {
                                children: /*#__PURE__*/ jsxRuntime.jsxs(UploadDropZone.DropZoneWithOverlay, {
                                    children: [
                                        /*#__PURE__*/ jsxRuntime.jsx(UploadDropZone.DropFilesMessage, {
                                            uploadDropZoneRef: uploadDropZoneRef,
                                            folderName: title
                                        }),
                                        /*#__PURE__*/ jsxRuntime.jsx(AssetsView, {
                                            view: view,
                                            folderId: currentFolderId,
                                            onAssetItemClick: openDetails
                                        })
                                    ]
                                })
                            })
                        ]
                    })
                })
            }),
            /*#__PURE__*/ jsxRuntime.jsx(CreateFolderDialog.CreateFolderDialog, {
                open: isCreateFolderDialogOpen,
                folderName: title,
                parentFolderId: currentFolderId,
                onClose: ()=>setIsCreateFolderDialogOpen(false)
            }),
            /*#__PURE__*/ jsxRuntime.jsx(ImportFromUrlDialog.ImportFromUrlDialog, {
                open: isUrlDialogOpen,
                onClose: ()=>setIsUrlDialogOpen(false),
                onUpload: handleUrlUpload
            }),
            /*#__PURE__*/ jsxRuntime.jsx(AssetDetailsDrawer.AssetDetailsDrawer, {})
        ]
    });
};

exports.AssetsPage = AssetsPage;
