// @strapi/upload@5.42.1/dist/admin/future/pages/Assets/constants.mjs
import { getTranslationKey } from '../../utils/translations.mjs';

const localStorageKeys = {
    view: `STRAPI_UPLOAD_LIBRARY_VIEW`
};
const viewOptions = {
    GRID: 0,
    TABLE: 1
};
const TABLE_HEADERS = [
    {
        name: 'name',
        label: {
            id: getTranslationKey('list.table.header.name'),
            defaultMessage: 'name'
        }
    },
    {
        name: 'createdAt',
        label: {
            id: getTranslationKey('list.table.header.creationDate'),
            defaultMessage: 'creation date'
        }
    },
    {
        name: 'updatedAt',
        label: {
            id: getTranslationKey('list.table.header.lastModified'),
            defaultMessage: 'last modified'
        }
    },
    {
        name: 'size',
        label: {
            id: getTranslationKey('list.table.header.size'),
            defaultMessage: 'size'
        }
    },
    {
        name: 'actions',
        label: {
            id: getTranslationKey('list.table.header.actions'),
            defaultMessage: 'actions'
        },
        isVisuallyHidden: true
    }
];

export { TABLE_HEADERS, localStorageKeys, viewOptions };
