// @strapi/utils@5.42.1/dist/validate/index.mjs
import { isArray, isObject } from 'lodash/fp';
import { getNonWritableAttributes, constants } from '../content-types.mjs';
import { ALLOWED_QUERY_PARAM_KEYS } from '../content-api-constants.mjs';
import { getExtraQueryKeysFromRoute, getExtraRootKeysFromRouteBody } from '../content-api-route-params.mjs';
import { pipe } from '../async.mjs';
import { throwInvalidKey } from './utils.mjs';
import * as index from './visitors/index.mjs';
export { index as visitors };
import { defaultValidatePopulate, defaultValidateFields, defaultValidateSort, defaultValidateFilters } from './validators.mjs';
import * as validators from './validators.mjs';
export { validators };
import traverseEntity from '../traverse-entity.mjs';
import traverseQueryFilters from '../traverse/query-filters.mjs';
import traverseQuerySort from '../traverse/query-sort.mjs';
import traverseQueryPopulate from '../traverse/query-populate.mjs';
import '../traverse/query-fields.mjs';
import { ValidationError } from '../errors.mjs';
import throwUnrecognizedFields from './visitors/throw-unrecognized-fields.mjs';
import throwRestrictedRelations from './visitors/throw-restricted-relations.mjs';
import throwRestrictedFields from './visitors/throw-restricted-fields.mjs';

const { ID_ATTRIBUTE, DOC_ID_ATTRIBUTE } = constants;
const createAPIValidators = (opts)=>{
    const { getModel } = opts || {};
    const validateInput = async (data, schema, options = {})=>{
        const { auth, route } = options;
        if (!schema) {
            throw new Error('Missing schema in validateInput');
        }
        if (isArray(data)) {
            await Promise.all(data.map((entry)=>validateInput(entry, schema, options)));
            return;
        }
        const allowedExtraRootKeys = getExtraRootKeysFromRouteBody(route);
        const nonWritableAttributes = getNonWritableAttributes(schema);
        const transforms = [
            (data)=>{
                if (isObject(data)) {
                    if (ID_ATTRIBUTE in data) {
                        throwInvalidKey({
                            key: ID_ATTRIBUTE
                        });
                    }
                    if (DOC_ID_ATTRIBUTE in data) {
                        throwInvalidKey({
                            key: DOC_ID_ATTRIBUTE
                        });
                    }
                }
                return data;
            },
            // non-writable attributes
            traverseEntity(throwRestrictedFields(nonWritableAttributes), {
                schema,
                getModel
            }),
            // unrecognized attributes (allowedExtraRootKeys = registered input param keys)
            traverseEntity(throwUnrecognizedFields, {
                schema,
                getModel,
                allowedExtraRootKeys
            })
        ];
        if (auth) {
            // restricted relations
            transforms.push(traverseEntity(throwRestrictedRelations(auth), {
                schema,
                getModel
            }));
        }
        // Apply validators from registry if exists
        opts?.validators?.input?.forEach((validator)=>transforms.push(validator(schema)));
        try {
            await pipe(...transforms)(data);
            // Validate extra root keys from route's body schema with Zod (throw on failure).
            //
            // Content-api sends the document payload as body.data; the controller calls validateInput(body.data, ctx),
            // so the input we receive here is the inner payload (keys like "relatedMedia", "name"), not the full body.
            // The route's body schema is z.object({ data: ... }), so its shape includes "data". We skip "data" because
            // the main document payload is already validated above by traverseEntity (throwUnrecognizedFields, etc.);
            // relation ops (connect/disconnect/set) are handled there, not by the route's Zod schema. We only run
            // Zod here for truly extra root keys added via addInputParams (e.g. clientMutationId).
            if (isObject(data) && route?.request?.body?.['application/json']) {
                const bodySchema = route.request.body['application/json'];
                if (typeof bodySchema === 'object' && 'shape' in bodySchema) {
                    const shape = bodySchema.shape;
                    const dataObj = data;
                    for (const key of Object.keys(shape)){
                        if (key === 'data' || !(key in dataObj)) continue;
                        const zodSchema = shape[key];
                        if (zodSchema && typeof zodSchema.parse === 'function') {
                            const result = zodSchema.safeParse(dataObj[key]);
                            if (!result.success) {
                                throw new ValidationError(result.error?.message ?? 'Validation failed', {
                                    key,
                                    path: null,
                                    source: 'body',
                                    param: key
                                });
                            }
                        }
                    }
                }
            }
        } catch (e) {
            if (e instanceof ValidationError) {
                e.details.source = 'body';
            }
            throw e;
        }
    };
    const validateQuery = async (query, schema, { auth, strictParams = false, route } = {})=>{
        if (!schema) {
            throw new Error('Missing schema in validateQuery');
        }
        if (strictParams) {
            const extraQueryKeys = getExtraQueryKeysFromRoute(route);
            const allowedKeys = [
                ...ALLOWED_QUERY_PARAM_KEYS,
                ...extraQueryKeys
            ];
            for (const key of Object.keys(query)){
                if (!allowedKeys.includes(key)) {
                    try {
                        throwInvalidKey({
                            key,
                            path: null
                        });
                    } catch (e) {
                        if (e instanceof ValidationError) {
                            e.details.source = 'query';
                            e.details.param = key;
                        }
                        throw e;
                    }
                }
            }
            // Validate extra query keys from route's request schema with Zod (throw on failure)
            const routeQuerySchema = route?.request?.query;
            if (routeQuerySchema) {
                for (const key of extraQueryKeys){
                    if (key in query) {
                        const zodSchema = routeQuerySchema[key];
                        if (zodSchema && typeof zodSchema.parse === 'function') {
                            const result = zodSchema.safeParse(query[key]);
                            if (!result.success) {
                                throw new ValidationError(result.error?.message ?? 'Invalid query param', {
                                    key,
                                    path: null,
                                    source: 'query',
                                    param: key
                                });
                            }
                        }
                    }
                }
            }
        }
        const { filters, sort, fields, populate } = query;
        if (filters) {
            await validateFilters(filters, schema, {
                auth
            });
        }
        if (sort) {
            await validateSort(sort, schema, {
                auth
            });
        }
        if (fields) {
            await validateFields(fields, schema);
        }
        // a wildcard is always valid; its conversion will be handled by the entity service and can be optimized with sanitizer
        if (populate && populate !== '*') {
            await validatePopulate(populate, schema);
        }
    };
    const validateFilters = async (filters, schema, { auth } = {})=>{
        if (!schema) {
            throw new Error('Missing schema in validateFilters');
        }
        if (isArray(filters)) {
            await Promise.all(filters.map((filter)=>validateFilters(filter, schema, {
                    auth
                })));
            return;
        }
        const transforms = [
            defaultValidateFilters({
                schema,
                getModel
            })
        ];
        if (auth) {
            transforms.push(traverseQueryFilters(throwRestrictedRelations(auth), {
                schema,
                getModel
            }));
        }
        try {
            await pipe(...transforms)(filters);
        } catch (e) {
            if (e instanceof ValidationError) {
                e.details.source = 'query';
                e.details.param = 'filters';
            }
            throw e;
        }
    };
    const validateSort = async (sort, schema, { auth } = {})=>{
        if (!schema) {
            throw new Error('Missing schema in validateSort');
        }
        const transforms = [
            defaultValidateSort({
                schema,
                getModel
            })
        ];
        if (auth) {
            transforms.push(traverseQuerySort(throwRestrictedRelations(auth), {
                schema,
                getModel
            }));
        }
        try {
            await pipe(...transforms)(sort);
        } catch (e) {
            if (e instanceof ValidationError) {
                e.details.source = 'query';
                e.details.param = 'sort';
            }
            throw e;
        }
    };
    const validateFields = async (fields, schema)=>{
        if (!schema) {
            throw new Error('Missing schema in validateFields');
        }
        const transforms = [
            defaultValidateFields({
                schema,
                getModel
            })
        ];
        try {
            await pipe(...transforms)(fields);
        } catch (e) {
            if (e instanceof ValidationError) {
                e.details.source = 'query';
                e.details.param = 'fields';
            }
            throw e;
        }
    };
    const validatePopulate = async (populate, schema, { auth } = {})=>{
        if (!schema) {
            throw new Error('Missing schema in sanitizePopulate');
        }
        const transforms = [
            defaultValidatePopulate({
                schema,
                getModel
            })
        ];
        if (auth) {
            transforms.push(traverseQueryPopulate(throwRestrictedRelations(auth), {
                schema,
                getModel
            }));
        }
        try {
            await pipe(...transforms)(populate);
        } catch (e) {
            if (e instanceof ValidationError) {
                e.details.source = 'query';
                e.details.param = 'populate';
            }
            throw e;
        }
    };
    return {
        input: validateInput,
        query: validateQuery,
        filters: validateFilters,
        sort: validateSort,
        fields: validateFields,
        populate: validatePopulate
    };
};

export { createAPIValidators };
