// @strapi/admin@5.42.1/dist/admin/admin/tests/utils.mjs
import { jsx, Fragment } from 'react/jsx-runtime';
import * as React from 'react';
import { configureStore } from '@reduxjs/toolkit';
import { fixtures } from '@strapi/admin-test-utils';
import { lightTheme, darkTheme } from '@strapi/design-system';
import { render as render$1, renderHook as renderHook$1 } from '@testing-library/react';
export { act, fireEvent, screen, waitFor } from '@testing-library/react';
import { userEvent } from '@testing-library/user-event';
import { DndProvider } from 'react-dnd';
import { HTML5Backend } from 'react-dnd-html5-backend';
import { setLogger, QueryClient, QueryClientProvider } from 'react-query';
import { Provider } from 'react-redux';
import { createMemoryRouter, RouterProvider } from 'react-router-dom';
import { GuidedTourContext } from '../src/components/GuidedTour/Context.mjs';
import { LanguageProvider } from '../src/components/LanguageProvider.mjs';
import { Theme } from '../src/components/Theme.mjs';
import { RBAC } from '../src/core/apis/rbac.mjs';
import { AppInfoProvider } from '../src/features/AppInfo.mjs';
import { AuthProvider } from '../src/features/Auth.mjs';
import { _internalConfigurationContextProvider as ConfigurationContextProvider } from '../src/features/Configuration.mjs';
import { NotificationsProvider } from '../src/features/Notifications.mjs';
import { StrapiAppProvider } from '../src/features/StrapiApp.mjs';
import { reducer } from '../src/reducer.mjs';
import { adminApi } from '../src/services/api.mjs';
export { server } from './server.mjs';
import { initialState } from './store.mjs';

setLogger({
    log: ()=>{},
    warn: ()=>{},
    error: ()=>{}
});
/**
 * RTL's `rerender()` updates the `children` passed to our test `render()` helper.
 * However, we keep a stable `routerRef` to avoid recreating the data router (which
 * can crash React Router). That means the route element tree would otherwise keep
 * rendering the initial children forever.
 *
 * We solve this by passing `children` via React context, so consumers inside the
 * router element tree still update on rerenders without recreating the router.
 */ const TestChildrenContext = /*#__PURE__*/ React.createContext(null);
const TestChildrenOutlet = ()=>{
    return /*#__PURE__*/ jsx(Fragment, {
        children: React.useContext(TestChildrenContext)
    });
};
const defaultTestStoreConfig = ()=>({
        preloadedState: initialState(),
        reducer: {
            [adminApi.reducerPath]: adminApi.reducer,
            admin_app: reducer
        },
        // @ts-expect-error – this fails.
        middleware: (getDefaultMiddleware)=>[
                ...getDefaultMiddleware({
                    // Disable timing checks for test env
                    immutableCheck: false,
                    serializableCheck: false
                }),
                adminApi.middleware
            ]
    });
const DEFAULT_PERMISSIONS = [
    ...fixtures.permissions.allPermissions,
    {
        id: 314,
        action: 'admin::users.read',
        subject: null,
        properties: {},
        conditions: [],
        actionParameters: {}
    }
];
const Providers = ({ children, initialEntries, storeConfig, permissions = [] })=>{
    /**
   * Keep these instances stable for the lifetime of the render.
   * Recreating the data router on rerenders can crash React Router.
   */ const queryClientRef = React.useRef(null);
    const storeRef = React.useRef(null);
    const routerRef = React.useRef(null);
    if (queryClientRef.current === null) {
        queryClientRef.current = new QueryClient({
            defaultOptions: {
                queries: {
                    retry: false
                }
            }
        });
    }
    if (storeRef.current === null) {
        storeRef.current = configureStore({
            ...defaultTestStoreConfig(),
            ...storeConfig
        });
    }
    let allPermissions;
    if (typeof permissions === 'function') {
        /**
     * The callback form can return `undefined` to mean "use defaults".
     * Guard against passing `undefined` into AuthProvider.
     */ allPermissions = permissions(DEFAULT_PERMISSIONS) ?? DEFAULT_PERMISSIONS;
    } else {
        allPermissions = [
            ...DEFAULT_PERMISSIONS,
            ...permissions
        ];
    }
    if (routerRef.current === null) {
        routerRef.current = createMemoryRouter([
            {
                path: '/*',
                element: /*#__PURE__*/ jsx(StrapiAppProvider, {
                    components: {},
                    rbac: new RBAC(),
                    widgets: {
                        widgets: [],
                        getAll: jest.fn(),
                        register: jest.fn()
                    },
                    customFields: {
                        customFields: {},
                        get: jest.fn().mockReturnValue({
                            name: 'color',
                            pluginId: 'mycustomfields',
                            type: 'text',
                            icon: jest.fn(),
                            intlLabel: {
                                id: 'mycustomfields.color.label',
                                defaultMessage: 'Color'
                            },
                            intlDescription: {
                                id: 'mycustomfields.color.description',
                                defaultMessage: 'Select any color'
                            },
                            components: {
                                Input: jest.fn().mockResolvedValue({
                                    default: jest.fn()
                                })
                            }
                        }),
                        getAll: jest.fn(),
                        register: jest.fn()
                    },
                    fields: {},
                    menu: [],
                    getAdminInjectedComponents: jest.fn(),
                    getPlugin: jest.fn(),
                    plugins: {},
                    runHookParallel: jest.fn(),
                    runHookWaterfall: jest.fn().mockImplementation((_name, initialValue)=>initialValue),
                    runHookSeries: jest.fn(),
                    settings: {},
                    children: /*#__PURE__*/ jsx(Provider, {
                        store: storeRef.current,
                        children: /*#__PURE__*/ jsx(AuthProvider, {
                            _defaultPermissions: allPermissions,
                            _disableRenewToken: true,
                            children: /*#__PURE__*/ jsx(QueryClientProvider, {
                                client: queryClientRef.current,
                                children: /*#__PURE__*/ jsx(DndProvider, {
                                    backend: HTML5Backend,
                                    children: /*#__PURE__*/ jsx(LanguageProvider, {
                                        messages: {},
                                        children: /*#__PURE__*/ jsx(Theme, {
                                            themes: {
                                                dark: darkTheme,
                                                light: lightTheme
                                            },
                                            children: /*#__PURE__*/ jsx(NotificationsProvider, {
                                                children: /*#__PURE__*/ jsx(GuidedTourContext, {
                                                    enabled: false,
                                                    children: /*#__PURE__*/ jsx(ConfigurationContextProvider, {
                                                        showReleaseNotification: false,
                                                        logos: {
                                                            auth: {
                                                                default: 'default'
                                                            },
                                                            menu: {
                                                                default: 'default'
                                                            }
                                                        },
                                                        updateProjectSettings: jest.fn(),
                                                        children: /*#__PURE__*/ jsx(AppInfoProvider, {
                                                            autoReload: true,
                                                            useYarn: true,
                                                            dependencies: {
                                                                '@strapi/plugin-documentation': '4.2.0',
                                                                '@strapi/provider-upload-cloudinary': '4.2.0'
                                                            },
                                                            strapiVersion: "4.1.0",
                                                            communityEdition: true,
                                                            shouldUpdateStrapi: false,
                                                            children: /*#__PURE__*/ jsx(TestChildrenOutlet, {})
                                                        })
                                                    })
                                                })
                                            })
                                        })
                                    })
                                })
                            })
                        })
                    })
                })
            }
        ], {
            initialEntries
        });
    }
    // en is the default locale of the admin app.
    return /*#__PURE__*/ jsx(TestChildrenContext.Provider, {
        value: children,
        children: /*#__PURE__*/ jsx(RouterProvider, {
            router: routerRef.current
        })
    });
};
// eslint-disable-next-line react/jsx-no-useless-fragment
const fallbackWrapper = ({ children })=>/*#__PURE__*/ jsx(Fragment, {
        children: children
    });
/**
 * @alpha
 * @description A custom render function that wraps the component with the necessary providers,
 * for use of testing components within the Strapi Admin.
 */ const render = (ui, { renderOptions, userEventOptions, initialEntries, providerOptions } = {})=>{
    const { wrapper: Wrapper = fallbackWrapper, ...restOptions } = renderOptions ?? {};
    return {
        ...render$1(ui, {
            wrapper: ({ children })=>/*#__PURE__*/ jsx(Providers, {
                    initialEntries: initialEntries,
                    ...providerOptions,
                    children: /*#__PURE__*/ jsx(Wrapper, {
                        children: children
                    })
                }),
            ...restOptions
        }),
        user: userEvent.setup({
            skipHover: true,
            ...userEventOptions
        })
    };
};
/**
 * @alpha
 * @description A custom render-hook function that wraps the component with the necessary providers,
 * for use of testing hooks within the Strapi Admin.
 */ const renderHook = (hook, options)=>{
    const { wrapper: Wrapper = fallbackWrapper, initialEntries, providerOptions, ...restOptions } = options ?? {};
    return renderHook$1(hook, {
        wrapper: ({ children })=>/*#__PURE__*/ jsx(Providers, {
                initialEntries: initialEntries,
                ...providerOptions,
                children: /*#__PURE__*/ jsx(Wrapper, {
                    children: children
                })
            }),
        ...restOptions
    });
};

export { defaultTestStoreConfig, render, renderHook };
