// @strapi/utils/dist/traverse-entity.js
'use strict';

var fp = require('lodash/fp');
var contentTypes = require('./content-types.js');

/**
 * Execute promises in parallel but throw errors in array index order.
 */ const parallelWithOrderedErrors = async (promises)=>{
    const results = await Promise.allSettled(promises);
    // Throw first error in array index order (matches sequential behavior)
    for(let i = 0; i < results.length; i += 1){
        const result = results[i];
        if (result.status === 'rejected') {
            throw result.reason;
        }
    }
    return results.map((r)=>r.value);
};
const traverseEntity = async (visitor, options, entity)=>{
    const { path = {
        raw: null,
        attribute: null,
        rawWithIndices: null
    }, schema, getModel, allowedExtraRootKeys } = options;
    let parent = options.parent;
    const traverseMorphRelationTarget = async (visitor, path, entry)=>{
        const targetSchema = getModel(entry.__type);
        const traverseOptions = {
            schema: targetSchema,
            path,
            getModel,
            parent,
            allowedExtraRootKeys
        };
        return traverseEntity(visitor, traverseOptions, entry);
    };
    const traverseRelationTarget = (schema)=>async (visitor, path, entry)=>{
            const traverseOptions = {
                schema,
                path,
                getModel,
                parent,
                allowedExtraRootKeys
            };
            return traverseEntity(visitor, traverseOptions, entry);
        };
    const traverseMediaTarget = async (visitor, path, entry)=>{
        const targetSchemaUID = 'plugin::upload.file';
        const targetSchema = getModel(targetSchemaUID);
        const traverseOptions = {
            schema: targetSchema,
            path,
            getModel,
            parent,
            allowedExtraRootKeys
        };
        return traverseEntity(visitor, traverseOptions, entry);
    };
    const traverseComponent = async (visitor, path, schema, entry)=>{
        const traverseOptions = {
            schema,
            path,
            getModel,
            parent,
            allowedExtraRootKeys
        };
        return traverseEntity(visitor, traverseOptions, entry);
    };
    const visitDynamicZoneEntry = async (visitor, path, entry)=>{
        const targetSchema = getModel(entry.__component);
        const traverseOptions = {
            schema: targetSchema,
            path,
            getModel,
            parent,
            allowedExtraRootKeys
        };
        return traverseEntity(visitor, traverseOptions, entry);
    };
    // End recursion
    if (!fp.isObject(entity) || fp.isNil(schema)) {
        return entity;
    }
    // Don't mutate the original entity object
    // only clone at 1st level as the next level will get clone when traversed
    const copy = fp.clone(entity);
    const visitorUtils = createVisitorUtils({
        data: copy
    });
    const keys = Object.keys(copy);
    for(let i = 0; i < keys.length; i += 1){
        const key = keys[i];
        // Retrieve the attribute definition associated to the key from the schema
        const attribute = schema.attributes[key];
        const newPath = {
            ...path
        };
        newPath.raw = fp.isNil(path.raw) ? key : `${path.raw}.${key}`;
        newPath.rawWithIndices = fp.isNil(path.rawWithIndices) ? key : `${path.rawWithIndices}.${key}`;
        if (!fp.isNil(attribute)) {
            newPath.attribute = fp.isNil(path.attribute) ? key : `${path.attribute}.${key}`;
        }
        // Visit the current attribute
        const visitorOptions = {
            data: copy,
            schema,
            key,
            value: copy[key],
            attribute,
            path: newPath,
            getModel,
            parent,
            allowedExtraRootKeys
        };
        await visitor(visitorOptions, visitorUtils);
        // Extract the value for the current key (after calling the visitor)
        const value = copy[key];
        // Ignore Nil values or attributes
        if (fp.isNil(value) || fp.isNil(attribute)) {
            continue;
        }
        if (contentTypes.isRelationalAttribute(attribute)) {
            parent = {
                schema,
                key,
                attribute,
                path: newPath
            };
            const isMorphRelation = attribute.relation.toLowerCase().startsWith('morph');
            const method = isMorphRelation ? traverseMorphRelationTarget : traverseRelationTarget(getModel(attribute.target));
            if (fp.isArray(value)) {
                // Process array items in parallel with ordered error handling
                copy[key] = await parallelWithOrderedErrors(value.map((item, i)=>{
                    const arrayPath = {
                        ...newPath,
                        rawWithIndices: fp.isNil(newPath.rawWithIndices) ? `${i}` : `${newPath.rawWithIndices}.${i}`
                    };
                    return method(visitor, arrayPath, item);
                }));
            } else {
                copy[key] = await method(visitor, newPath, value);
            }
            continue;
        }
        if (contentTypes.isMediaAttribute(attribute)) {
            parent = {
                schema,
                key,
                attribute,
                path: newPath
            };
            if (fp.isArray(value)) {
                // Process media array items in parallel with ordered error handling
                copy[key] = await parallelWithOrderedErrors(value.map((item, i)=>{
                    const arrayPath = {
                        ...newPath,
                        rawWithIndices: fp.isNil(newPath.rawWithIndices) ? `${i}` : `${newPath.rawWithIndices}.${i}`
                    };
                    return traverseMediaTarget(visitor, arrayPath, item);
                }));
            } else {
                copy[key] = await traverseMediaTarget(visitor, newPath, value);
            }
            continue;
        }
        if (attribute.type === 'component') {
            parent = {
                schema,
                key,
                attribute,
                path: newPath
            };
            const targetSchema = getModel(attribute.component);
            if (fp.isArray(value)) {
                // Process component array items in parallel with ordered error handling
                copy[key] = await parallelWithOrderedErrors(value.map((item, i)=>{
                    const arrayPath = {
                        ...newPath,
                        rawWithIndices: fp.isNil(newPath.rawWithIndices) ? `${i}` : `${newPath.rawWithIndices}.${i}`
                    };
                    return traverseComponent(visitor, arrayPath, targetSchema, item);
                }));
            } else {
                copy[key] = await traverseComponent(visitor, newPath, targetSchema, value);
            }
            continue;
        }
        if (attribute.type === 'dynamiczone' && fp.isArray(value)) {
            parent = {
                schema,
                key,
                attribute,
                path: newPath
            };
            // Process dynamic zone items in parallel with ordered error handling
            copy[key] = await parallelWithOrderedErrors(value.map((item, i)=>{
                const arrayPath = {
                    ...newPath,
                    rawWithIndices: fp.isNil(newPath.rawWithIndices) ? `${i}` : `${newPath.rawWithIndices}.${i}`
                };
                return visitDynamicZoneEntry(visitor, arrayPath, item);
            }));
            continue;
        }
    }
    return copy;
};
const createVisitorUtils = ({ data })=>({
        remove (key) {
            delete data[key];
        },
        set (key, value) {
            data[key] = value;
        }
    });
var traverseEntity$1 = fp.curry(traverseEntity);

module.exports = traverseEntity$1;
