// @strapi/content-manager@5.42.1/dist/admin/pages/EditView/components/FormInputs/Component/NonRepeatable.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var strapiAdmin = require('@strapi/admin/strapi-admin');
var designSystem = require('@strapi/design-system');
var reactIntl = require('react-intl');
var objects = require('../../../../../utils/objects.js');
var FormLayout = require('../../FormLayout.js');
var ComponentContext = require('../ComponentContext.js');

function _interopNamespaceDefault(e) {
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespaceDefault(React);

const NonRepeatableComponent = ({ attribute, name, children, layout })=>{
    const componentId = strapiAdmin.useForm('NonRepeatableComponent', (state)=>objects.getIn(state.values, `${name}.id`));
    const level = ComponentContext.useComponent('NonRepeatableComponent', (state)=>state.level);
    const isNested = level > 0;
    const isMobile = strapiAdmin.useIsMobile();
    return /*#__PURE__*/ jsxRuntime.jsx(ComponentContext.ComponentProvider, {
        id: componentId,
        uid: attribute.component,
        level: level + 1,
        type: "component",
        children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Box, {
            background: 'neutral100',
            padding: {
                initial: 4,
                medium: 6
            },
            hasRadius: isNested,
            borderColor: isNested || isMobile ? 'neutral200' : undefined,
            children: /*#__PURE__*/ jsxRuntime.jsx(NonRepeatableComponentFields, {
                attribute: attribute,
                name: name,
                layout: layout,
                children: children
            })
        })
    });
};
const NonRepeatableComponentFields = /*#__PURE__*/ React__namespace.memo(({ attribute, children, layout, name })=>{
    const { formatMessage } = reactIntl.useIntl();
    return /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
        direction: "column",
        alignItems: "stretch",
        gap: 6,
        children: layout.map((row, index)=>{
            return /*#__PURE__*/ jsxRuntime.jsx(FormLayout.ResponsiveGridRoot, {
                gap: {
                    initial: 3,
                    medium: 4
                },
                children: row.map(({ size, ...field })=>{
                    /**
                 * Layouts are built from schemas so they don't understand the complete
                 * schema tree, for components we append the parent name to the field name
                 * because this is the structure for the data & permissions also understand
                 * the nesting involved.
                 */ const completeFieldName = `${name}.${field.name}`;
                    const translatedLabel = formatMessage({
                        id: `content-manager.components.${attribute.component}.${field.name}`,
                        defaultMessage: field.label
                    });
                    return /*#__PURE__*/ jsxRuntime.jsx(FormLayout.ResponsiveGridItem, {
                        col: size,
                        s: 12,
                        xs: 12,
                        direction: "column",
                        alignItems: "stretch",
                        children: children({
                            ...field,
                            label: translatedLabel,
                            name: completeFieldName
                        })
                    }, completeFieldName);
                })
            }, index);
        })
    });
});
NonRepeatableComponentFields.displayName = 'NonRepeatableComponentFields';
const MemoizedNonRepeatableComponent = /*#__PURE__*/ React__namespace.memo(NonRepeatableComponent);

exports.NonRepeatableComponent = MemoizedNonRepeatableComponent;
