// @strapi/admin/dist/admin/admin/src/components/Layouts/ActionLayout.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
require('react');
var designSystem = require('@strapi/design-system');
var styled = require('styled-components');
var theme = require('../../constants/theme.js');

/** On mobile: first item stays left, rest stick to the right (margin-right: auto on first child). */ const EndActionsFlex = styled.styled(designSystem.Flex)`
  width: 100%;
  & > *:first-child {
    margin-right: auto;
  }

  ${({ theme })=>theme.breakpoints.medium} {
    width: auto;
    & > *:first-child {
      margin-right: 0;
    }
  }
`;
const ActionLayout = ({ startActions, endActions, bottomActions })=>{
    if (!startActions && !endActions) {
        return null;
    }
    return /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
        direction: {
            initial: 'column',
            medium: 'row'
        },
        gap: {
            initial: 4,
            medium: 3
        },
        alignItems: "flex-start",
        paddingBottom: {
            initial: 2,
            medium: 4
        },
        paddingTop: 0,
        paddingLeft: theme.RESPONSIVE_DEFAULT_SPACING,
        paddingRight: theme.RESPONSIVE_DEFAULT_SPACING,
        wrap: {
            initial: 'nowrap',
            medium: 'wrap'
        },
        children: [
            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
                gap: 2,
                wrap: "wrap",
                flex: 1,
                width: "100%",
                children: startActions
            }),
            /*#__PURE__*/ jsxRuntime.jsx(EndActionsFlex, {
                gap: {
                    initial: 3,
                    medium: 2
                },
                shrink: 0,
                wrap: "wrap",
                children: endActions
            }),
            bottomActions && /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
                gap: 2,
                wrap: "wrap",
                width: "100%",
                children: bottomActions
            })
        ]
    });
};

exports.ActionLayout = ActionLayout;
