// @strapi/content-manager/dist/admin/pages/EditView/components/Panels.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var strapiAdmin = require('@strapi/admin/strapi-admin');
var designSystem = require('@strapi/design-system');
var reactIntl = require('react-intl');
var reactRouterDom = require('react-router-dom');
var InjectionZone = require('../../../components/InjectionZone.js');
var useDocument = require('../../../hooks/useDocument.js');
var router = require('../../../router.js');
var DocumentActions = require('./DocumentActions.js');

function _interopNamespaceDefault(e) {
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespaceDefault(React);

const PanelsItems = ({ panels, setVisiblePanels, visiblePanelsLength })=>{
    React__namespace.useEffect(()=>{
        if (setVisiblePanels && visiblePanelsLength !== panels.length) {
            setVisiblePanels(panels);
        }
    }, [
        panels,
        panels.length,
        setVisiblePanels,
        visiblePanelsLength
    ]);
    return /*#__PURE__*/ jsxRuntime.jsx(jsxRuntime.Fragment, {
        children: panels.map(({ content, id, ...description })=>/*#__PURE__*/ jsxRuntime.jsx(Panel, {
                ...description,
                children: content
            }, id))
    });
};
/* -------------------------------------------------------------------------------------------------
 * Panels
 * -----------------------------------------------------------------------------------------------*/ const Panels = ({ withActions = true })=>{
    const isCloning = reactRouterDom.useMatch(router.CLONE_PATH) !== null;
    const [{ query: { status } }] = strapiAdmin.useQueryParams({
        status: 'draft'
    });
    const { model, id, document, meta, collectionType } = useDocument.useDoc();
    const plugins = strapiAdmin.useStrapiApp('Panels', (state)=>state.plugins);
    // Optional context consumer
    const setVisiblePanels = usePanelsContext('Panels', (s)=>s.setVisiblePanels, false);
    const visiblePanels = usePanelsContext('Panels', (s)=>s.visiblePanels, false);
    const allPanels = plugins['content-manager'].apis.getEditViewSidePanels();
    const filteredPanels = allPanels.filter((panel)=>panel.type !== 'actions');
    const panelsToDisplay = withActions ? allPanels : filteredPanels;
    const props = {
        activeTab: status,
        model,
        documentId: id,
        document: isCloning ? undefined : document,
        meta: isCloning ? undefined : meta,
        collectionType
    };
    return /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
        direction: "column",
        alignItems: "stretch",
        gap: 2,
        children: /*#__PURE__*/ jsxRuntime.jsx(strapiAdmin.DescriptionComponentRenderer, {
            props: props,
            descriptions: panelsToDisplay,
            children: (panels)=>/*#__PURE__*/ jsxRuntime.jsx(PanelsItems, {
                    panels: panels,
                    setVisiblePanels: setVisiblePanels,
                    visiblePanelsLength: visiblePanels?.length
                })
        })
    });
};
/* -------------------------------------------------------------------------------------------------
 * Default Action Panels (CE)
 * -----------------------------------------------------------------------------------------------*/ const ActionsPanel = ()=>{
    const { formatMessage } = reactIntl.useIntl();
    return {
        title: formatMessage({
            id: 'content-manager.containers.edit.panels.default.title',
            defaultMessage: 'Entry'
        }),
        content: /*#__PURE__*/ jsxRuntime.jsx(ActionsPanelContent, {})
    };
};
ActionsPanel.type = 'actions';
const ActionsPanelContent = ()=>{
    const isCloning = reactRouterDom.useMatch(router.CLONE_PATH) !== null;
    const [{ query: { status = 'draft' } }] = strapiAdmin.useQueryParams();
    const { model, id, document, meta, collectionType } = useDocument.useDoc();
    const plugins = strapiAdmin.useStrapiApp('ActionsPanel', (state)=>state.plugins);
    const props = {
        activeTab: status,
        model,
        documentId: id,
        document: isCloning ? undefined : document,
        meta: isCloning ? undefined : meta,
        collectionType
    };
    return /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
        direction: "column",
        gap: 2,
        width: "100%",
        children: [
            /*#__PURE__*/ jsxRuntime.jsx(strapiAdmin.DescriptionComponentRenderer, {
                props: props,
                descriptions: plugins['content-manager'].apis.getDocumentActions('panel'),
                children: (actions)=>/*#__PURE__*/ jsxRuntime.jsx(DocumentActions.DocumentActions, {
                        actions: actions
                    })
            }),
            /*#__PURE__*/ jsxRuntime.jsx(InjectionZone.InjectionZone, {
                area: "editView.right-links",
                slug: model
            })
        ]
    });
};
const Panel = /*#__PURE__*/ React__namespace.forwardRef(({ children, title }, ref)=>{
    const isDesktop = strapiAdmin.useIsDesktop();
    return /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
        ref: ref,
        tag: "aside",
        "aria-labelledby": "additional-information",
        background: {
            initial: 'transparent',
            large: 'neutral0'
        },
        borderColor: {
            initial: 'transparent',
            large: 'neutral150'
        },
        hasRadius: isDesktop,
        padding: {
            initial: 0,
            large: 4
        },
        shadow: {
            initial: 'none',
            large: 'tableShadow'
        },
        gap: {
            initial: 4,
            large: 3
        },
        direction: "column",
        justifyContent: "stretch",
        alignItems: "flex-start",
        children: [
            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                tag: "h2",
                variant: "sigma",
                textTransform: "uppercase",
                textColor: "neutral600",
                children: title
            }),
            children
        ]
    });
});
/* -------------------------------------------------------------------------------------------------
 * PanelsContext
 * -----------------------------------------------------------------------------------------------*/ const [PanelsProviderImpl, usePanelsContext] = strapiAdmin.createContext('PanelsContext');
const PanelsProvider = ({ children })=>{
    const [visiblePanels, setVisiblePanels] = React__namespace.useState([]);
    return /*#__PURE__*/ jsxRuntime.jsx(PanelsProviderImpl, {
        visiblePanels: visiblePanels,
        setVisiblePanels: setVisiblePanels,
        children: children
    });
};

exports.ActionsPanel = ActionsPanel;
exports.ActionsPanelContent = ActionsPanelContent;
exports.Panels = Panels;
exports.PanelsProvider = PanelsProvider;
exports.usePanelsContext = usePanelsContext;
