// @strapi/upload/dist/admin/future/pages/Assets/AssetsPage.mjs
import { jsxs, Fragment, jsx } from 'react/jsx-runtime';
import { useState, useRef, useCallback } from 'react';
import * as ToggleGroup from '@radix-ui/react-toggle-group';
import { usePersistentState, Layouts, useElementOnScreen } from '@strapi/admin/strapi-admin';
import { Box, VisuallyHidden, Flex, SimpleMenu, MenuItem, Loader, Typography } from '@strapi/design-system';
import { List, GridFour, Folder, Files, Link, ChevronDown } from '@strapi/icons';
import { useIntl } from 'react-intl';
import { styled } from 'styled-components';
import { useUploadFilesStreamMutation, useUploadFromUrlsMutation } from '../../services/api.mjs';
import { useGetFoldersQuery } from '../../services/folders.mjs';
import { getTranslationKey } from '../../utils/translations.mjs';
import { useAssetDetailsParam, AssetDetailsDrawer } from './components/AssetDetails/AssetDetailsDrawer.mjs';
import { AssetsGrid } from './components/AssetsGrid.mjs';
import { AssetsTable } from './components/AssetsTable.mjs';
import { CreateFolderDialog } from './components/CreateFolderDialog.mjs';
import { DropZoneWithOverlay, DropFilesMessage } from './components/DropZone/UploadDropZone.mjs';
import { UploadDropZoneProvider } from './components/DropZone/UploadDropZoneContext.mjs';
import { ImportFromUrlDialog } from './components/ImportFromUrlDialog.mjs';
import { localStorageKeys, viewOptions } from './constants.mjs';
import { useFolderInfo } from './hooks/useFolderInfo.mjs';
import { useFolderNavigation } from './hooks/useFolderNavigation.mjs';
import { useInfiniteAssets } from './hooks/useInfiniteAssets.mjs';

const INTERSECTION_OPTIONS = {
    threshold: 0.1
};
const AssetsView = ({ view, folderId, onAssetItemClick })=>{
    const { formatMessage } = useIntl();
    const { assets, isLoading: isLoadingAssets, isFetchingMore, hasNextPage, fetchNextPage, error } = useInfiniteAssets({
        folder: folderId
    });
    const { data: folders = [], isLoading: isLoadingFolders } = useGetFoldersQuery({
        parentId: folderId
    });
    const isGridView = view === viewOptions.GRID;
    const isLoading = isLoadingAssets || isLoadingFolders;
    const loadMoreRef = useElementOnScreen(useCallback((isVisible)=>{
        if (isVisible && hasNextPage && !isFetchingMore) {
            fetchNextPage();
        }
    }, [
        hasNextPage,
        isFetchingMore,
        fetchNextPage
    ]), INTERSECTION_OPTIONS);
    if (isLoading) {
        return /*#__PURE__*/ jsx(Flex, {
            justifyContent: "center",
            padding: 8,
            children: /*#__PURE__*/ jsx(Loader, {
                children: formatMessage({
                    id: 'app.loading',
                    defaultMessage: 'Loading...'
                })
            })
        });
    }
    if (error) {
        return /*#__PURE__*/ jsx(Box, {
            padding: 8,
            children: /*#__PURE__*/ jsx(Typography, {
                textColor: "danger600",
                children: formatMessage({
                    id: getTranslationKey('list.assets.error'),
                    defaultMessage: 'An error occurred while fetching assets.'
                })
            })
        });
    }
    if (folders.length === 0 && assets.length === 0) {
        return /*#__PURE__*/ jsx(Box, {
            padding: 8,
            children: /*#__PURE__*/ jsx(Typography, {
                textColor: "neutral600",
                children: formatMessage({
                    id: 'app.components.EmptyStateLayout.content-document',
                    defaultMessage: 'No content found'
                })
            })
        });
    }
    return /*#__PURE__*/ jsxs(Fragment, {
        children: [
            isGridView ? /*#__PURE__*/ jsx(AssetsGrid, {
                folders: folders,
                assets: assets,
                onAssetItemClick: onAssetItemClick
            }) : /*#__PURE__*/ jsx(AssetsTable, {
                assets: assets,
                folders: folders,
                onAssetItemClick: onAssetItemClick
            }),
            /*#__PURE__*/ jsx("div", {
                ref: loadMoreRef,
                style: {
                    height: 1
                }
            }),
            isFetchingMore && /*#__PURE__*/ jsx(Flex, {
                justifyContent: "center",
                padding: 4,
                children: /*#__PURE__*/ jsx(Loader, {
                    children: formatMessage({
                        id: getTranslationKey('list.assets.loading-more'),
                        defaultMessage: 'Loading more assets...'
                    })
                })
            })
        ]
    });
};
/* -------------------------------------------------------------------------------------------------
 * AssetsPage
 * -----------------------------------------------------------------------------------------------*/ const StyledToggleGroup = styled(ToggleGroup.Root)`
  display: flex;
  border: 1px solid ${({ theme })=>theme.colors.neutral200};
  border-radius: ${({ theme })=>theme.borderRadius};
  overflow: hidden;
`;
const StyledToggleItem = styled(ToggleGroup.Item)`
  display: flex;
  align-items: center;
  gap: ${({ theme })=>theme.spaces[2]};
  padding: ${({ theme })=>`${theme.spaces[2]} ${theme.spaces[4]}`};
  border: none;
  background: ${({ theme })=>theme.colors.neutral0};
  color: ${({ theme })=>theme.colors.neutral800};
  cursor: pointer;
  font-size: ${({ theme })=>theme.fontSizes[1]};
  font-weight: ${({ theme })=>theme.fontWeights.semiBold};

  &:hover {
    background: ${({ theme })=>theme.colors.primary100};
  }

  &[data-state='on'] {
    background: ${({ theme })=>theme.colors.neutral150};
  }

  svg {
    width: 1.6rem;
    height: 1.6rem;
  }
`;
const HeaderWrapper = styled(Box)`
  [data-strapi-header] {
    background: ${({ theme })=>theme.colors.neutral0};

    h1 {
      font-size: 1.8rem;
    }
  }
`;
const AssetsPage = ()=>{
    const { formatMessage } = useIntl();
    const { openDetails } = useAssetDetailsParam();
    const { currentFolderId } = useFolderNavigation();
    const { title, itemCount } = useFolderInfo(currentFolderId);
    const itemCountLabel = formatMessage({
        id: getTranslationKey('header.content.item-count'),
        defaultMessage: '{count, plural, =1 {# item} other {# items}}'
    }, {
        count: itemCount
    });
    const [isCreateFolderDialogOpen, setIsCreateFolderDialogOpen] = useState(false);
    // View state
    const [view, setView] = usePersistentState(localStorageKeys.view, viewOptions.GRID);
    const isGridView = view === viewOptions.GRID;
    // Dialog state
    const [isUrlDialogOpen, setIsUrlDialogOpen] = useState(false);
    // Refs
    const fileInputRef = useRef(null);
    const uploadDropZoneRef = useRef(null);
    // Upload handlers
    const [uploadFilesStream] = useUploadFilesStreamMutation();
    const [uploadFromUrls] = useUploadFromUrlsMutation();
    const uploadFilesToFolder = async (files, folderId)=>{
        if (files.length === 0) return;
        const formData = new FormData();
        const fileInfoArray = [];
        files.forEach((file)=>{
            formData.append('files', file);
            fileInfoArray.push({
                name: file.name,
                caption: null,
                alternativeText: null,
                folder: folderId
            });
        });
        formData.append('fileInfo', JSON.stringify(fileInfoArray));
        try {
            await uploadFilesStream({
                formData,
                totalFiles: files.length
            }).unwrap();
        } catch (error) {
        // Error is already dispatched to store from the API queryFn
        }
    };
    const handleFileSelect = ()=>{
        fileInputRef.current?.click();
    };
    const handleFileChange = async (e)=>{
        const files = e.target.files;
        if (files && files.length > 0) {
            await uploadFilesToFolder(Array.from(files), currentFolderId);
        }
        e.target.value = '';
    };
    const handleDrop = async (files)=>{
        await uploadFilesToFolder(files, currentFolderId);
    };
    const handleUrlUpload = async (urls)=>{
        try {
            await uploadFromUrls({
                urls,
                folderId: currentFolderId
            }).unwrap();
        } catch (error) {
        // Error is already dispatched to store from the API queryFn
        }
    };
    return /*#__PURE__*/ jsxs(Fragment, {
        children: [
            /*#__PURE__*/ jsx(UploadDropZoneProvider, {
                onDrop: handleDrop,
                children: /*#__PURE__*/ jsx(Box, {
                    ref: uploadDropZoneRef,
                    children: /*#__PURE__*/ jsxs(Layouts.Root, {
                        minHeight: "100vh",
                        background: "neutral0",
                        children: [
                            /*#__PURE__*/ jsx(VisuallyHidden, {
                                children: /*#__PURE__*/ jsx("input", {
                                    type: "file",
                                    ref: fileInputRef,
                                    onChange: handleFileChange,
                                    multiple: true
                                })
                            }),
                            /*#__PURE__*/ jsx(HeaderWrapper, {
                                children: /*#__PURE__*/ jsx(Layouts.Header, {
                                    title: `${title} (${itemCountLabel})`,
                                    primaryAction: /*#__PURE__*/ jsxs(SimpleMenu, {
                                        popoverPlacement: "bottom-end",
                                        variant: "default",
                                        endIcon: /*#__PURE__*/ jsx(ChevronDown, {}),
                                        label: formatMessage({
                                            id: getTranslationKey('new'),
                                            defaultMessage: 'New'
                                        }),
                                        children: [
                                            /*#__PURE__*/ jsx(MenuItem, {
                                                onSelect: ()=>setIsCreateFolderDialogOpen(true),
                                                startIcon: /*#__PURE__*/ jsx(Folder, {}),
                                                children: formatMessage({
                                                    id: getTranslationKey('folder.create.title'),
                                                    defaultMessage: 'New folder'
                                                })
                                            }),
                                            /*#__PURE__*/ jsx(MenuItem, {
                                                onSelect: handleFileSelect,
                                                startIcon: /*#__PURE__*/ jsx(Files, {}),
                                                children: formatMessage({
                                                    id: getTranslationKey('import-files'),
                                                    defaultMessage: 'Import files'
                                                })
                                            }),
                                            /*#__PURE__*/ jsx(MenuItem, {
                                                onSelect: ()=>setIsUrlDialogOpen(true),
                                                startIcon: /*#__PURE__*/ jsx(Link, {}),
                                                children: formatMessage({
                                                    id: getTranslationKey('import-from-url'),
                                                    defaultMessage: 'Import from URL'
                                                })
                                            })
                                        ]
                                    }),
                                    subtitle: /*#__PURE__*/ jsxs(Flex, {
                                        justifyContent: "space-between",
                                        alignItems: "center",
                                        gap: 4,
                                        width: "100%",
                                        children: [
                                            /*#__PURE__*/ jsx(Flex, {
                                                gap: 4,
                                                alignItems: "center",
                                                children: "TODO: Filters and search"
                                            }),
                                            /*#__PURE__*/ jsxs(Flex, {
                                                gap: 4,
                                                alignItems: "center",
                                                children: [
                                                    /*#__PURE__*/ jsx(Box, {
                                                        children: "TODO: Sort"
                                                    }),
                                                    /*#__PURE__*/ jsxs(StyledToggleGroup, {
                                                        type: "single",
                                                        value: isGridView ? 'grid' : 'table',
                                                        onValueChange: (value)=>value && setView(value === 'grid' ? viewOptions.GRID : viewOptions.TABLE),
                                                        "aria-label": formatMessage({
                                                            id: getTranslationKey('view.switch.label'),
                                                            defaultMessage: 'View options'
                                                        }),
                                                        children: [
                                                            /*#__PURE__*/ jsxs(StyledToggleItem, {
                                                                value: "table",
                                                                "aria-label": formatMessage({
                                                                    id: getTranslationKey('view.table'),
                                                                    defaultMessage: 'Table view'
                                                                }),
                                                                children: [
                                                                    /*#__PURE__*/ jsx(List, {}),
                                                                    formatMessage({
                                                                        id: getTranslationKey('view.table'),
                                                                        defaultMessage: 'Table view'
                                                                    })
                                                                ]
                                                            }),
                                                            /*#__PURE__*/ jsxs(StyledToggleItem, {
                                                                value: "grid",
                                                                "aria-label": formatMessage({
                                                                    id: getTranslationKey('view.grid'),
                                                                    defaultMessage: 'Grid view'
                                                                }),
                                                                children: [
                                                                    /*#__PURE__*/ jsx(GridFour, {}),
                                                                    formatMessage({
                                                                        id: getTranslationKey('view.grid'),
                                                                        defaultMessage: 'Grid view'
                                                                    })
                                                                ]
                                                            })
                                                        ]
                                                    })
                                                ]
                                            })
                                        ]
                                    })
                                })
                            }),
                            /*#__PURE__*/ jsx(Layouts.Content, {
                                children: /*#__PURE__*/ jsxs(DropZoneWithOverlay, {
                                    children: [
                                        /*#__PURE__*/ jsx(DropFilesMessage, {
                                            uploadDropZoneRef: uploadDropZoneRef,
                                            folderName: title
                                        }),
                                        /*#__PURE__*/ jsx(AssetsView, {
                                            view: view,
                                            folderId: currentFolderId,
                                            onAssetItemClick: openDetails
                                        })
                                    ]
                                })
                            })
                        ]
                    })
                })
            }),
            /*#__PURE__*/ jsx(CreateFolderDialog, {
                open: isCreateFolderDialogOpen,
                folderName: title,
                parentFolderId: currentFolderId,
                onClose: ()=>setIsCreateFolderDialogOpen(false)
            }),
            /*#__PURE__*/ jsx(ImportFromUrlDialog, {
                open: isUrlDialogOpen,
                onClose: ()=>setIsUrlDialogOpen(false),
                onUpload: handleUrlUpload
            }),
            /*#__PURE__*/ jsx(AssetDetailsDrawer, {})
        ]
    });
};

export { AssetsPage };
