// @strapi/upload/dist/admin/pages/SettingsPage/SettingsPage.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var strapiAdmin = require('@strapi/admin/strapi-admin');
var ee = require('@strapi/admin/strapi-admin/ee');
var designSystem = require('@strapi/design-system');
var icons = require('@strapi/icons');
var isEqual = require('lodash/isEqual');
var reactIntl = require('react-intl');
var reactQuery = require('react-query');
var constants = require('../../constants.js');
var useAIMetadataJob = require('../../hooks/useAIMetadataJob.js');
var useSettings = require('../../hooks/useSettings.js');
var useTracking = require('../../hooks/useTracking.js');
require('byte-size');
require('date-fns');
var getTrad = require('../../utils/getTrad.js');
require('qs');
require('../../utils/typeFromMime.js');
require('../../utils/urlYupSchema.js');
var init = require('./init.js');
var reducer = require('./reducer.js');

function _interopNamespaceDefault(e) {
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespaceDefault(React);

// TODO: find a better naming convention for the file that was an index file before
/* -------------------------------------------------------------------------------------------------
 * MetadataAction
 * -----------------------------------------------------------------------------------------------*/ const BetaStatus = (props)=>{
    const { formatMessage } = reactIntl.useIntl();
    return /*#__PURE__*/ jsxRuntime.jsx(designSystem.Status, {
        size: "S",
        variant: "secondary",
        style: {
            textTransform: 'uppercase',
            display: 'flex'
        },
        ...props,
        children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
            tag: "span",
            variant: "pi",
            fontWeight: "bold",
            children: formatMessage({
                id: 'app.components.Status.beta',
                defaultMessage: 'Beta'
            })
        })
    });
};
const MetadataAction = ({ job, imagesWithoutMetadataCount, isConfirmDialogOpen, onConfirmDialogChange, onGenerate })=>{
    const { formatMessage } = reactIntl.useIntl();
    // If there's an active job processing
    if (job?.status === 'processing') {
        return /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
            gap: 2,
            alignItems: "center",
            children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                variant: "pi",
                textColor: "neutral600",
                children: formatMessage({
                    id: getTrad.getTrad('settings.form.aiMetadata.generatingMetadata'),
                    defaultMessage: 'AI is generating your metadata'
                })
            })
        });
    }
    // Only show completed state when all images have metadata
    if (imagesWithoutMetadataCount === 0) {
        return /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
            variant: "pi",
            textColor: "primary600",
            children: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                gap: 2,
                alignItems: "center",
                children: [
                    /*#__PURE__*/ jsxRuntime.jsx(icons.Check, {
                        width: "16px",
                        height: "16px"
                    }),
                    formatMessage({
                        id: getTrad.getTrad('settings.form.aiMetadata.metadataGenerated'),
                        defaultMessage: 'Your metadata has been generated'
                    })
                ]
            })
        });
    }
    return /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Dialog.Root, {
        open: isConfirmDialogOpen,
        onOpenChange: onConfirmDialogChange,
        children: [
            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Dialog.Trigger, {
                children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.TextButton, {
                    disabled: imagesWithoutMetadataCount === 0,
                    children: formatMessage({
                        id: getTrad.getTrad('settings.form.aiMetadata.generateButton'),
                        defaultMessage: 'Generate metadata'
                    })
                })
            }),
            /*#__PURE__*/ jsxRuntime.jsx(strapiAdmin.ConfirmDialog, {
                variant: "success-light",
                onConfirm: onGenerate,
                title: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                    gap: 2,
                    children: [
                        /*#__PURE__*/ jsxRuntime.jsx(BetaStatus, {}),
                        formatMessage({
                            id: getTrad.getTrad('settings.form.aiMetadata.confirmDialog.title'),
                            defaultMessage: 'Generate AI Metadata'
                        })
                    ]
                }),
                children: formatMessage({
                    id: getTrad.getTrad('settings.form.aiMetadata.confirmDialog.message'),
                    defaultMessage: 'This will start a process in the background to generate captions and alternative text for {count, plural, one {# image} other {# images}}. AI can make mistakes, be sure to review the generated content.'
                }, {
                    count: imagesWithoutMetadataCount
                })
            })
        ]
    });
};
/* -------------------------------------------------------------------------------------------------
 * SettingsPage
 * -----------------------------------------------------------------------------------------------*/ const SettingsPage = ()=>{
    const { formatMessage } = reactIntl.useIntl();
    const { toggleNotification } = strapiAdmin.useNotification();
    const { put, post, get } = strapiAdmin.useFetchClient();
    const [isConfirmDialogOpen, setIsConfirmDialogOpen] = React__namespace.useState(false);
    const { trackUsage } = useTracking.useTracking();
    const [{ initialData, modifiedData }, dispatch] = React__namespace.useReducer(reducer.reducer, reducer.initialState, init.init);
    const { data, isLoading, refetch } = useSettings.useSettings();
    const isAIAvailable = ee.useAIAvailability();
    const { data: imageCountResponse, isLoading: isLoadingImagesWithoutMetadataCount } = reactQuery.useQuery([
        'ai-metadata-count'
    ], async ()=>{
        const { data } = await get('/upload/actions/generate-ai-metadata/count');
        return data;
    }, {
        enabled: isAIAvailable && !!data?.aiMetadata,
        retry: false
    });
    const imagesWithoutMetadataCount = imageCountResponse?.imagesWithoutMetadataCount ?? 0;
    React__namespace.useEffect(()=>{
        if (data) {
            dispatch({
                type: 'GET_DATA_SUCCEEDED',
                data
            });
        }
    }, [
        data
    ]);
    const isSaveButtonDisabled = isEqual(initialData, modifiedData);
    const { mutateAsync, isLoading: isSubmitting } = reactQuery.useMutation(async (body)=>{
        const { data } = await put('/upload/settings', body);
        return data;
    }, {
        onSuccess () {
            refetch();
            toggleNotification({
                type: 'success',
                message: formatMessage({
                    id: 'notification.form.success.fields'
                })
            });
        },
        onError (err) {
            toggleNotification({
                type: 'danger',
                message: err.message || formatMessage({
                    id: 'notification.error'
                })
            });
        }
    });
    // Poll for latest active job - notifications are handled inside the hook
    const { data: aiMetadataJob, refetch: refetchAiMetadataJob } = useAIMetadataJob.useAIMetadataJob({
        enabled: isAIAvailable
    });
    const { mutateAsync: startGenerateAIMetadata } = reactQuery.useMutation(async ()=>{
        const { data } = await post('/upload/actions/generate-ai-metadata', {});
        return data;
    }, {
        onSuccess () {
            setIsConfirmDialogOpen(false);
            // Refetch job status to start polling the new job
            refetchAiMetadataJob();
        },
        onError (err) {
            toggleNotification({
                type: 'danger',
                message: err?.message || formatMessage({
                    id: 'notification.error'
                })
            });
            setIsConfirmDialogOpen(false);
        }
    });
    const handleSubmit = async (e)=>{
        e.preventDefault();
        if (isSaveButtonDisabled) {
            return;
        }
        await mutateAsync(modifiedData);
    };
    const handleChange = ({ target: { name, value } })=>{
        dispatch({
            type: 'ON_CHANGE',
            keys: name,
            value
        });
    };
    if (isLoading) {
        return /*#__PURE__*/ jsxRuntime.jsx(strapiAdmin.Page.Loading, {});
    }
    return /*#__PURE__*/ jsxRuntime.jsxs(strapiAdmin.Page.Main, {
        tabIndex: -1,
        children: [
            /*#__PURE__*/ jsxRuntime.jsx(strapiAdmin.Page.Title, {
                children: formatMessage({
                    id: getTrad.getTrad('page.title'),
                    defaultMessage: 'Settings - Media Library'
                })
            }),
            /*#__PURE__*/ jsxRuntime.jsxs("form", {
                onSubmit: handleSubmit,
                children: [
                    /*#__PURE__*/ jsxRuntime.jsx(strapiAdmin.Layouts.Header, {
                        title: formatMessage({
                            id: getTrad.getTrad('settings.header.label'),
                            defaultMessage: 'Media Library'
                        }),
                        primaryAction: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Button, {
                            disabled: isSaveButtonDisabled,
                            loading: isSubmitting,
                            type: "submit",
                            startIcon: /*#__PURE__*/ jsxRuntime.jsx(icons.Check, {}),
                            size: "S",
                            fullWidth: true,
                            children: formatMessage({
                                id: 'global.save',
                                defaultMessage: 'Save'
                            })
                        }),
                        subtitle: formatMessage({
                            id: getTrad.getTrad('settings.sub-header.label'),
                            defaultMessage: 'Configure the settings for the Media Library'
                        })
                    }),
                    /*#__PURE__*/ jsxRuntime.jsx(strapiAdmin.Layouts.Content, {
                        children: /*#__PURE__*/ jsxRuntime.jsx(strapiAdmin.Layouts.Root, {
                            children: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                                direction: "column",
                                alignItems: "stretch",
                                gap: 4,
                                children: [
                                    isAIAvailable && /*#__PURE__*/ jsxRuntime.jsx(designSystem.Box, {
                                        background: "neutral0",
                                        padding: 6,
                                        shadow: "filterShadow",
                                        hasRadius: true,
                                        children: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                                            direction: "column",
                                            alignItems: "stretch",
                                            gap: 1,
                                            children: [
                                                /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Grid.Root, {
                                                    gap: 6,
                                                    children: [
                                                        /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Grid.Item, {
                                                            col: 8,
                                                            xs: 12,
                                                            direction: "column",
                                                            alignItems: "stretch",
                                                            children: [
                                                                /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                                                                    gap: 2,
                                                                    children: [
                                                                        /*#__PURE__*/ jsxRuntime.jsx(designSystem.Box, {
                                                                            color: "alternative700",
                                                                            children: /*#__PURE__*/ jsxRuntime.jsx(icons.Sparkle, {})
                                                                        }),
                                                                        /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                                                                            variant: "delta",
                                                                            tag: "h2",
                                                                            children: formatMessage({
                                                                                id: getTrad.getTrad('settings.form.aiMetadata.label'),
                                                                                defaultMessage: 'Generate AI captions and alt texts automatically on upload!'
                                                                            })
                                                                        })
                                                                    ]
                                                                }),
                                                                /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
                                                                    paddingTop: 1,
                                                                    children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                                                                        variant: "pi",
                                                                        textColor: "neutral600",
                                                                        children: formatMessage({
                                                                            id: getTrad.getTrad('settings.form.aiMetadata.description'),
                                                                            defaultMessage: 'Enable this feature to save time, optimize your SEO and increase accessibility by letting our AI generate captions and alternative texts for you.'
                                                                        })
                                                                    })
                                                                })
                                                            ]
                                                        }),
                                                        /*#__PURE__*/ jsxRuntime.jsx(designSystem.Grid.Item, {
                                                            col: 4,
                                                            xs: 12,
                                                            direction: "column",
                                                            alignItems: "end",
                                                            justifyContent: 'center',
                                                            children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Field.Root, {
                                                                name: "aiMetadata",
                                                                minWidth: "200px",
                                                                children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Toggle, {
                                                                    checked: modifiedData?.aiMetadata,
                                                                    offLabel: formatMessage({
                                                                        id: 'app.components.ToggleCheckbox.disabled-label',
                                                                        defaultMessage: 'Disabled'
                                                                    }),
                                                                    onLabel: formatMessage({
                                                                        id: 'app.components.ToggleCheckbox.enabled-label',
                                                                        defaultMessage: 'Enabled'
                                                                    }),
                                                                    onChange: (e)=>{
                                                                        handleChange({
                                                                            target: {
                                                                                name: 'aiMetadata',
                                                                                value: e.target.checked
                                                                            }
                                                                        });
                                                                    }
                                                                })
                                                            })
                                                        })
                                                    ]
                                                }),
                                                initialData?.aiMetadata && !isLoadingImagesWithoutMetadataCount && Boolean(imageCountResponse?.totalImages) && /*#__PURE__*/ jsxRuntime.jsxs(jsxRuntime.Fragment, {
                                                    children: [
                                                        /*#__PURE__*/ jsxRuntime.jsx(designSystem.Divider, {
                                                            marginTop: 4,
                                                            marginBottom: 4
                                                        }),
                                                        /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                                                            justifyContent: "space-between",
                                                            alignItems: "center",
                                                            gap: 2,
                                                            children: [
                                                                /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                                                                    gap: 2,
                                                                    children: [
                                                                        /*#__PURE__*/ jsxRuntime.jsx(BetaStatus, {
                                                                            size: "XS"
                                                                        }),
                                                                        /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                                                                            variant: "pi",
                                                                            textColor: "neutral500",
                                                                            children: imagesWithoutMetadataCount === 0 ? formatMessage({
                                                                                id: getTrad.getTrad('settings.form.aiMetadata.allAssetsHaveMetadata'),
                                                                                defaultMessage: 'All assets have caption and alt text'
                                                                            }) : formatMessage({
                                                                                id: getTrad.getTrad('settings.form.aiMetadata.imagesWithoutMetadata'),
                                                                                defaultMessage: '{count, plural, one {# image lacks captions or alternative text} other {# images lack captions or alternative text}}'
                                                                            }, {
                                                                                count: imagesWithoutMetadataCount
                                                                            })
                                                                        })
                                                                    ]
                                                                }),
                                                                /*#__PURE__*/ jsxRuntime.jsx(MetadataAction, {
                                                                    job: aiMetadataJob ?? null,
                                                                    imagesWithoutMetadataCount: imagesWithoutMetadataCount,
                                                                    isConfirmDialogOpen: isConfirmDialogOpen,
                                                                    onConfirmDialogChange: setIsConfirmDialogOpen,
                                                                    onGenerate: async ()=>{
                                                                        await startGenerateAIMetadata();
                                                                        trackUsage('didGenerateMetadataRetroactively');
                                                                    }
                                                                })
                                                            ]
                                                        })
                                                    ]
                                                })
                                            ]
                                        })
                                    }),
                                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Box, {
                                        background: "neutral0",
                                        padding: 6,
                                        shadow: "filterShadow",
                                        hasRadius: true,
                                        children: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                                            direction: "column",
                                            alignItems: "stretch",
                                            gap: 4,
                                            children: [
                                                /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
                                                    children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                                                        variant: "delta",
                                                        tag: "h2",
                                                        children: formatMessage({
                                                            id: getTrad.getTrad('settings.blockTitle'),
                                                            defaultMessage: 'Asset management'
                                                        })
                                                    })
                                                }),
                                                /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Grid.Root, {
                                                    gap: 6,
                                                    children: [
                                                        /*#__PURE__*/ jsxRuntime.jsx(designSystem.Grid.Item, {
                                                            col: 6,
                                                            xs: 12,
                                                            direction: "column",
                                                            alignItems: "stretch",
                                                            children: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Field.Root, {
                                                                hint: formatMessage({
                                                                    id: getTrad.getTrad('settings.form.responsiveDimensions.description'),
                                                                    defaultMessage: 'Enabling this option will generate multiple formats (small, medium and large) of the uploaded asset.'
                                                                }),
                                                                name: "responsiveDimensions",
                                                                children: [
                                                                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Field.Label, {
                                                                        children: formatMessage({
                                                                            id: getTrad.getTrad('settings.form.responsiveDimensions.label'),
                                                                            defaultMessage: 'Responsive friendly upload'
                                                                        })
                                                                    }),
                                                                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Toggle, {
                                                                        checked: modifiedData?.responsiveDimensions,
                                                                        offLabel: formatMessage({
                                                                            id: 'app.components.ToggleCheckbox.off-label',
                                                                            defaultMessage: 'Off'
                                                                        }),
                                                                        onLabel: formatMessage({
                                                                            id: 'app.components.ToggleCheckbox.on-label',
                                                                            defaultMessage: 'On'
                                                                        }),
                                                                        onChange: (e)=>{
                                                                            handleChange({
                                                                                target: {
                                                                                    name: 'responsiveDimensions',
                                                                                    value: e.target.checked
                                                                                }
                                                                            });
                                                                        }
                                                                    }),
                                                                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Field.Hint, {})
                                                                ]
                                                            })
                                                        }),
                                                        /*#__PURE__*/ jsxRuntime.jsx(designSystem.Grid.Item, {
                                                            col: 6,
                                                            xs: 12,
                                                            direction: "column",
                                                            alignItems: "stretch",
                                                            children: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Field.Root, {
                                                                hint: formatMessage({
                                                                    id: getTrad.getTrad('settings.form.sizeOptimization.description'),
                                                                    defaultMessage: 'Enabling this option will reduce the image size and slightly reduce its quality.'
                                                                }),
                                                                name: "sizeOptimization",
                                                                children: [
                                                                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Field.Label, {
                                                                        children: formatMessage({
                                                                            id: getTrad.getTrad('settings.form.sizeOptimization.label'),
                                                                            defaultMessage: 'Size optimization'
                                                                        })
                                                                    }),
                                                                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Toggle, {
                                                                        checked: modifiedData?.sizeOptimization,
                                                                        offLabel: formatMessage({
                                                                            id: 'app.components.ToggleCheckbox.off-label',
                                                                            defaultMessage: 'Off'
                                                                        }),
                                                                        onLabel: formatMessage({
                                                                            id: 'app.components.ToggleCheckbox.on-label',
                                                                            defaultMessage: 'On'
                                                                        }),
                                                                        onChange: (e)=>{
                                                                            handleChange({
                                                                                target: {
                                                                                    name: 'sizeOptimization',
                                                                                    value: e.target.checked
                                                                                }
                                                                            });
                                                                        }
                                                                    }),
                                                                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Field.Hint, {})
                                                                ]
                                                            })
                                                        }),
                                                        /*#__PURE__*/ jsxRuntime.jsx(designSystem.Grid.Item, {
                                                            col: 6,
                                                            xs: 12,
                                                            direction: "column",
                                                            alignItems: "stretch",
                                                            children: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Field.Root, {
                                                                hint: formatMessage({
                                                                    id: getTrad.getTrad('settings.form.autoOrientation.description'),
                                                                    defaultMessage: 'Enabling this option will automatically rotate the image according to EXIF orientation tag.'
                                                                }),
                                                                name: "autoOrientation",
                                                                children: [
                                                                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Field.Label, {
                                                                        children: formatMessage({
                                                                            id: getTrad.getTrad('settings.form.autoOrientation.label'),
                                                                            defaultMessage: 'Auto orientation'
                                                                        })
                                                                    }),
                                                                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Toggle, {
                                                                        checked: modifiedData?.autoOrientation,
                                                                        offLabel: formatMessage({
                                                                            id: 'app.components.ToggleCheckbox.off-label',
                                                                            defaultMessage: 'Off'
                                                                        }),
                                                                        onLabel: formatMessage({
                                                                            id: 'app.components.ToggleCheckbox.on-label',
                                                                            defaultMessage: 'On'
                                                                        }),
                                                                        onChange: (e)=>{
                                                                            handleChange({
                                                                                target: {
                                                                                    name: 'autoOrientation',
                                                                                    value: e.target.checked
                                                                                }
                                                                            });
                                                                        }
                                                                    }),
                                                                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Field.Hint, {})
                                                                ]
                                                            })
                                                        })
                                                    ]
                                                })
                                            ]
                                        })
                                    })
                                ]
                            })
                        })
                    })
                ]
            })
        ]
    });
};
const ProtectedSettingsPage = ()=>/*#__PURE__*/ jsxRuntime.jsx(strapiAdmin.Page.Protect, {
        permissions: constants.PERMISSIONS.settings,
        children: /*#__PURE__*/ jsxRuntime.jsx(SettingsPage, {})
    });

exports.ProtectedSettingsPage = ProtectedSettingsPage;
exports.SettingsPage = SettingsPage;
