// @strapi/admin/dist/admin/admin/src/components/SearchInput.mjs
import { jsx } from 'react/jsx-runtime';
import * as React from 'react';
import { Box, IconButton, SearchForm, Searchbar } from '@strapi/design-system';
import { Search } from '@strapi/icons';
import { useIntl } from 'react-intl';
import { useTracking } from '../features/Tracking.mjs';
import { useIsMobile } from '../hooks/useMediaQuery.mjs';
import { useQueryParams } from '../hooks/useQueryParams.mjs';

const SearchInput = ({ disabled, label, placeholder, trackedEvent, trackedEventDetails })=>{
    const inputRef = React.useRef(null);
    const iconButtonRef = React.useRef(null);
    const [{ query }, setQuery] = useQueryParams();
    const [value, setValue] = React.useState(query?._q || '');
    const [isOpen, setIsOpen] = React.useState(!!value);
    const { formatMessage } = useIntl();
    const { trackUsage } = useTracking();
    const isMobile = useIsMobile();
    const handleToggle = ()=>setIsOpen((prev)=>!prev);
    React.useLayoutEffect(()=>{
        if (isOpen && inputRef.current) {
            inputRef.current.focus();
        }
    }, [
        isOpen
    ]);
    const handleClear = ()=>{
        setValue('');
        setQuery({
            _q: ''
        }, 'remove');
    };
    const handleSubmit = (e)=>{
        e.preventDefault();
        // Ensure value is a string
        if (value) {
            if (trackedEvent) {
                trackUsage(trackedEvent, trackedEventDetails);
            }
            setQuery({
                _q: encodeURIComponent(value),
                page: 1
            });
        } else {
            handleToggle();
            setQuery({
                _q: ''
            }, 'remove');
        }
    };
    const renderSearchForm = (opts)=>/*#__PURE__*/ jsx(SearchForm, {
            onSubmit: handleSubmit,
            children: /*#__PURE__*/ jsx(Searchbar, {
                ref: inputRef,
                name: "search",
                onChange: (e)=>setValue(e.target.value),
                value: value,
                clearLabel: formatMessage({
                    id: 'clearLabel',
                    defaultMessage: 'Clear'
                }),
                onClear: handleClear,
                placeholder: isMobile ? undefined : placeholder,
                disabled: disabled,
                onBlur: opts?.enableBlurClose ? (e)=>{
                    if (!e.currentTarget.contains(e.relatedTarget) && e.currentTarget.value === '') {
                        setIsOpen(false);
                    }
                } : undefined,
                children: label
            })
        });
    if (isMobile) {
        // Add wrapper so that the Searchbar takes up the rest of the available space.
        return /*#__PURE__*/ jsx(Box, {
            width: "100%",
            children: renderSearchForm()
        });
    }
    if (isOpen) {
        return renderSearchForm({
            enableBlurClose: true
        });
    }
    return /*#__PURE__*/ jsx(IconButton, {
        ref: iconButtonRef,
        disabled: disabled,
        label: formatMessage({
            id: 'global.search',
            defaultMessage: 'Search'
        }),
        onClick: handleToggle,
        children: /*#__PURE__*/ jsx(Search, {})
    });
};

export { SearchInput };
