// @strapi/upload@5.42.1/dist/admin/future/services/api.js
'use strict';

var strapiAdmin = require('@strapi/admin/strapi-admin');
var uploadProgress = require('../store/uploadProgress.js');
var files = require('../utils/files.js');

/**
 * Stores original File objects for retry functionality.
 *
 * Similar to abortControllers, File objects cannot be stored in Redux state
 * (they are not serializable). This Map allows us to retry cancelled uploads
 * by retrieving the original files using the uploadId.
 */ const uploadedFiles = new Map();
/**
 * Registers files for a specific upload to enable retry.
 */ const registerUploadedFiles = (uploadId, files)=>{
    uploadedFiles.set(uploadId, files);
};
/**
 * Retrieves stored files for an upload.
 */ const getUploadedFiles = (uploadId)=>{
    return uploadedFiles.get(uploadId);
};
/**
 * Manages abort controllers for in-flight uploads.
 *
 * Design decision: Uses a Map to track uploads by their unique uploadId.
 * This approach is necessary because:
 * 1. Redux state cannot store function references (abort controllers)
 * 2. RTK Query's signal is only accessible within the queryFn
 * 3. The upload is triggered in AssetsPage but cancelled from UploadProgressDialog
 *
 * The uploadId ensures we abort the correct upload even if multiple uploads
 * are queued, though the current UI prevents simultaneous uploads.
 */ const abortControllers = new Map();
/**
 * Registers an abort controller for a specific upload.
 * Called internally when an upload starts.
 */ const registerAbortController = (uploadId, controller)=>{
    abortControllers.set(uploadId, controller);
};
/**
 * Removes an abort controller when an upload completes or is aborted.
 */ const unregisterAbortController = (uploadId)=>{
    abortControllers.delete(uploadId);
};
/**
 * Aborts an upload by its uploadId.
 * Called from the UploadProgressDialog when the user clicks cancel or close.
 */ const abortUpload = (uploadId)=>{
    const controller = abortControllers.get(uploadId);
    if (controller) {
        controller.abort();
        unregisterAbortController(uploadId);
    }
};
/**
 * Parses a raw SSE text chunk into event/data pairs.
 *
 * SSE format:
 *   event: <eventName>\n
 *   data: <json>\n
 *   \n
 */ const parseSSEEvents = (chunk)=>{
    const events = [];
    const blocks = chunk.split('\n\n').filter(Boolean);
    for (const block of blocks){
        let event = '';
        let data = '';
        for (const line of block.split('\n')){
            if (line.startsWith('event: ')) {
                event = line.slice(7);
            } else if (line.startsWith('data: ')) {
                data = line.slice(6);
            }
        }
        if (event && data) {
            events.push({
                event,
                data
            });
        }
    }
    return events;
};
/**
 * Makes a streaming upload request to the server.
 *
 * We use fetch directly instead of RTK Query's fetchBaseQuery because:
 * 1. We need access to the raw Response to read the body as a stream
 * 2. RTK Query's baseQuery awaits the full response and parses it as JSON,
 *    which doesn't work for Server-Sent Events (SSE) streaming
 * 3. The stream must be read incrementally via response.body.getReader()
 *    to dispatch progress updates as files upload
 */ const fetchUploadStream = async ({ token, formData, signal })=>{
    const backendURL = window.strapi.backendURL;
    const headers = {};
    if (token) {
        headers.Authorization = `Bearer ${token}`;
    }
    return fetch(`${backendURL}/upload/unstable/stream`, {
        method: 'POST',
        headers,
        body: formData,
        signal
    });
};
/**
 * Makes a streaming upload-from-URLs request to the server.
 * Sends URLs as JSON body instead of FormData.
 */ const fetchUrlUploadStream = async ({ token, urls, folderId, signal })=>{
    const backendURL = window.strapi.backendURL;
    const headers = {
        'Content-Type': 'application/json'
    };
    if (token) {
        headers.Authorization = `Bearer ${token}`;
    }
    return fetch(`${backendURL}/upload/unstable/stream-from-urls`, {
        method: 'POST',
        headers,
        body: JSON.stringify({
            urls,
            folderId
        }),
        signal
    });
};
/**
 * Processes an SSE stream from the upload endpoint.
 * Dispatches Redux actions for each file event and returns the final result.
 *
 * @param options.response - The fetch Response object with SSE body
 * @param options.dispatch - Redux dispatch function
 * @param options.indexMapper - Optional function to map server indices to state indices (for retry)
 * @param options.logPrefix - Optional prefix for console logs
 * @returns The stream result or null if no files completed
 */ const processSSEStream = async ({ response, dispatch, indexMapper = (i)=>i })=>{
    const reader = response.body.getReader();
    const decoder = new TextDecoder();
    let streamResult = null;
    let buffer = '';
    while(true){
        const { done, value } = await reader.read();
        if (done) {
            break;
        }
        buffer += decoder.decode(value, {
            stream: true
        });
        // Process complete SSE events from the buffer
        const lastDoubleNewline = buffer.lastIndexOf('\n\n');
        if (lastDoubleNewline === -1) {
            continue;
        }
        const completePart = buffer.slice(0, lastDoubleNewline + 2);
        buffer = buffer.slice(lastDoubleNewline + 2);
        const events = parseSSEEvents(completePart);
        for (const { event, data } of events){
            const parsed = JSON.parse(data);
            const mappedIndex = indexMapper(parsed.index);
            switch(event){
                case 'file:fetching':
                    {
                        // URL is being fetched server-side - mark as uploading (processing)
                        dispatch(uploadProgress.setFileUploading({
                            name: parsed.url,
                            index: mappedIndex,
                            total: parsed.total,
                            size: 0
                        }));
                        break;
                    }
                case 'file:uploading':
                    {
                        const payload = parsed;
                        dispatch(uploadProgress.setFileUploading({
                            name: payload.name,
                            index: mappedIndex,
                            total: payload.total,
                            size: payload.size
                        }));
                        break;
                    }
                case 'file:complete':
                    {
                        const payload = parsed;
                        dispatch(uploadProgress.setFileComplete({
                            index: mappedIndex,
                            file: payload.file
                        }));
                        break;
                    }
                case 'file:error':
                    {
                        const payload = parsed;
                        dispatch(uploadProgress.setFileError({
                            index: mappedIndex,
                            name: payload.name,
                            message: payload.message
                        }));
                        break;
                    }
                case 'stream:complete':
                    {
                        const payload = parsed;
                        streamResult = {
                            data: payload.data,
                            errors: payload.errors
                        };
                        break;
                    }
                default:
                    console.error(`[SSE Upload] unknown event: ${event}`, parsed);
            }
        }
    }
    return streamResult;
};
/**
 * Performs the actual streaming upload to the server.
 * Shared by uploadFilesStream, retryCancelledFilesStream, and uploadFromUrls.
 *
 * @param options - Upload configuration
 * @returns The upload result or error
 */ const performStreamUpload = async ({ token, formData, abortController, uploadId, dispatch, indexMapper = (i)=>i, onUploadFailed = (message)=>dispatch(uploadProgress.setUploadFailed({
        message
    })) })=>{
    try {
        const response = await fetchUploadStream({
            token,
            formData,
            signal: abortController.signal
        });
        if (!response.ok || !response.body) {
            unregisterAbortController(uploadId);
            let errorMessage = 'Upload request failed';
            try {
                const errorData = await response.json();
                if (errorData.error?.message) {
                    errorMessage = errorData.error.message;
                } else if (errorData.message) {
                    errorMessage = errorData.message;
                }
            } catch  {
                errorMessage = `Upload failed with status ${response.status}`;
            }
            onUploadFailed(errorMessage);
            return {
                error: {
                    name: 'UnknownError',
                    message: errorMessage,
                    status: response.status
                }
            };
        }
        const streamResult = await processSSEStream({
            response,
            dispatch,
            indexMapper
        });
        unregisterAbortController(uploadId);
        if (streamResult && streamResult.data.length > 0) {
            return {
                data: streamResult
            };
        }
        return {
            data: {
                data: [],
                errors: []
            }
        };
    } catch (err) {
        unregisterAbortController(uploadId);
        if (err instanceof DOMException && err.name === 'AbortError') {
            return {
                error: {
                    name: 'UnknownError',
                    message: 'Upload cancelled'
                }
            };
        }
        const errorMessage = err instanceof Error ? err.message : 'Network error occurred';
        onUploadFailed(errorMessage);
        return {
            error: {
                name: 'UnknownError',
                message: errorMessage
            }
        };
    }
};
const uploadApi = strapiAdmin.adminApi.enhanceEndpoints({
    addTagTypes: [
        'Asset',
        'Folder'
    ]
}).injectEndpoints({
    endpoints: (builder)=>({
            /**
       * Stream upload files to the /upload/unstable/stream endpoint.
       * Reads SSE stream for per-file progress updates.
       */ uploadFilesStream: builder.mutation({
                queryFn: async ({ formData, totalFiles }, { dispatch, getState })=>{
                    const token = getState().admin_app?.token;
                    // Extract file names and sizes from FormData
                    const files = formData.getAll('files');
                    const fileInfoJson = formData.get('fileInfo');
                    const fileInfo = JSON.parse(fileInfoJson);
                    const fileNames = fileInfo.map((info)=>info.name);
                    const fileSizes = files.map((file)=>file.size);
                    // Open the progress dialog
                    dispatch(uploadProgress.openUploadProgress({
                        totalFiles,
                        fileNames,
                        fileSizes
                    }));
                    dispatch(uploadProgress.updateProgress(0));
                    // Get the uploadId from state after dispatching
                    const uploadId = getState().uploadProgress.uploadId;
                    // Store original files for retry functionality
                    registerUploadedFiles(uploadId, files);
                    // Create abort controller for this upload
                    const abortController = new AbortController();
                    registerAbortController(uploadId, abortController);
                    return performStreamUpload({
                        token,
                        formData,
                        abortController,
                        uploadId,
                        dispatch
                    });
                },
                invalidatesTags: [
                    {
                        type: 'Asset',
                        id: 'LIST'
                    }
                ]
            }),
            /**
       * Retry uploading cancelled files.
       * Retrieves original File objects and re-uploads only the cancelled ones.
       */ retryCancelledFilesStream: builder.mutation({
                queryFn: async (_, { dispatch, getState })=>{
                    const token = getState().admin_app?.token;
                    const { uploadId, files: stateFiles } = getState().uploadProgress;
                    // Get cancelled files with their original indices
                    const cancelledFiles = stateFiles.filter((f)=>f.status === 'cancelled');
                    if (cancelledFiles.length === 0) {
                        return {
                            error: {
                                name: 'UnknownError',
                                message: 'No cancelled files to retry'
                            }
                        };
                    }
                    // Get the original File objects
                    const originalFiles = getUploadedFiles(uploadId);
                    if (!originalFiles) {
                        return {
                            error: {
                                name: 'UnknownError',
                                message: 'Original files not found'
                            }
                        };
                    }
                    // Build mapping from new index to original index
                    const indexMapping = cancelledFiles.map((f)=>f.index);
                    const filesToRetry = cancelledFiles.map((f)=>originalFiles[f.index]);
                    // Reset cancelled files to pending
                    dispatch(uploadProgress.retryCancelledFiles());
                    // Build FormData for retry
                    const formData = new FormData();
                    const fileInfoArray = filesToRetry.map((file)=>({
                            name: file.name,
                            caption: null,
                            alternativeText: null,
                            folder: null
                        }));
                    filesToRetry.forEach((file)=>{
                        formData.append('files', file);
                    });
                    formData.append('fileInfo', JSON.stringify(fileInfoArray));
                    // Create abort controller for this retry
                    const abortController = new AbortController();
                    registerAbortController(uploadId, abortController);
                    // Custom error handler: mark individual retried files as failed
                    const onUploadFailed = (message)=>{
                        for (const originalIndex of indexMapping){
                            dispatch(uploadProgress.setFileError({
                                index: originalIndex,
                                name: stateFiles[originalIndex].name,
                                message
                            }));
                        }
                    };
                    return performStreamUpload({
                        token,
                        formData,
                        abortController,
                        uploadId,
                        dispatch,
                        indexMapper: (serverIndex)=>indexMapping[serverIndex],
                        onUploadFailed
                    });
                },
                invalidatesTags: [
                    {
                        type: 'Asset',
                        id: 'LIST'
                    }
                ]
            }),
            /**
       * Upload files from URLs.
       * Sends URLs to the server which fetches and uploads them.
       */ uploadFromUrls: builder.mutation({
                queryFn: async ({ urls, folderId }, { dispatch, getState })=>{
                    const token = getState().admin_app?.token;
                    // Extract filenames from URLs for the progress dialog
                    const fileNames = urls.map((url)=>files.getFilenameFromUrl(url));
                    // Open progress dialog with all URLs as pending files
                    dispatch(uploadProgress.openUploadProgress({
                        totalFiles: urls.length,
                        fileNames
                    }));
                    dispatch(uploadProgress.updateProgress(0));
                    // Get the uploadId from state after dispatching
                    const uploadId = getState().uploadProgress.uploadId;
                    // Create abort controller for this upload
                    const abortController = new AbortController();
                    registerAbortController(uploadId, abortController);
                    try {
                        // Send URLs to server for fetching and uploading
                        const response = await fetchUrlUploadStream({
                            token,
                            urls,
                            folderId,
                            signal: abortController.signal
                        });
                        if (!response.ok || !response.body) {
                            unregisterAbortController(uploadId);
                            let errorMessage = 'Upload request failed';
                            try {
                                const errorData = await response.json();
                                if (errorData.error?.message) {
                                    errorMessage = errorData.error.message;
                                } else if (errorData.message) {
                                    errorMessage = errorData.message;
                                }
                            } catch  {
                                errorMessage = `Upload failed with status ${response.status}`;
                            }
                            dispatch(uploadProgress.setUploadFailed({
                                message: errorMessage
                            }));
                            return {
                                error: {
                                    name: 'UnknownError',
                                    message: errorMessage,
                                    status: response.status
                                }
                            };
                        }
                        // Process SSE stream from server
                        const streamResult = await processSSEStream({
                            response,
                            dispatch
                        });
                        unregisterAbortController(uploadId);
                        if (streamResult && streamResult.data.length > 0) {
                            return {
                                data: streamResult
                            };
                        }
                        return {
                            data: {
                                data: [],
                                errors: []
                            }
                        };
                    } catch (err) {
                        unregisterAbortController(uploadId);
                        if (err instanceof DOMException && err.name === 'AbortError') {
                            return {
                                error: {
                                    name: 'UnknownError',
                                    message: 'Upload cancelled'
                                }
                            };
                        }
                        const errorMessage = err instanceof Error ? err.message : 'Network error occurred';
                        dispatch(uploadProgress.setUploadFailed({
                            message: errorMessage
                        }));
                        return {
                            error: {
                                name: 'UnknownError',
                                message: errorMessage
                            }
                        };
                    }
                },
                invalidatesTags: [
                    {
                        type: 'Asset',
                        id: 'LIST'
                    }
                ]
            })
        })
});
const { useUploadFilesStreamMutation, useRetryCancelledFilesStreamMutation, useUploadFromUrlsMutation } = uploadApi;

exports.abortUpload = abortUpload;
exports.uploadApi = uploadApi;
exports.useRetryCancelledFilesStreamMutation = useRetryCancelledFilesStreamMutation;
exports.useUploadFilesStreamMutation = useUploadFilesStreamMutation;
exports.useUploadFromUrlsMutation = useUploadFromUrlsMutation;
