// @strapi/upload@5.42.1/dist/admin/future/services/folders.js
'use strict';

var api = require('./api.js');

const foldersApi = api.uploadApi.injectEndpoints({
    endpoints: (builder)=>({
            getFolders: builder.query({
                query: (params = {})=>{
                    const { parentId } = params;
                    const queryParams = {};
                    if (parentId != null) {
                        queryParams['filters'] = {
                            $and: [
                                {
                                    parent: {
                                        id: parentId
                                    }
                                }
                            ]
                        };
                    } else {
                        queryParams['filters'] = {
                            $and: [
                                {
                                    parent: {
                                        id: {
                                            $null: true
                                        }
                                    }
                                }
                            ]
                        };
                    }
                    return {
                        url: '/upload/folders',
                        method: 'GET',
                        config: {
                            params: queryParams
                        }
                    };
                },
                transformResponse: (response)=>// TODO dont want this cast
                    response.data,
                providesTags: (results)=>{
                    if (results) {
                        return [
                            ...results.map(({ id })=>({
                                    type: 'Folder',
                                    id
                                })),
                            {
                                type: 'Folder',
                                id: 'LIST'
                            }
                        ];
                    }
                    return [
                        {
                            type: 'Folder',
                            id: 'LIST'
                        }
                    ];
                }
            }),
            createFolder: builder.mutation({
                query: (body)=>({
                        url: '/upload/folders',
                        method: 'POST',
                        data: body
                    }),
                transformResponse: (response)=>response.data,
                invalidatesTags: [
                    {
                        type: 'Folder',
                        id: 'LIST'
                    }
                ]
            }),
            getFolder: builder.query({
                query: ({ id })=>({
                        url: `/upload/folders/${id}`,
                        method: 'GET',
                        config: {
                            params: {
                                populate: {
                                    parent: {
                                        populate: {
                                            parent: '*'
                                        }
                                    },
                                    children: {
                                        count: true
                                    },
                                    files: {
                                        count: true
                                    }
                                }
                            }
                        }
                    }),
                transformResponse: (response)=>response.data,
                providesTags: (_result, _error, { id })=>[
                        {
                            type: 'Folder',
                            id
                        }
                    ]
            })
        })
});
const { useCreateFolderMutation, useGetFoldersQuery, useGetFolderQuery } = foldersApi;

exports.useCreateFolderMutation = useCreateFolderMutation;
exports.useGetFolderQuery = useGetFolderQuery;
exports.useGetFoldersQuery = useGetFoldersQuery;
