// @strapi/upload@5.42.1/dist/admin/future/components/Drawer.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var Dialog = require('@radix-ui/react-dialog');
var designSystem = require('@strapi/design-system');
var icons = require('@strapi/icons');
var reactIntl = require('react-intl');
var styledComponents = require('styled-components');

function _interopNamespaceDefault(e) {
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespaceDefault(React);
var Dialog__namespace = /*#__PURE__*/_interopNamespaceDefault(Dialog);

/** Duration of the close animation in ms. Use for timing cleanup (e.g. removing URL params). */ const DRAWER_CLOSE_ANIMATION_MS = 300;
/* -------------------------------------------------------------------------------------------------
 * Animations
 * -----------------------------------------------------------------------------------------------*/ // Direction: up
const slideUpFromBottomIn = styledComponents.keyframes`
  from {
    opacity: 0;
    transform: translateY(100%);
  }
  to {
    opacity: 1;
    transform: translateY(0);
  }
`;
const slideUpFromBottomOut = styledComponents.keyframes`
  from {
    opacity: 1;
    transform: translateY(0);
  }
  to {
    opacity: 0;
    transform: translateY(100%);
  }
`;
// Direction: left
const slideLeftFromRightIn = styledComponents.keyframes`
  from {
    opacity: 0;
    transform: translateX(100%);
  }
  to {
    opacity: 1;
    transform: translateX(0);
  }
`;
const slideLeftFromRightOut = styledComponents.keyframes`
  from {
    opacity: 1;
    transform: translateX(0);
  }
  to {
    opacity: 0;
    transform: translateX(100%);
  }
`;
const DrawerContainer = styledComponents.styled(designSystem.Flex)`
  flex-direction: column;
  position: fixed;
  bottom: 0;
  right: 0;
  padding: ${({ theme })=>theme.spaces[2]};
  max-width: 100%;
  z-index: 1000;
  overflow: hidden;
  width: ${({ width })=>width ?? '400px'};
  max-height: ${({ maxHeight })=>maxHeight ?? '100vh'};

  &:focus {
    outline: none;
  }

  @media (prefers-reduced-motion: no-preference) {
    &[data-state='open'] {
      animation: ${({ $animationDirection })=>$animationDirection === 'up' ? slideUpFromBottomIn : slideLeftFromRightIn}
        ${DRAWER_CLOSE_ANIMATION_MS}ms cubic-bezier(0.32, 0.72, 0, 1) forwards;
    }

    &[data-state='closed'] {
      animation: ${({ $animationDirection })=>$animationDirection === 'up' ? slideUpFromBottomOut : slideLeftFromRightOut}
        ${DRAWER_CLOSE_ANIMATION_MS}ms cubic-bezier(0.32, 0.72, 0, 1) forwards;
      pointer-events: none;
    }
  }
`;
const DrawerContent = styledComponents.styled(designSystem.Box)`
  display: flex;
  flex-direction: column;
  flex: 1;
  min-height: 0;
  width: 100%;
  background-color: ${({ theme })=>theme.colors.neutral0};
  border-radius: ${({ theme })=>theme.borderRadius};
  box-shadow: ${({ theme })=>theme.shadows.popupShadow};
  overflow: hidden;
  border: 1px solid ${({ theme })=>theme.colors.neutral150};
`;
const CollapsibleContent = styledComponents.styled(designSystem.Box)`
  display: grid;
  flex: 1;
  min-height: 0;
  grid-template-rows: ${({ $isVisible })=>$isVisible ? '1fr' : '0fr'};
  transition: grid-template-rows 0.3s ease-in-out;

  > div {
    overflow: ${({ $isVisible })=>$isVisible ? 'auto' : 'hidden'};
    min-height: 0;
  }
`;
const CloseIconButton = styledComponents.styled(designSystem.IconButton)`
  &:hover {
    background: transparent;
  }
`;
const DrawerBody = /*#__PURE__*/ React__namespace.forwardRef(({ animationDirection, children, ...props }, ref)=>/*#__PURE__*/ jsxRuntime.jsx(Dialog__namespace.Content, {
        ref: ref,
        forceMount: true,
        asChild: true,
        onPointerDownOutside: (e)=>e.preventDefault(),
        onInteractOutside: (e)=>e.preventDefault(),
        "data-animation-direction": animationDirection,
        children: /*#__PURE__*/ jsxRuntime.jsx(DrawerContainer, {
            $animationDirection: animationDirection,
            ...props,
            children: /*#__PURE__*/ jsxRuntime.jsx(DrawerContent, {
                children: children
            })
        })
    }));
DrawerBody.displayName = 'DrawerBody';
const DrawerRoot = ({ isVisible, onClose, children })=>/*#__PURE__*/ jsxRuntime.jsx(Dialog__namespace.Root, {
        open: isVisible,
        onOpenChange: (nextVisible)=>!nextVisible && onClose?.(),
        modal: false,
        children: /*#__PURE__*/ jsxRuntime.jsx(Dialog__namespace.Portal, {
            children: children
        })
    });
const DrawerScrollableContent = ({ children, isContentExpanded = true })=>/*#__PURE__*/ jsxRuntime.jsx(CollapsibleContent, {
        $isVisible: isContentExpanded,
        "data-collapsed": !isContentExpanded,
        children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.ScrollArea, {
            children: children
        })
    });
const DrawerCloseButton = ({ onClose, label, children })=>{
    const { formatMessage } = reactIntl.useIntl();
    const labelMessage = label ?? formatMessage({
        id: 'global.close',
        defaultMessage: 'Close'
    });
    return /*#__PURE__*/ jsxRuntime.jsx(CloseIconButton, {
        onClick: onClose,
        label: labelMessage,
        variant: "ghost",
        children: children ?? /*#__PURE__*/ jsxRuntime.jsx(icons.Cross, {})
    });
};
const Drawer = {
    Root: DrawerRoot,
    Body: DrawerBody,
    ScrollableContent: DrawerScrollableContent,
    CloseButton: DrawerCloseButton,
    Title: Dialog__namespace.Title,
    Description: Dialog__namespace.Description
};

exports.DRAWER_CLOSE_ANIMATION_MS = DRAWER_CLOSE_ANIMATION_MS;
exports.Drawer = Drawer;
