// @strapi/upload/dist/server/controllers/admin-upload.js
'use strict';

var os = require('os');
var path = require('path');
var fse = require('fs-extra');
var _ = require('lodash');
var utils = require('@strapi/utils');
var index = require('../utils/index.js');
var constants = require('../constants.js');
var upload = require('./validation/admin/upload.js');
var findEntityAndCheckPermissions = require('./utils/find-entity-and-check-permissions.js');
var mimeValidation = require('../utils/mime-validation.js');

var adminUpload = {
    async bulkUpdateFileInfo (ctx) {
        const { state: { userAbility, user }, request: { body } } = ctx;
        const { updates } = await upload.validateBulkUpdateBody(body);
        const uploadService = index.getService('upload');
        const results = await utils.async.map(updates, async ({ id, fileInfo })=>{
            const { pm } = await findEntityAndCheckPermissions.findEntityAndCheckPermissions(userAbility, constants.ACTIONS.update, constants.FILE_MODEL_UID, id);
            const updated = await uploadService.updateFileInfo(id, fileInfo, {
                user
            });
            return pm.sanitizeOutput(updated, {
                action: constants.ACTIONS.read
            });
        });
        ctx.body = results;
    },
    async updateFileInfo (ctx) {
        const { state: { userAbility, user }, query: { id }, request: { body } } = ctx;
        if (typeof id !== 'string') {
            throw new utils.errors.ValidationError('File id is required');
        }
        const uploadService = index.getService('upload');
        const { pm } = await findEntityAndCheckPermissions.findEntityAndCheckPermissions(userAbility, constants.ACTIONS.update, constants.FILE_MODEL_UID, id);
        const data = await upload.validateUploadBody(body);
        const file = await uploadService.updateFileInfo(id, data.fileInfo, {
            user
        });
        ctx.body = await pm.sanitizeOutput(file, {
            action: constants.ACTIONS.read
        });
    },
    async replaceFile (ctx) {
        const { state: { userAbility, user }, query: { id }, request: { body, files: { files } = {} } } = ctx;
        if (typeof id !== 'string') {
            throw new utils.errors.ValidationError('File id is required');
        }
        const uploadService = index.getService('upload');
        const { pm } = await findEntityAndCheckPermissions.findEntityAndCheckPermissions(userAbility, constants.ACTIONS.update, constants.FILE_MODEL_UID, id);
        if (Array.isArray(files)) {
            throw new utils.errors.ApplicationError('Cannot replace a file with multiple ones');
        }
        const { validFiles, filteredBody, errors: validationErrors } = await mimeValidation.prepareUploadRequest(files, body, strapi);
        if (validFiles.length === 0) {
            throw new utils.errors.ValidationError(validationErrors[0].message);
        }
        const data = await upload.validateUploadBody(filteredBody);
        const replacedFile = await uploadService.replace(id, {
            data,
            file: validFiles[0]
        }, {
            user
        });
        // Sign file urls for private providers
        const signedFile = await index.getService('file').signFileUrls(replacedFile);
        ctx.body = await pm.sanitizeOutput(signedFile, {
            action: constants.ACTIONS.read
        });
    },
    async uploadFiles (ctx) {
        const { state: { userAbility, user }, request: { body, files: { files } = {} } } = ctx;
        const uploadService = index.getService('upload');
        const pm = strapi.service('admin::permission').createPermissionsManager({
            ability: userAbility,
            action: constants.ACTIONS.create,
            model: constants.FILE_MODEL_UID
        });
        if (!pm.isAllowed) {
            return ctx.forbidden();
        }
        const { validFiles, filteredBody, errors: validationErrors } = await mimeValidation.prepareUploadRequest(files, body, strapi);
        if (validFiles.length === 0) {
            throw new utils.errors.ValidationError(validationErrors[0].message);
        }
        const isMultipleFiles = validFiles.length > 1;
        const data = await upload.validateUploadBody(filteredBody, isMultipleFiles);
        let filesArray = validFiles;
        if (data.fileInfo && Array.isArray(data.fileInfo) && filesArray.length === data.fileInfo.length) {
            // Reorder filesArray to match data.fileInfo order
            const alignedFilesArray = data.fileInfo.map((info)=>{
                return filesArray.find((file)=>file.originalFilename === info.name);
            }).filter(Boolean);
            filesArray = alignedFilesArray;
        }
        // Upload files first to get thumbnails
        const uploadedFiles = await uploadService.upload({
            data,
            files: filesArray
        }, {
            user
        });
        if (uploadedFiles.some((file)=>file.mime?.startsWith('image/'))) {
            await index.getService('metrics').trackUsage('didUploadImage');
        }
        const aiMetadataService = index.getService('aiMetadata');
        // AFTER upload - generate AI metadata for images
        if (await aiMetadataService.isEnabled()) {
            try {
                const metadataResults = await aiMetadataService.processFiles(uploadedFiles);
                // Update the uploaded files with AI metadata
                await aiMetadataService.updateFilesWithAIMetadata(uploadedFiles, metadataResults, user);
            } catch (error) {
                strapi.log.warn('AI metadata generation failed, proceeding without AI enhancements', {
                    error: error instanceof Error ? error.message : String(error)
                });
            }
        }
        // Sign file urls for private providers
        const signedFiles = await utils.async.map(uploadedFiles, index.getService('file').signFileUrls);
        ctx.body = await pm.sanitizeOutput(signedFiles, {
            action: constants.ACTIONS.read
        });
        ctx.status = 201;
    },
    /**
   * @experimental
   * Stream upload files with SSE streaming for per-file progress
   *
   * Streams Server-Sent Events as each file is validated and uploaded:
   * - file:uploading — when processing starts for a file
   * - file:complete  — when a file is successfully uploaded
   * - file:error     — when a file fails validation or upload
   * - stream:complete — final summary with all results
   *
   */ async unstable_uploadFilesStream (ctx) {
        const { state: { userAbility, user }, request: { body, files: { files } = {} } } = ctx;
        const uploadService = index.getService('upload');
        const pm = strapi.service('admin::permission').createPermissionsManager({
            ability: userAbility,
            action: constants.ACTIONS.create,
            model: constants.FILE_MODEL_UID
        });
        if (!pm.isAllowed) {
            return ctx.forbidden();
        }
        if (_.isEmpty(files) || !Array.isArray(files) && files.size === 0) {
            throw new utils.errors.ApplicationError('Files are empty');
        }
        // Take manual control of the response for SSE streaming
        ctx.respond = false;
        const res = ctx.res;
        res.writeHead(200, {
            'Content-Type': 'text/event-stream',
            'Cache-Control': 'no-cache',
            Connection: 'keep-alive'
        });
        const writeSSE = (event, data)=>{
            res.write(`event: ${event}\ndata: ${JSON.stringify(data)}\n\n`);
        };
        // Normalize files to an array
        const filesArray = Array.isArray(files) ? files : [
            files
        ];
        const total = filesArray.length;
        // Parse fileInfo from body
        // Multipart forms send fileInfo as either:
        //   - An array of JSON strings (one per file)
        //   - A single JSON string (one file, or a JSON-encoded array)
        let parsedFileInfo = [];
        if (body?.fileInfo) {
            const raw = body.fileInfo;
            if (Array.isArray(raw)) {
                parsedFileInfo = raw.map((fi)=>typeof fi === 'string' ? JSON.parse(fi) : fi);
            } else if (typeof raw === 'string') {
                const parsed = JSON.parse(raw);
                // Handle case where a single string contains a JSON array
                parsedFileInfo = Array.isArray(parsed) ? parsed : [
                    parsed
                ];
            } else {
                parsedFileInfo = [
                    raw
                ];
            }
        }
        const uploadErrors = [];
        const successfulFiles = [];
        // Process each file sequentially with inline validation
        for(let i = 0; i < filesArray.length; i += 1){
            const file = filesArray[i];
            const fileName = file.originalFilename || 'unknown';
            const fileInfo = parsedFileInfo[i] || {
                name: fileName,
                caption: null,
                alternativeText: null,
                folder: null
            };
            writeSSE('file:uploading', {
                name: fileName,
                index: i,
                total,
                size: file.size || 0
            });
            try {
                // Validate this single file using security checks
                const { validFiles, errors: validationErrors } = await mimeValidation.prepareUploadRequest(file, {
                    fileInfo: JSON.stringify(fileInfo)
                }, strapi);
                if (validFiles.length === 0) {
                    const errorMessage = validationErrors[0]?.message || 'Validation failed';
                    uploadErrors.push({
                        name: fileName,
                        message: errorMessage
                    });
                    writeSSE('file:error', {
                        name: fileName,
                        index: i,
                        message: errorMessage
                    });
                } else {
                    // Validate using the already-parsed single fileInfo object directly
                    const data = await upload.validateUploadBody({
                        fileInfo
                    }, false);
                    const [uploadedFile] = await uploadService.upload({
                        data,
                        files: [
                            validFiles[0]
                        ]
                    }, {
                        user
                    });
                    // Sign file url
                    const signedFile = await index.getService('file').signFileUrls(uploadedFile);
                    successfulFiles.push(signedFile);
                    writeSSE('file:complete', {
                        name: fileName,
                        index: i,
                        file: signedFile
                    });
                }
            } catch (error) {
                const errorMessage = error instanceof Error ? error.message : String(error);
                uploadErrors.push({
                    name: fileName,
                    message: errorMessage
                });
                writeSSE('file:error', {
                    name: fileName,
                    index: i,
                    message: errorMessage
                });
            }
        }
        // Track image upload metric once if any images were uploaded
        if (successfulFiles.some((file)=>file.mime?.startsWith('image/'))) {
            await index.getService('metrics').trackUsage('didUploadImage');
        }
        // Send final stream summary
        writeSSE('stream:complete', {
            data: await pm.sanitizeOutput(successfulFiles, {
                action: constants.ACTIONS.read
            }),
            errors: uploadErrors
        });
        res.end();
    },
    /**
   * @experimental
   * Upload files from URLs with SSE streaming for per-file progress
   *
   * Accepts JSON body with URLs and fetches them server-side.
   * Streams Server-Sent Events as each URL is fetched and uploaded:
   * - file:fetching  — when starting to fetch a URL
   * - file:uploading — when upload starts for a fetched file
   * - file:complete  — when a file is successfully uploaded
   * - file:error     — when a URL fetch or upload fails
   * - stream:complete — final summary with all results
   */ async unstable_uploadFromUrls (ctx) {
        const { state: { userAbility, user }, request: { body } } = ctx;
        const uploadService = index.getService('upload');
        const fileService = index.getService('file');
        const pm = strapi.service('admin::permission').createPermissionsManager({
            ability: userAbility,
            action: constants.ACTIONS.create,
            model: constants.FILE_MODEL_UID
        });
        if (!pm.isAllowed) {
            return ctx.forbidden();
        }
        // Parse and validate request body
        const { urls, folderId } = body;
        if (!urls || !Array.isArray(urls) || urls.length === 0) {
            throw new utils.errors.ApplicationError('URLs are required');
        }
        if (urls.length > 20) {
            throw new utils.errors.ApplicationError('Maximum 20 URLs allowed per request');
        }
        // Take manual control of the response for SSE streaming
        ctx.respond = false;
        const res = ctx.res;
        res.writeHead(200, {
            'Content-Type': 'text/event-stream',
            'Cache-Control': 'no-cache',
            Connection: 'keep-alive'
        });
        const writeSSE = (event, data)=>{
            res.write(`event: ${event}\ndata: ${JSON.stringify(data)}\n\n`);
        };
        const total = urls.length;
        const uploadErrors = [];
        const successfulFiles = [];
        // Create temp directory for fetched files
        const tmpWorkingDirectory = await fse.mkdtemp(path.join(os.tmpdir(), 'strapi-url-upload-'));
        const { sizeLimit } = strapi.config.get('plugin::upload');
        try {
            // Process each URL sequentially
            for(let i = 0; i < urls.length; i += 1){
                const url = urls[i];
                writeSSE('file:fetching', {
                    url,
                    index: i,
                    total
                });
                try {
                    // Fetch URL to temp file
                    const { file } = await fileService.fetchUrlToInputFile(url, tmpWorkingDirectory, sizeLimit);
                    const fileName = file.originalFilename;
                    writeSSE('file:uploading', {
                        name: fileName,
                        index: i,
                        total,
                        size: file.size
                    });
                    // Validate using security checks
                    const fileInfo = {
                        name: fileName,
                        caption: null,
                        alternativeText: null,
                        folder: folderId ?? null
                    };
                    const { validFiles, errors: validationErrors } = await mimeValidation.prepareUploadRequest(file, {
                        fileInfo: JSON.stringify(fileInfo)
                    }, strapi);
                    if (validFiles.length === 0) {
                        const errorMessage = validationErrors[0]?.message || 'Validation failed';
                        uploadErrors.push({
                            name: fileName,
                            message: errorMessage
                        });
                        writeSSE('file:error', {
                            name: fileName,
                            url,
                            index: i,
                            message: errorMessage
                        });
                    } else {
                        // Upload the file
                        const data = await upload.validateUploadBody({
                            fileInfo
                        }, false);
                        const [uploadedFile] = await uploadService.upload({
                            data,
                            files: [
                                validFiles[0]
                            ]
                        }, {
                            user
                        });
                        // Sign file url
                        const signedFile = await fileService.signFileUrls(uploadedFile);
                        successfulFiles.push(signedFile);
                        writeSSE('file:complete', {
                            name: fileName,
                            index: i,
                            file: signedFile
                        });
                    }
                } catch (error) {
                    const errorMessage = error instanceof Error ? error.message : String(error);
                    uploadErrors.push({
                        name: url,
                        message: errorMessage
                    });
                    writeSSE('file:error', {
                        url,
                        index: i,
                        message: errorMessage
                    });
                }
            }
            // Track image upload metric once if any images were uploaded
            if (successfulFiles.some((file)=>file.mime?.startsWith('image/'))) {
                await index.getService('metrics').trackUsage('didUploadImage');
            }
            // Send final stream summary
            writeSSE('stream:complete', {
                data: await pm.sanitizeOutput(successfulFiles, {
                    action: constants.ACTIONS.read
                }),
                errors: uploadErrors
            });
        } finally{
            // Clean up temp directory
            await fse.remove(tmpWorkingDirectory);
        }
        res.end();
    },
    // TODO: split into multiple endpoints
    async upload (ctx) {
        const { query: { id }, request: { files: { files } = {} } } = ctx;
        if (_.isEmpty(files) || !Array.isArray(files) && files.size === 0) {
            if (id) {
                return this.updateFileInfo(ctx);
            }
            throw new utils.errors.ApplicationError('Files are empty');
        }
        await (id ? this.replaceFile : this.uploadFiles)(ctx);
    }
};

module.exports = adminUpload;
