// @strapi/review-workflows@5.42.1/dist/server/services/stage-permissions.js
'use strict';

var fp = require('lodash/fp');
var utils = require('@strapi/utils');
var index = require('../utils/index.js');
var workflows = require('../constants/workflows.js');

const { ApplicationError } = utils.errors;
const validActions = [
    workflows.STAGE_TRANSITION_UID
];
var stagePermissions = (({ strapi })=>{
    const roleService = index.getAdminService('role');
    const permissionService = index.getAdminService('permission');
    const getUserRoles = ()=>{
        const requestState = strapi.requestContext.get()?.state;
        if (!requestState) {
            return null;
        }
        return requestState.user?.roles ?? null;
    };
    return {
        async register ({ roleId, action, fromStage }) {
            if (!validActions.includes(action)) {
                throw new ApplicationError(`Invalid action ${action}`);
            }
            const permissions = await roleService.addPermissions(roleId, [
                {
                    action,
                    actionParameters: {
                        from: fromStage
                    }
                }
            ]);
            // TODO: Filter response
            return permissions;
        },
        async registerMany (permissions) {
            return utils.async.map(permissions, this.register);
        },
        async registerTo ({ roleId, action, toStage }) {
            if (!validActions.includes(action)) {
                throw new ApplicationError(`Invalid action ${action}`);
            }
            const permissions = await roleService.addPermissions(roleId, [
                {
                    action,
                    actionParameters: {
                        to: toStage
                    }
                }
            ]);
            return permissions;
        },
        async registerManyTo (permissions) {
            return utils.async.map(permissions, this.registerTo);
        },
        async unregister (permissions) {
            const permissionIds = permissions.map(fp.prop('id'));
            await permissionService.deleteByIds(permissionIds);
        },
        can (action, fromStage) {
            const requestState = strapi.requestContext.get()?.state;
            if (!requestState) {
                return false;
            }
            const userRoles = getUserRoles();
            if (!userRoles) {
                return false;
            }
            if (userRoles.some((role)=>role.code === 'strapi-super-admin')) {
                return true;
            }
            return requestState.userAbility.can({
                name: action,
                params: {
                    from: fromStage
                }
            });
        },
        async canTransitionToStage (toStageId) {
            const userRoles = getUserRoles();
            if (!userRoles) {
                return false;
            }
            if (userRoles.some((role)=>role.code === 'strapi-super-admin')) {
                return true;
            }
            const targetStage = await strapi.db.query(workflows.STAGE_MODEL_UID).findOne({
                where: {
                    id: toStageId
                },
                populate: {
                    permissions: {
                        populate: [
                            'role'
                        ]
                    }
                }
            });
            if (!targetStage) {
                return false;
            }
            return this.canTransitionToStageWithPermissions(targetStage);
        },
        /**
     * Check if the current user can transition to a stage using pre-loaded permissions.
     * The stage must already have its permissions populated with roles.
     */ canTransitionToStageWithPermissions (stage) {
            const userRoles = getUserRoles();
            if (!userRoles) {
                return false;
            }
            if (userRoles.some((role)=>role.code === 'strapi-super-admin')) {
                return true;
            }
            const toPermissions = (stage.permissions || []).filter((p)=>p.actionParameters?.to);
            const userRoleIds = new Set(userRoles.map((role)=>role.id));
            return toPermissions.some((p)=>{
                const roleId = typeof p.role === 'object' ? p.role?.id : p.role;
                return roleId !== undefined && userRoleIds.has(roleId);
            });
        }
    };
});

module.exports = stagePermissions;
