// @strapi/upload@5.42.1/dist/admin/future/pages/Assets/components/DropZone/UploadDropZone.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var designSystem = require('@strapi/design-system');
var icons = require('@strapi/icons');
var reactIntl = require('react-intl');
var styledComponents = require('styled-components');
var translations = require('../../../../utils/translations.js');
var UploadDropZoneContext = require('./UploadDropZoneContext.js');

function _interopNamespaceDefault(e) {
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespaceDefault(React);

/* -------------------------------------------------------------------------------------------------
 * DropZoneOverlay
 * -----------------------------------------------------------------------------------------------*/ const setOpacity = (hex, alpha)=>`${hex}${Math.floor(alpha * 255).toString(16).padStart(2, '0')}`;
const DropZoneOverlay = styledComponents.styled(designSystem.Box)`
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background: ${({ theme })=>setOpacity(theme.colors.primary200, 0.3)};
  border: 1px solid ${({ theme })=>theme.colors.primary700};
  border-radius: ${({ theme })=>theme.borderRadius};
  z-index: 1;
  pointer-events: none;
`;
const DropZoneWithOverlay = ({ children })=>{
    const { isDragging } = UploadDropZoneContext.useUploadDropZone();
    return /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Box, {
        position: "relative",
        children: [
            isDragging && /*#__PURE__*/ jsxRuntime.jsx(DropZoneOverlay, {}),
            children
        ]
    });
};
/* -------------------------------------------------------------------------------------------------
 * DropZoneMessage
 * -----------------------------------------------------------------------------------------------*/ const DropFilesMessageImpl = styledComponents.styled(designSystem.Box)`
  position: fixed;
  bottom: ${({ theme })=>theme.spaces[8]};
  left: 50%;
  transform: translateX(calc(-50% + ${({ $leftContentWidth })=>$leftContentWidth / 2}px));
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  gap: ${({ theme })=>theme.spaces[2]};
  background: ${({ theme })=>theme.colors.primary600};
  padding: ${({ theme })=>theme.spaces[4]} ${({ theme })=>theme.spaces[6]};
  border-radius: ${({ theme })=>theme.borderRadius};
  z-index: 2;
`;
const DropFilesMessage = ({ uploadDropZoneRef, folderName })=>{
    const { formatMessage } = reactIntl.useIntl();
    const { isDragging } = UploadDropZoneContext.useUploadDropZone();
    // Dropzone message position (relative to main content)
    const [leftContentWidth, setLeftContentWidth] = React__namespace.useState(0);
    // Calculate the left content width to position the dropzone message correctly
    React__namespace.useEffect(()=>{
        if (!uploadDropZoneRef?.current) return;
        const updateRect = ()=>{
            const rect = uploadDropZoneRef.current?.getBoundingClientRect();
            if (rect) {
                setLeftContentWidth((prev)=>prev !== rect.left ? rect.left : prev);
            }
        };
        updateRect();
        const resizeObserver = new ResizeObserver(updateRect);
        resizeObserver.observe(uploadDropZoneRef.current);
        return ()=>resizeObserver.disconnect();
    }, [
        uploadDropZoneRef
    ]);
    if (!isDragging) return null;
    return /*#__PURE__*/ jsxRuntime.jsxs(DropFilesMessageImpl, {
        $leftContentWidth: leftContentWidth,
        children: [
            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                textColor: "neutral0",
                children: formatMessage({
                    id: translations.getTranslationKey('dropzone.upload.message'),
                    defaultMessage: 'Drop here to upload to'
                })
            }),
            /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                gap: 2,
                alignItems: "center",
                children: [
                    /*#__PURE__*/ jsxRuntime.jsx(icons.Folder, {
                        width: 20,
                        height: 20,
                        fill: "neutral0"
                    }),
                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                        textColor: "neutral0",
                        fontWeight: "semiBold",
                        children: folderName
                    })
                ]
            })
        ]
    });
};

exports.DropFilesMessage = DropFilesMessage;
exports.DropZoneWithOverlay = DropZoneWithOverlay;
