// @strapi/upload/dist/admin/components/EditAssetDialog/PreviewBox/FocalPointActions.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var designSystem = require('@strapi/design-system');
var icons = require('@strapi/icons');
var reactIntl = require('react-intl');
require('byte-size');
require('date-fns');
var getTrad = require('../../../utils/getTrad.js');
require('qs');
require('../../../utils/typeFromMime.js');
require('../../../utils/urlYupSchema.js');
var PreviewComponents = require('./PreviewComponents.js');

const FocalPointActions = ({ onCancel, onValidate, onReset })=>{
    const { formatMessage } = reactIntl.useIntl();
    return /*#__PURE__*/ jsxRuntime.jsx(designSystem.FocusTrap, {
        onEscape: onCancel,
        children: /*#__PURE__*/ jsxRuntime.jsx(PreviewComponents.FocalPointActionRow, {
            justifyContent: "flex-end",
            paddingLeft: 3,
            paddingRight: 3,
            children: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                gap: 1,
                children: [
                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.IconButton, {
                        label: formatMessage({
                            id: getTrad.getTrad('control-card.stop-focal-point'),
                            defaultMessage: 'Cancel focal point selection'
                        }),
                        onClick: onCancel,
                        children: /*#__PURE__*/ jsxRuntime.jsx(icons.Cross, {})
                    }),
                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.IconButton, {
                        label: formatMessage({
                            id: getTrad.getTrad('control-card.reset-focal-point'),
                            defaultMessage: 'Reset to center'
                        }),
                        onClick: onReset,
                        children: /*#__PURE__*/ jsxRuntime.jsx(icons.ArrowsCounterClockwise, {})
                    }),
                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.IconButton, {
                        label: formatMessage({
                            id: getTrad.getTrad('control-card.save-focal-point'),
                            defaultMessage: 'Save focal point'
                        }),
                        onClick: onValidate,
                        children: /*#__PURE__*/ jsxRuntime.jsx(icons.Check, {})
                    })
                ]
            })
        })
    });
};

exports.FocalPointActions = FocalPointActions;
