// @strapi/upload/dist/admin/components/EditAssetDialog/PreviewBox/FocalPointActions.mjs
import { jsx, jsxs } from 'react/jsx-runtime';
import { FocusTrap, Flex, IconButton } from '@strapi/design-system';
import { Cross, ArrowsCounterClockwise, Check } from '@strapi/icons';
import { useIntl } from 'react-intl';
import 'byte-size';
import 'date-fns';
import { getTrad } from '../../../utils/getTrad.mjs';
import 'qs';
import '../../../utils/typeFromMime.mjs';
import '../../../utils/urlYupSchema.mjs';
import { FocalPointActionRow } from './PreviewComponents.mjs';

const FocalPointActions = ({ onCancel, onValidate, onReset })=>{
    const { formatMessage } = useIntl();
    return /*#__PURE__*/ jsx(FocusTrap, {
        onEscape: onCancel,
        children: /*#__PURE__*/ jsx(FocalPointActionRow, {
            justifyContent: "flex-end",
            paddingLeft: 3,
            paddingRight: 3,
            children: /*#__PURE__*/ jsxs(Flex, {
                gap: 1,
                children: [
                    /*#__PURE__*/ jsx(IconButton, {
                        label: formatMessage({
                            id: getTrad('control-card.stop-focal-point'),
                            defaultMessage: 'Cancel focal point selection'
                        }),
                        onClick: onCancel,
                        children: /*#__PURE__*/ jsx(Cross, {})
                    }),
                    /*#__PURE__*/ jsx(IconButton, {
                        label: formatMessage({
                            id: getTrad('control-card.reset-focal-point'),
                            defaultMessage: 'Reset to center'
                        }),
                        onClick: onReset,
                        children: /*#__PURE__*/ jsx(ArrowsCounterClockwise, {})
                    }),
                    /*#__PURE__*/ jsx(IconButton, {
                        label: formatMessage({
                            id: getTrad('control-card.save-focal-point'),
                            defaultMessage: 'Save focal point'
                        }),
                        onClick: onValidate,
                        children: /*#__PURE__*/ jsx(Check, {})
                    })
                ]
            })
        })
    });
};

export { FocalPointActions };
