// @strapi/content-manager@5.42.1/dist/admin/pages/EditView/components/FormInputs/DynamicZone/ComponentPicker.mjs
import { jsxs, jsx } from 'react/jsx-runtime';
import { useIsMobile } from '@strapi/admin/strapi-admin';
import { Flex, Typography, Box, Accordion } from '@strapi/design-system';
import { useIntl } from 'react-intl';
import { getTranslation } from '../../../../../utils/translations.mjs';
import { ComponentCategory } from './ComponentCategory.mjs';

const ComponentPicker = ({ dynamicComponentsByCategory = {}, isOpen, onClickAddComponent })=>{
    const { formatMessage } = useIntl();
    const isMobile = useIsMobile();
    const handleAddComponentToDz = (componentUid)=>()=>{
            onClickAddComponent(componentUid);
        };
    if (!isOpen) {
        return null;
    }
    return /*#__PURE__*/ jsxs(Flex, {
        paddingTop: {
            initial: 0,
            medium: 6
        },
        paddingBottom: {
            initial: 0,
            medium: 6
        },
        paddingLeft: {
            initial: 0,
            medium: 5
        },
        paddingRight: {
            initial: 0,
            medium: 5
        },
        background: isMobile ? 'transparent' : 'neutral0',
        shadow: isMobile ? 'none' : 'tableShadow',
        borderColor: isMobile ? 'transparent' : 'neutral150',
        hasRadius: !isMobile,
        gap: {
            initial: 4,
            medium: 2
        },
        direction: "column",
        alignItems: "stretch",
        children: [
            /*#__PURE__*/ jsx(Flex, {
                justifyContent: "center",
                children: /*#__PURE__*/ jsx(Typography, {
                    fontWeight: "bold",
                    textColor: "neutral600",
                    fontSize: 1,
                    children: formatMessage({
                        id: getTranslation('components.DynamicZone.ComponentPicker-label'),
                        defaultMessage: 'Pick one component'
                    })
                })
            }),
            /*#__PURE__*/ jsx(Box, {
                children: /*#__PURE__*/ jsx(Accordion.Root, {
                    defaultValue: Object.keys(dynamicComponentsByCategory)[0],
                    children: Object.entries(dynamicComponentsByCategory).map(([category, components], index)=>/*#__PURE__*/ jsx(ComponentCategory, {
                            category: category,
                            components: components,
                            onAddComponent: handleAddComponentToDz,
                            variant: index % 2 === 1 ? 'primary' : 'secondary'
                        }, category))
                })
            })
        ]
    });
};

export { ComponentPicker };
