// @strapi/content-manager@5.42.1/dist/admin/pages/EditView/components/FormInputs/DynamicZone/DynamicComponent.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var strapiAdmin = require('@strapi/admin/strapi-admin');
var designSystem = require('@strapi/design-system');
var Icons = require('@strapi/icons');
var reactDndHtml5Backend = require('react-dnd-html5-backend');
var reactIntl = require('react-intl');
var styledComponents = require('styled-components');
var ComponentIcon = require('../../../../../components/ComponentIcon.js');
var dragAndDrop = require('../../../../../constants/dragAndDrop.js');
var useDocumentContext = require('../../../../../hooks/useDocumentContext.js');
var useDocumentLayout = require('../../../../../hooks/useDocumentLayout.js');
var useDragAndDrop = require('../../../../../hooks/useDragAndDrop.js');
var objects = require('../../../../../utils/objects.js');
var translations = require('../../../../../utils/translations.js');
var FormLayout = require('../../FormLayout.js');
var InputRenderer = require('../../InputRenderer.js');

function _interopNamespaceDefault(e) {
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespaceDefault(React);

const DynamicComponent = ({ componentUid, disabled, index, name, onRemoveComponentClick, onMoveComponent, onGrabItem, onDropItem, onCancel, dynamicComponentsByCategory = {}, onAddComponent, totalLength, children })=>{
    const { formatMessage } = reactIntl.useIntl();
    const { currentDocumentMeta } = useDocumentContext.useDocumentContext('DynamicComponent');
    const isDesktop = strapiAdmin.useIsDesktop();
    const { edit: { components } } = useDocumentLayout.useDocumentLayout(currentDocumentMeta.model);
    const { mainField = 'id' } = components[componentUid]?.settings ?? {};
    const mainFieldValue = strapiAdmin.useForm('DynamicComponent', (state)=>objects.getIn(state.values, `${name}.${index}.${mainField}`));
    const displayedValue = mainField === 'id' || !mainFieldValue ? '' : String(mainFieldValue).trim();
    const displayTitle = displayedValue.length > 0 ? `- ${displayedValue}` : displayedValue;
    const [category] = componentUid.split('.');
    const { icon, displayName } = (dynamicComponentsByCategory[category] ?? []).find((component)=>component.uid === componentUid) ?? {
        icon: null,
        displayName: null
    };
    const [{ handlerId, isDragging, handleKeyDown }, boxRef, dropRef, dragRef, dragPreviewRef] = useDragAndDrop.useDragAndDrop(!disabled, {
        type: `${dragAndDrop.ItemTypes.DYNAMIC_ZONE}_${name}`,
        index,
        item: {
            index,
            displayedValue: `${displayName} ${displayTitle}`,
            icon
        },
        onMoveItem: onMoveComponent,
        onDropItem,
        onGrabItem,
        onCancel
    });
    React__namespace.useEffect(()=>{
        dragPreviewRef(reactDndHtml5Backend.getEmptyImage(), {
            captureDraggingState: false
        });
    }, [
        dragPreviewRef,
        index
    ]);
    /**
   * We don't need the accordion's to communicate with each other,
   * so a unique value for their state is enough.
   */ const accordionValue = React__namespace.useId();
    const componentPath = `${name}.${index}`;
    const hasValue = strapiAdmin.useForm('DynamicComponent', (state)=>objects.getIn(state.values, componentPath) != null);
    const isNewItem = strapiAdmin.useForm('DynamicComponent', (state)=>objects.getIn(state.values, componentPath)?.id == null);
    const rawError = strapiAdmin.useForm('DynamicComponent', (state)=>objects.getIn(state.errors, componentPath));
    const [collapseToOpen, setCollapseToOpen] = React__namespace.useState(isNewItem ? accordionValue : '');
    React__namespace.useEffect(()=>{
        if (rawError && hasValue) {
            setCollapseToOpen(accordionValue);
        }
    }, [
        rawError,
        hasValue,
        accordionValue
    ]);
    const composedBoxRefs = designSystem.useComposedRefs(boxRef, dropRef);
    const canMoveUp = index > 0;
    const canMoveDown = index < totalLength - 1;
    const handleRemoveCurrentComponent = React__namespace.useCallback(()=>{
        onRemoveComponentClick(index);
    }, [
        onRemoveComponentClick,
        index
    ]);
    const accordionActions = disabled ? null : /*#__PURE__*/ jsxRuntime.jsxs(jsxRuntime.Fragment, {
        children: [
            /*#__PURE__*/ jsxRuntime.jsx(designSystem.IconButton, {
                variant: "ghost",
                label: formatMessage({
                    id: translations.getTranslation('components.DynamicZone.delete-label'),
                    defaultMessage: 'Delete {name}'
                }, {
                    name: displayTitle
                }),
                onClick: handleRemoveCurrentComponent,
                children: /*#__PURE__*/ jsxRuntime.jsx(Icons.Trash, {})
            }),
            isDesktop && /*#__PURE__*/ jsxRuntime.jsx(designSystem.IconButton, {
                variant: "ghost",
                onClick: (e)=>e.stopPropagation(),
                "data-handler-id": handlerId,
                ref: dragRef,
                label: formatMessage({
                    id: translations.getTranslation('components.DragHandle-label'),
                    defaultMessage: 'Drag'
                }),
                onKeyDown: handleKeyDown,
                children: /*#__PURE__*/ jsxRuntime.jsx(Icons.Drag, {})
            }),
            !isDesktop && /*#__PURE__*/ jsxRuntime.jsxs(jsxRuntime.Fragment, {
                children: [
                    canMoveUp && /*#__PURE__*/ jsxRuntime.jsx(designSystem.IconButton, {
                        variant: "ghost",
                        onClick: (e)=>{
                            e.stopPropagation();
                            onMoveComponent(index - 1, index);
                        },
                        disabled: !canMoveUp,
                        label: formatMessage({
                            id: translations.getTranslation('components.DynamicZone.move-up'),
                            defaultMessage: 'Move up'
                        }),
                        children: /*#__PURE__*/ jsxRuntime.jsx(Icons.ArrowUp, {})
                    }),
                    canMoveDown && /*#__PURE__*/ jsxRuntime.jsx(designSystem.IconButton, {
                        variant: "ghost",
                        onClick: (e)=>{
                            e.stopPropagation();
                            onMoveComponent(index + 1, index);
                        },
                        disabled: !canMoveDown,
                        label: formatMessage({
                            id: translations.getTranslation('components.DynamicZone.move-down'),
                            defaultMessage: 'Move down'
                        }),
                        children: /*#__PURE__*/ jsxRuntime.jsx(Icons.ArrowDown, {})
                    })
                ]
            }),
            /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Menu.Root, {
                children: [
                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Menu.Trigger, {
                        size: "S",
                        endIcon: null,
                        paddingLeft: 0,
                        paddingRight: 0,
                        children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.IconButton, {
                            variant: "ghost",
                            label: formatMessage({
                                id: translations.getTranslation('components.DynamicZone.more-actions'),
                                defaultMessage: 'More actions'
                            }),
                            tag: "span",
                            children: /*#__PURE__*/ jsxRuntime.jsx(Icons.More, {
                                "aria-hidden": true,
                                focusable: false
                            })
                        })
                    }),
                    /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Menu.Content, {
                        children: [
                            /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Menu.SubRoot, {
                                children: [
                                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Menu.SubTrigger, {
                                        children: formatMessage({
                                            id: translations.getTranslation('components.DynamicZone.add-item-above'),
                                            defaultMessage: 'Add component above'
                                        })
                                    }),
                                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Menu.SubContent, {
                                        children: Object.entries(dynamicComponentsByCategory).map(([category, components])=>/*#__PURE__*/ jsxRuntime.jsxs(React__namespace.Fragment, {
                                                children: [
                                                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Menu.Label, {
                                                        children: category
                                                    }),
                                                    components.map(({ displayName, uid })=>/*#__PURE__*/ jsxRuntime.jsx(designSystem.Menu.Item, {
                                                            onSelect: ()=>onAddComponent(uid, index),
                                                            children: displayName
                                                        }, uid))
                                                ]
                                            }, category))
                                    })
                                ]
                            }),
                            /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Menu.SubRoot, {
                                children: [
                                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Menu.SubTrigger, {
                                        children: formatMessage({
                                            id: translations.getTranslation('components.DynamicZone.add-item-below'),
                                            defaultMessage: 'Add component below'
                                        })
                                    }),
                                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Menu.SubContent, {
                                        children: Object.entries(dynamicComponentsByCategory).map(([category, components])=>/*#__PURE__*/ jsxRuntime.jsxs(React__namespace.Fragment, {
                                                children: [
                                                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Menu.Label, {
                                                        children: category
                                                    }),
                                                    components.map(({ displayName, uid })=>/*#__PURE__*/ jsxRuntime.jsx(designSystem.Menu.Item, {
                                                            onSelect: ()=>onAddComponent(uid, index + 1),
                                                            children: displayName
                                                        }, uid))
                                                ]
                                            }, category))
                                    })
                                ]
                            })
                        ]
                    })
                ]
            })
        ]
    });
    const accordionTitle = displayTitle ? `${displayName} ${displayTitle}` : displayName;
    return /*#__PURE__*/ jsxRuntime.jsxs(ComponentContainer, {
        tag: "li",
        width: "100%",
        children: [
            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
                justifyContent: "center",
                children: /*#__PURE__*/ jsxRuntime.jsx(Rectangle, {
                    background: "neutral200"
                })
            }),
            /*#__PURE__*/ jsxRuntime.jsx(StyledBox, {
                ref: composedBoxRefs,
                hasRadius: true,
                children: isDragging ? /*#__PURE__*/ jsxRuntime.jsx(Preview, {}) : /*#__PURE__*/ jsxRuntime.jsx(designSystem.Accordion.Root, {
                    value: collapseToOpen,
                    onValueChange: setCollapseToOpen,
                    children: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Accordion.Item, {
                        value: accordionValue,
                        children: [
                            /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Accordion.Header, {
                                children: [
                                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Accordion.Trigger, {
                                        icon: icon && ComponentIcon.COMPONENT_ICONS[icon] ? ComponentIcon.COMPONENT_ICONS[icon] : ComponentIcon.COMPONENT_ICONS.dashboard,
                                        children: accordionTitle
                                    }),
                                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Accordion.Actions, {
                                        children: accordionActions
                                    })
                                ]
                            }),
                            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Accordion.Content, {
                                children: /*#__PURE__*/ jsxRuntime.jsx(AccordionContentRadius, {
                                    background: "neutral0",
                                    children: /*#__PURE__*/ jsxRuntime.jsx(DynamicComponentFields, {
                                        componentUid: componentUid,
                                        index: index,
                                        layout: components[componentUid]?.layout,
                                        name: name,
                                        children: children
                                    })
                                })
                            })
                        ]
                    })
                })
            })
        ]
    });
};
// TODO: Delete once https://github.com/strapi/design-system/pull/858
// is merged and released.
const StyledBox = styledComponents.styled(designSystem.Box)`
  > div:first-child {
    box-shadow: ${({ theme })=>theme.shadows.tableShadow};
  }
`;
const AccordionContentRadius = styledComponents.styled(designSystem.Box)`
  border-radius: 0 0 ${({ theme })=>theme.spaces[1]} ${({ theme })=>theme.spaces[1]};
`;
const Rectangle = styledComponents.styled(designSystem.Box)`
  width: ${({ theme })=>theme.spaces[2]};
  height: ${({ theme })=>theme.spaces[4]};
`;
const Preview = styledComponents.styled.span`
  display: block;
  background-color: ${({ theme })=>theme.colors.primary100};
  outline: 1px dashed ${({ theme })=>theme.colors.primary500};
  outline-offset: -1px;
  padding: ${({ theme })=>theme.spaces[6]};
`;
const ComponentContainer = styledComponents.styled(designSystem.Box)`
  list-style: none;
  padding: 0;
  margin: 0;
`;
const DynamicComponentFields = /*#__PURE__*/ React__namespace.memo(({ children, componentUid, index, layout, name })=>{
    const { formatMessage } = reactIntl.useIntl();
    return /*#__PURE__*/ jsxRuntime.jsx(designSystem.Box, {
        padding: {
            initial: 4,
            medium: 6
        },
        children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Grid.Root, {
            gap: 4,
            children: layout?.map((row, rowInd)=>{
                return /*#__PURE__*/ jsxRuntime.jsx(designSystem.Grid.Item, {
                    col: 12,
                    xs: 12,
                    direction: "column",
                    alignItems: "stretch",
                    children: /*#__PURE__*/ jsxRuntime.jsx(FormLayout.ResponsiveGridRoot, {
                        gap: 4,
                        children: row.map(({ size, ...field })=>{
                            const fieldName = `${name}.${index}.${field.name}`;
                            const fieldWithTranslatedLabel = {
                                ...field,
                                label: formatMessage({
                                    id: `content-manager.components.${componentUid}.${field.name}`,
                                    defaultMessage: field.label
                                })
                            };
                            return /*#__PURE__*/ jsxRuntime.jsx(FormLayout.ResponsiveGridItem, {
                                col: size,
                                s: 12,
                                xs: 12,
                                direction: "column",
                                alignItems: "stretch",
                                children: children ? children({
                                    ...fieldWithTranslatedLabel,
                                    name: fieldName
                                }) : /*#__PURE__*/ jsxRuntime.jsx(InputRenderer.InputRenderer, {
                                    ...fieldWithTranslatedLabel,
                                    name: fieldName
                                })
                            }, fieldName);
                        })
                    })
                }, rowInd);
            })
        })
    });
});
DynamicComponentFields.displayName = 'DynamicComponentFields';
const MemoizedDynamicComponent = /*#__PURE__*/ React__namespace.memo(DynamicComponent);

exports.DynamicComponent = MemoizedDynamicComponent;
