// @strapi/upload/dist/admin/future/pages/Assets/components/CreateFolderDialog.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var strapiAdmin = require('@strapi/admin/strapi-admin');
var designSystem = require('@strapi/design-system');
var reactIntl = require('react-intl');
var styledComponents = require('styled-components');
var folders = require('../../../services/folders.js');
var translations = require('../../../utils/translations.js');

const StyledModalContent = styledComponents.styled(designSystem.Modal.Content)`
  max-width: 51.6rem;
`;
const CreateFolderDialog = ({ open, folderName, parentFolderId, onClose })=>{
    const { formatMessage } = reactIntl.useIntl();
    const { toggleNotification } = strapiAdmin.useNotification();
    const [name, setName] = React.useState('');
    const [fieldError, setFieldError] = React.useState();
    const [createFolder, { isLoading }] = folders.useCreateFolderMutation();
    React.useEffect(()=>{
        if (open) {
            setName('');
            setFieldError(undefined);
        }
    }, [
        open
    ]);
    const handleSubmit = async (e)=>{
        e.preventDefault();
        const trimmedName = name.trim();
        if (!trimmedName) {
            setFieldError(formatMessage({
                id: translations.getTranslationKey('folder.create.form.error.name-required'),
                defaultMessage: 'Name is required'
            }));
            return;
        }
        try {
            await createFolder({
                name: trimmedName,
                parent: parentFolderId
            }).unwrap();
            toggleNotification({
                type: 'success',
                message: formatMessage({
                    id: translations.getTranslationKey('folder.create.success'),
                    defaultMessage: 'Folder has been created'
                })
            });
            onClose();
        } catch (err) {
            const apiError = err;
            if (apiError?.message) {
                setFieldError(apiError.message);
            } else {
                toggleNotification({
                    type: 'danger',
                    message: formatMessage({
                        id: translations.getTranslationKey('folder.create.form.error.unknown'),
                        defaultMessage: 'An error occurred while creating the folder'
                    })
                });
            }
        }
    };
    return /*#__PURE__*/ jsxRuntime.jsx(designSystem.Modal.Root, {
        open: open,
        onOpenChange: onClose,
        children: /*#__PURE__*/ jsxRuntime.jsxs(StyledModalContent, {
            children: [
                /*#__PURE__*/ jsxRuntime.jsx(designSystem.Modal.Header, {
                    children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Modal.Title, {
                        children: formatMessage({
                            id: translations.getTranslationKey('folder.create.title-in'),
                            defaultMessage: 'New folder in {folderName}'
                        }, {
                            folderName
                        })
                    })
                }),
                /*#__PURE__*/ jsxRuntime.jsxs("form", {
                    onSubmit: handleSubmit,
                    children: [
                        /*#__PURE__*/ jsxRuntime.jsx(designSystem.Modal.Body, {
                            children: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Field.Root, {
                                error: fieldError,
                                name: "name",
                                required: true,
                                children: [
                                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Field.Label, {
                                        children: formatMessage({
                                            id: translations.getTranslationKey('folder.form.name.label'),
                                            defaultMessage: 'Folder name'
                                        })
                                    }),
                                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.TextInput, {
                                        value: name,
                                        onChange: (e)=>{
                                            setName(e.target.value);
                                            setFieldError(undefined);
                                        },
                                        autoFocus: true
                                    }),
                                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Field.Error, {})
                                ]
                            })
                        }),
                        /*#__PURE__*/ jsxRuntime.jsx(designSystem.Modal.Footer, {
                            children: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                                gap: 2,
                                justifyContent: "space-between",
                                width: "100%",
                                children: [
                                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Button, {
                                        variant: "tertiary",
                                        onClick: onClose,
                                        type: "button",
                                        children: formatMessage({
                                            id: 'app.components.Button.cancel',
                                            defaultMessage: 'Cancel'
                                        })
                                    }),
                                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Button, {
                                        type: "submit",
                                        loading: isLoading,
                                        children: formatMessage({
                                            id: translations.getTranslationKey('folder.create.submit'),
                                            defaultMessage: 'Create folder'
                                        })
                                    })
                                ]
                            })
                        })
                    ]
                })
            ]
        })
    });
};

exports.CreateFolderDialog = CreateFolderDialog;
